{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Gene order and gene functional category abundances\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Gita/anaconda/lib/python3.4/site-packages/matplotlib/__init__.py:872: UserWarning: axes.color_cycle is deprecated and replaced with axes.prop_cycle; please use the latter.\n",
      "  warnings.warn(self.msg_depr % (key, alt_key))\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import ete3 as ete\n",
    "import os\n",
    "import Bio\n",
    "import random\n",
    "import scipy.stats as stats\n",
    "from functools import *\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "import matplotlib.patches as mpatches\n",
    "import seaborn as sns\n",
    "sns.set_context(\"paper\")\n",
    "%matplotlib inline\n",
    "\n",
    "\n",
    "#as part of quality control the following viruses had to be omitted from our analysis\n",
    "#they contain misannotations or mislabeled hosts, etc. \n",
    "badFolders = ['Acyrthosiphon_pisum_secondary_endosymbiont_phage_1_uid14047', 'Aureococcus_anophagefferens_virus_MM_2014_uid258005', 'Dragonfly_associated_microphage_1_uid177547', 'Enterobacter_phage_PG7_uid240014', 'Escherichia_phage_vB_EcoM_JS09_uid248321', 'Lactobacillus_phage_LL_H_uid19803', 'Providencia_phage_Redjac_uid177540', 'Pseudoalteromonas_phage_RIO_1_uid206039', 'Sputnik_virophage_2_uid243067', 'Sputnik_virophage_3_uid243065', 'Staphylococcus_phage_6ec_uid253318', 'Streptococcus_phage_DCC1738_uid253219', 'Streptococcus_phage_IC1_uid253220', 'Streptococcus_phage_K13_uid253223', 'Streptococcus_phage_SMP_uid18529', 'Vibrio_phage_CP_T1_uid181062', 'Vibrio_phage_vpms1_uid212709', 'Vibrio_phage_VPUSM_8_uid227006',  'Abaca_bunchy_top_virus_uid28697','Shallot_yellow_stripe_virus_uid15745','Equine_rhinitis_A_virus_uid15205','African_pouched_rat_arterivirus_uid274797','Spilanthes_yellow_vein_virus_uid19779', 'Velvet_bean_severe_mosaic_virus_uid41175','Paramecium_bursaria_Chlorella_virus_FR483_uid18305','Human_endogenous_retrovirus_K113_uid222261','Acholeplasma_phage_MV_L1_uid14573']\n",
    "#viruses whose lineage is improperly annotated (they don't belong to any Baltimore group)\n",
    "unknowns= ['Bacillus_phage_phBC6A51_uid15021', 'Bacillus_phage_phBC6A52_uid15022', 'Badger_feces_associated_gemycircularvirus_uid281125', 'Chaetoceros_DNA_virus_7_uid237558', 'Chaetoceros_lorenzianus_DNA_Virus_uid63565', 'Chaetoceros_tenuissimus_DNA_virus_uid60753', 'Cladosporium_cladosporioides_virus_1_uid258308', 'Enterobacteria_phage_YYZ_2008_uid32231', 'European_mountain_ash_ringspot_associated_virus_uid39973', 'Faecal_associated_gemycircularvirus_3_uid268557', 'Faecal_associated_gemycircularvirus_4_uid268556', 'Faecal_associated_gemycircularvirus_5_uid268555', 'Faecal_associated_gemycircularvirus_6_uid268554', 'Faecal_associated_gemycircularvirus_7_uid268553', 'Gentian_ovary_ring_spot_virus_uid256090', 'Geobacillus_virus_E2_uid19797', 'Glossina_pallidipes_salivary_gland_hypertrophy_virus_uid28839', 'Haloarcula_hispanica_icosahedral_virus_2_uid109269', 'Haloarcula_phage_SH1_uid15535', 'Halovirus_PH1_uid196975', 'Halovirus_VNH_1_uid262927', 'Human_genital_associated_circular_DNA_virus_1_uid281399', 'Lactobacillus_johnsonii_prophage_Lj771_uid28145', 'Magnaporthe_oryzae_RNA_virus_uid272442', 'Mongoose_feces_associated_gemycircularvirus_a_uid281407', 'Mongoose_feces_associated_gemycircularvirus_b_uid281406', 'Mongoose_feces_associated_gemycircularvirus_c_uid281404', 'Mongoose_feces_associated_gemycircularvirus_d_uid281405', 'Mycoplasma_phage_phiMFV1_uid14387', 'Penicillium_roqueforti_dsRNA_mycovirus_1_uid258162', 'Phage_Gifsy_1_uid32269', 'Phage_Gifsy_2_uid32271', 'Pigeonpea_cryptic_virus_uid244664', 'Planaria_asexual_strain_specific_virus_like_element_type_1_uid14140', 'Pseudomonas_phage_phi_Pto_bp6g_uid240724', 'Rhizoctonia_fumigata_mycovirus_uid283068', 'Rhodococcus_phage_REQ2_uid81171', 'Rhodococcus_phage_REQ3_uid81175', 'Rose_rosette_virus_uid64937', 'Rosellinia_necatrix_fusarivirus_1_uid255787', 'Rosellinia_necatrix_megabirnavirus_1_W779_uid41609', 'Salisaeta_icosahedral_phage_1_uid167575', 'Salmonella_phage_Fels_1_uid29267', 'Sodalis_phage_phiSG1_uid16583', 'Staphylococcus_phage_phi2958PVL_uid32173', 'Staphylococcus_phage_tp310_1_uid20659', 'Staphylococcus_phage_tp310_3_uid20663', 'Stenotrophomonas_phage_phiSMA9_uid15493', 'Streptococcus_phage_20617_uid239271', 'Streptococcus_phage_phiBHN167_uid227353', 'Streptococcus_pyogenes_phage_315_1_uid14533', 'Streptococcus_pyogenes_phage_315_2_uid14528', 'Streptococcus_pyogenes_phage_315_3_uid14529', 'Streptococcus_pyogenes_phage_315_4_uid14530', 'Streptococcus_pyogenes_phage_315_5_uid14531', 'Streptococcus_pyogenes_phage_315_6_uid14532', 'Tanay_virus_uid246129', 'Thermococcus_prieurii_virus_1_uid84407', 'Thermus_phage_IN93_uid14235', 'Thermus_phage_P23_77_uid40235']\n",
    "badFolders=badFolders+unknowns\n",
    "#these are satellite viruses \n",
    "sats = ['Ageratum_conyzoides_associated_symptomless_alphasatellite_uid259293', 'Ageratum_enation_alphasatellite_uid181994', 'Ageratum_leaf_curl_betasatellite_uid195929', 'Ageratum_leaf_curl_Cameroon_betasatellite_uid36669', 'Ageratum_yellow_leaf_curl_betasatellite_uid14439', 'Ageratum_yellow_vein_China_alphasatellite_uid237561', 'Ageratum_yellow_vein_China_virus_associated_DNA_beta_uid15515', 'Ageratum_yellow_vein_Singapore_alphasatellite_uid14232', 'Ageratum_yellow_vein_virus_satellite_DNA_beta_uid14444', 'Alternanthera_yellow_vein_virus_satellite_DNA_beta_uid19833', 'Andrographis_yellow_vein_leaf_curl_betasatellite_uid243492', 'Bhendi_yellow_vein_India_betasatellite__India_Aurangabad_OY164_2006__uid61557', 'Bhendi_yellow_vein_mosaic_betasatellite__India_Coimbator_OYCO1_2005__uid61777', 'Bhendi_yellow_vein_mosaic_virus_associated_alphasatellite_uid174781', 'Bhendi_yellow_vein_mosaic_virus_satellite_DNA_beta_uid14445', 'Black_medic_leafroll_alphasatellite_1_uid243500', 'Cardamom_bushy_dwarf_virus_satellite_uid230064', 'Cassava_mosaic_Madagascar_alphasatellite_uid175666', 'Chili_leaf_curl_Bhatinda_betasatellite_uid206467', 'Chilli_leaf_curl_alphasatellite_uid253224', 'Chilli_leaf_curl_Multan_alphasatellite_uid39933', 'Chilli_leaf_curl_virus_satellite_DNA_beta_uid14441', 'Cleome_leaf_crumple_virus_associated_DNA_1_uid60045', 'Corchorus_yellow_vein_mosaic_betasatellite_uid192608', 'Cotton_leaf_curl_Burewala_alphasatellite_uid45935', 'Cotton_leaf_curl_Burewala_betasatellite_uid45933', 'Cotton_leaf_curl_Gezira_alphasatellite_uid42507', 'Cotton_leaf_curl_Gezira_virus_satellite_DNA_beta_uid15166', 'Cotton_leaf_curl_Multan_betasatellite_uid15780', 'Cotton_leaf_curl_virus_associated_DNA_1_isolate_Lucknow_uid65305', 'Cotton_leaf_curl_virus_associated_DNA_beta_uid14438', 'Cotton_leaf_curl_virus_betasatellite_uid162497', 'Cowpea_severe_leaf_curl_associated_DNA_beta_uid15157', 'Croton_yellow_vein_mosaic_alphasatellite_uid45931', 'Croton_yellow_vein_mosaic_betasatellite_uid18249', 'Cuban_alphasatellite_1_uid210798', 'Dragonfly_associated_alphasatellite_uid181244', 'Emilia_yellow_vein_virus_associated_DNA_beta_uid37893', 'Erectites_yellow_mosaic_virus_satellite_DNA_beta_uid19827', 'Eupatorium_yellow_vein_virus_satellite_DNA_beta_uid14447', 'Faba_bean_necrotic_stunt_alphasatellite_1_uid243499', 'Faba_bean_necrotic_stunt_alphasatellite_2_uid243498', 'French_bean_leaf_curl_betasatellite_Kanpur_uid169556', 'Gossypium_darwinii_symptomless_alphasatellite_uid39593', 'Gossypium_davidsonii_symptomless_alphasatellite_uid39589', 'Gossypium_mustilinum_symptomless_alphasatellite_uid39591', 'Grapevine_satellite_virus_uid208539', 'Guar_leaf_curl_alphasatellite_uid193981', 'Hedyotis_uncinella_yellow_mosaic_betasatellite_uid230991', 'Honeysuckle_yellow_vein_mosaic_disease_associated_satellite_DNA_beta_uid19863', 'Honeysuckle_yellow_vein_mosaic_virus_satellite_DNA_beta_uid14620', 'Leucas_zeylanica_yellow_vein_virus_satellite_DNA_beta_uid41305', 'Ludwigia_leaf_distortion_betasatellite__India_Amadalavalasa_Hibiscus_2007__uid29233', 'Ludwigia_yellow_vein_virus_associated_DNA_beta_uid15561', 'Luffa_puckering_and_leaf_distortion_associated_betasatellite__India_Gurdaspur_Okra_2013__uid253218', 'Luffa_puckering_and_leaf_distortion_associated_DNA_beta_uid15779', 'Malachra_yellow_vein_mosaic_virus_associated_satellite_DNA_beta_uid28727', 'Malvastrum_leaf_curl_betasatellite_uid16301', 'Malvastrum_leaf_curl_Guangdong_betasatellite_uid243827', 'Malvastrum_yellow_vein_Yunnan_virus_satellite_DNA_beta_uid14567', 'Melon_chlorotic_mosaic_virus_associated_alphasatellite_uid51413', 'Mesta_yellow_vein_mosaic_virus_alphasatellite__India_Hoshiarpur_Okra_2013__uid253217', 'Mesta_yellow_vein_mosaic_virus_associated_alphasatellite_uid174780', 'Mesta_yellow_vein_mosaic_virus_associated_DNA_beta_uid21015', 'Milk_vetch_dwarf_C10_alphasatellite_uid240682', 'Mirabilis_leaf_curl_India_virus_associated_betasatellite_uid255800', 'Mungbean_yellow_mosaic_India_virus_associated_betasatellite__India__Faizabad__Cow_Pea_2012__uid177773', 'Nanovirus_like_particle_uid14386', 'Okra_enation_leaf_curl_betasatellite__India_Sonipat_EL10_2006__uid61781', 'Okra_leaf_curl_Mali_virus_satellite_DNA_beta_uid20323', 'Okra_yellow_vein_disease_associated_sequence_uid14443', 'Olive_viral_satellite_RNA_uid230268', 'Papaya_leaf_curl_alphasatellite_uid233414', 'Papaya_leaf_curl_China_virus_satellite_DNA_beta_uid19819', 'Pepper_leaf_curl_virus_satellite_DNA_beta_uid28283', 'Pepper_leaf_curl_Yunnan_virus_satellite_DNA_beta_uid29415', 'Potato_apical_leaf_curl_disease_associated_satellite_DNA_beta_uid18323', 'Radish_leaf_curl_virus_betasatellite_uid28281', 'Sida_leaf_curl_virus_satellite_DNA_beta_uid19823', 'Sida_yellow_mosaic_virus_China_associated_DNA_beta_uid15514', 'Sida_yellow_vein_China_alphasatellite_uid238950', 'Siegesbeckia_yellow_vein_virus_associated_DNA_beta_uid17269', 'Subterranean_clover_stunt_C6_alphasatellite_uid14180', 'Synedrella_leaf_curl_alphasatellite_uid263034', 'Tobacco_leaf_chlorosis_betasatellite_uid178075', 'Tobacco_leaf_curl_disease_associated_sequence_uid14442', 'Tobacco_leaf_curl_PUSA_alphasatellite_uid56023', 'Tobacco_leaf_curl_virus_associated_DNA_beta_uid45925', 'Tomato_leaf_curl_Bangladesh_betasatellite_uid56017', 'Tomato_leaf_curl_Gandhinagar_betasatellite_uid231683', 'Tomato_leaf_curl_Hajipur_betasatellite_uid175587', 'Tomato_leaf_curl_Joydebpur_virus_satellite_DNA_beta_uid28273', 'Tomato_leaf_curl_Karnataka_alphasatellite_uid181995', 'Tomato_leaf_curl_Pakistan_virus_associated_DNA_1_uid38463', 'Tomato_leaf_curl_Patna_betasatellite_uid36541', 'Tomato_leaf_curl_Togo_betasatellite_Togo_2006__uid60629', 'Tomato_leaf_curl_Yemen_betasatellite_uid177643', 'Tomato_yellow_leaf_curl_China_betasatellite_uid181248', 'Tomato_yellow_leaf_curl_Vietnam_virus_satellite_DNA_beta_uid19829', 'TYLCCNV_Y322__satellite_DNA_beta_uid16338', 'Vernonia_yellow_vein_betasatellite_uid41303', 'Vernonia_yellow_vein_Fujian_virus_alphasatellite_uid72145', 'Vernonia_yellow_vein_Fujian_virus_betasatellite_uid72143', 'Zinnia_leaf_curl_disease_associated_sequence_uid14440', 'Zinnia_leaf_curl_virus_associated_DNA_beta_uid14538']\n",
    "\n",
    "\n",
    "def getTaxidFromGenomeFolder(genome_folder):\n",
    "    #Takes an rpt file for a genome and returns the taxid in it.\n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        with open(os.path.join('all_rpt', folder, os.listdir(os.path.join('all_rpt', folder))[0])) as rpt_file:\n",
    "            for line in rpt_file:\n",
    "                if line.startswith('Taxid: '):\n",
    "                    return int(line[7:-1])\n",
    "                \n",
    "\n",
    "def getHostTaxidFromGenomeFolder(genome_folder):\n",
    "    #Takes a folder containing an asn file and returns the taxid of the host of the virus described by that asn file. \n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        with open('folderToHostTaxid.txt', 'r') as taxid2host:\n",
    "            for line in taxid2host:\n",
    "                if genome_folder in line:\n",
    "                    return int(line.split('\\t')[1][:-1])\n",
    "        return 0\n",
    "    \n",
    "\n",
    "def getGenomeLengthFromGenomeFolder(genome_folder):\n",
    "    total = 0\n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        for file in os.listdir(os.path.join('all_ptt', folder)):\n",
    "            with open(os.path.join('all_ptt', folder, file)) as ptt_file:\n",
    "                line = ptt_file.readline()\n",
    "                index = line.find('- 1..')\n",
    "                if index == -1:\n",
    "                    return -1\n",
    "                num = line[index + 5 : -1]\n",
    "                total += int(num)\n",
    "        return total\n",
    "    \n",
    "\n",
    "def getCodingNumberAndOverlapNumberFromGenomeFolder(folder):\n",
    "    totalCoding = 0\n",
    "    totalOverlap = 0\n",
    "    for file in os.listdir(os.path.join('all_ptt', folder)):\n",
    "        #these folders contain genomes with misannotation of gene overlap,non-coding DNA, or host information from NCBI. \n",
    "        if folder not in badFolders and folder not in sats:\n",
    "            with open(os.path.join('all_ptt', folder, file)) as ptt_file:\n",
    "                line = ptt_file.readline()\n",
    "                index = line.find('- 1..')\n",
    "                repliconLength = int(line[index + 5 : -1])\n",
    "                indicesThatOverlap = [0] * repliconLength\n",
    "                line = ptt_file.readline()\n",
    "                line = ptt_file.readline()\n",
    "\n",
    "                uniqueIndices = set([])\n",
    "                for line in ptt_file:                                     \n",
    "                    #this part is to collect beginning and end indices for genes\n",
    "                    line = line.split('\\t')[0]\n",
    "                    [start, end] = [int(x) for x in line.split('..')]\n",
    "                    #this is put in place to distinguish linear from circular genomes. Here, we're looking for linear genomes\n",
    "                    if start < end:                    \n",
    "                        for i in range(start, end):\n",
    "                            if i in uniqueIndices:\n",
    "                                indicesThatOverlap[i] = 1\n",
    "                            else:\n",
    "                                uniqueIndices.add(i)\n",
    "                    #here were are looking for circular genomes to be able to properly count noncoding bases\n",
    "                    elif start > end:\n",
    "                        for i in range(start, repliconLength):\n",
    "                            if i in uniqueIndices:\n",
    "                                indicesThatOverlap[i] = 1\n",
    "                            else:\n",
    "                                uniqueIndices.add(i)\n",
    "                        for i in range(1, end):\n",
    "                            if i in uniqueIndices:\n",
    "                                indicesThatOverlap[i] = 1\n",
    "                            else:\n",
    "                                uniqueIndices.add(i)          \n",
    "                \n",
    "                totalCoding += len(uniqueIndices)\n",
    "                #totalOverlap += len(bigList) - len(uniqueIndices)\n",
    "                totalOverlap += sum(indicesThatOverlap)\n",
    "    return (totalCoding, totalOverlap)\n",
    "\n",
    "\n",
    "def getProtienNamesFromGenomeFolder(folder):\n",
    "    protVec=[]\n",
    "    for file in os.listdir(os.path.join('all_ptt', folder)):\n",
    "        #these folders contain genomes with misannotation of gene overlap, non-coding DNA, or host information from NCBI. \n",
    "        if folder not in badFolders and folder not in sats:\n",
    "            with open(os.path.join('all_ptt', folder, file)) as ptt_file:\n",
    "                line = ptt_file.readline()\n",
    "                line = ptt_file.readline()\n",
    "                line = ptt_file.readline()               \n",
    "                for line in ptt_file:                    \n",
    "                    #this is to collect the name of proteins from ptt files                   \n",
    "                    protVec.append(line.split('\\t')[-1].rstrip())                                    \n",
    "    return (protVec)\n",
    "\n",
    "\n",
    "genomeLengths = []\n",
    "lengthData = []\n",
    "counter = 0\n",
    "length_lists = []\n",
    "hostTaxidList = []\n",
    "taxidList = []\n",
    "folders = []\n",
    "proteinNames=[]\n",
    "dbObject = ete.NCBITaxa()\n",
    "\n",
    "for folder in os.listdir('all_faa'):\n",
    "    #these entries had problems with their gene overlap, protein, and noncoding annotation so they're excluded\n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        length_lists.append([])\n",
    "        for fileName in os.listdir(os.path.join('all_faa', folder)):\n",
    "            with open(os.path.join('all_faa', folder, fileName), \"r\") as faa_file:\n",
    "                proteins = SeqIO.parse(faa_file, 'fasta')\n",
    "                for protein in proteins:\n",
    "                    length_lists[counter].append(float(len(protein.seq)))\n",
    "\n",
    "        hostTaxid = getHostTaxidFromGenomeFolder(folder)\n",
    "        # These taxids come viruses with unidentified hosts \n",
    "        if hostTaxid and hostTaxid != 11086 and hostTaxid != 10036 and hostTaxid != 37965 and hostTaxid != 10640  and hostTaxid != 32644 and hostTaxid != 1 and hostTaxid != 212035 and hostTaxid != 1051671:\n",
    "            taxidList.append(getTaxidFromGenomeFolder(folder))\n",
    "            hostTaxidList.append(hostTaxid)\n",
    "            genomeLengths.append(getGenomeLengthFromGenomeFolder(folder))\n",
    "            folders.append(folder)\n",
    "            proteinNames.append(getProtienNamesFromGenomeFolder(folder))\n",
    "            counter += 1\n",
    "        else:\n",
    "            length_lists = length_lists[:-1]\n",
    "        \n",
    "        #Uncomment the following lines to see the code's progress while its running\n",
    "        \"\"\"\n",
    "        if counter % 100 == 0:\n",
    "            print(str(int(np.round(float(counter) * 100.0 / 2701.0))) + '% done.')\n",
    "        \"\"\"\n",
    "\n",
    "taxidLineages = [dbObject.get_lineage(t) for t in taxidList]\n",
    "englishLineages = [[str(rank) for rank in dbObject.translate_to_names(lin)]\n",
    "                   for lin in taxidLineages]\n",
    "hostTaxidLineages = [dbObject.get_lineage(t) for t in hostTaxidList]\n",
    "hostEnglishLineages = [[str(rank) for rank in dbObject.translate_to_names(lin)] \n",
    "                       for lin in hostTaxidLineages]\n",
    "\n",
    "for i in range(len(length_lists)):\n",
    "    lengthData.append({'length_list' : length_lists[i],\n",
    "                       'taxid' : taxidList[i],\n",
    "                       'hostTaxid' : hostTaxidList[i],\n",
    "                       'englishLineage' : englishLineages[i],\n",
    "                       'taxidLineage' : taxidLineages[i],\n",
    "                       'hostEnglishLineage' : hostEnglishLineages[i],\n",
    "                       'hostTaxidLineage' : hostTaxidLineages[i],\n",
    "                       'mean' : np.mean(length_lists[i]),\n",
    "                       'std' : np.std(length_lists[i]),\n",
    "                       'median' : np.median(length_lists[i]),\n",
    "                       'numProteins' : len(length_lists[i]),\n",
    "                       'genomeLength' : genomeLengths[i],\n",
    "                       'proteinNames': proteinNames[i],\n",
    "                       'folder' : folders[i]})\n",
    "                       \n",
    "lengthData = [virus for virus in lengthData if virus['genomeLength'] != -1 and virus['median'] != -1]\n",
    "\n",
    "for (i, virus) in enumerate(lengthData):\n",
    "    #print(str(i) + '/' + str(len(lengthData)))\n",
    "    (totalCoding, totalOverlap) = getCodingNumberAndOverlapNumberFromGenomeFolder(virus['folder'])\n",
    "    virus['percentNonCoding'] = 100 * (virus['genomeLength'] - totalCoding) / virus['genomeLength']\n",
    "    virus['percentOverlap'] = 100 * totalOverlap / virus['genomeLength']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Examining protein coding gene annotations that are shared between different viral groups to identify problematic search terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'C1',\n",
       " 'C2',\n",
       " 'C3',\n",
       " 'C4',\n",
       " 'C5',\n",
       " 'Coat protein',\n",
       " 'DNA maturation protein',\n",
       " 'DNA packaging protein',\n",
       " 'ORF1',\n",
       " 'ORF10',\n",
       " 'ORF11',\n",
       " 'ORF12',\n",
       " 'ORF13',\n",
       " 'ORF14',\n",
       " 'ORF15',\n",
       " 'ORF2',\n",
       " 'ORF3',\n",
       " 'ORF4',\n",
       " 'ORF5',\n",
       " 'ORF6',\n",
       " 'ORF7',\n",
       " 'ORF8',\n",
       " 'ORF9',\n",
       " 'Orf1',\n",
       " 'Orf2',\n",
       " 'Orf3',\n",
       " 'REP',\n",
       " 'Rep',\n",
       " 'Rep protein',\n",
       " 'RepA',\n",
       " 'VP1',\n",
       " 'VP1 protein',\n",
       " 'VP2',\n",
       " 'VP2 protein',\n",
       " 'VP3',\n",
       " 'VP4',\n",
       " 'VP4 precursor',\n",
       " 'assembly protein',\n",
       " 'capsid',\n",
       " 'capsid protein',\n",
       " 'capsid protein VP2',\n",
       " 'cell lysis protein',\n",
       " 'coat protein',\n",
       " 'gpA',\n",
       " 'gpC',\n",
       " 'hypothetical protein',\n",
       " 'major capsid protein',\n",
       " 'major coat protein',\n",
       " 'major structural protein',\n",
       " 'mannosyl-glycoprotein endo-beta-N-acetylglucosaminidase',\n",
       " 'minor capsid protein',\n",
       " 'minor structural protein',\n",
       " 'nonstructural protein',\n",
       " 'peptidase',\n",
       " 'phage replication protein',\n",
       " 'putative DNA binding protein',\n",
       " 'putative Rep protein',\n",
       " 'putative adsorption protein',\n",
       " 'putative capsid protein',\n",
       " 'putative coat protein',\n",
       " 'putative integrase',\n",
       " 'putative major capsid protein',\n",
       " 'putative major coat protein',\n",
       " 'putative minor capsid protein',\n",
       " 'putative replication protein',\n",
       " 'putative structural protein',\n",
       " 'putative transcription regulator',\n",
       " 'putative transposase',\n",
       " 'receptor-binding protein',\n",
       " 'regulatory protein',\n",
       " 'replicase',\n",
       " 'replication initiation protein',\n",
       " 'replication initiator protein',\n",
       " 'replication protein',\n",
       " 'single-stranded DNA binding protein',\n",
       " 'ssDNA binding protein',\n",
       " 'ssDNA-binding protein',\n",
       " 'structural protein',\n",
       " 'structural protein VP1',\n",
       " 'structural protein VP2',\n",
       " 'structural protein VP3',\n",
       " 'transactivation protein',\n",
       " 'transcriptional activator',\n",
       " 'transcriptional regulator'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#the shared protein coding genes between dsDNA, ssDNA viruses\n",
    "protMatds= [v['proteinNames'] for v in lengthData if 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "protMatss= [v['proteinNames'] for v in lengthData if 'ssDNA viruses' in v['englishLineage']]\n",
    "\n",
    "ds = [i for vec in protMatds for i in vec]\n",
    "ss = [i for vec in protMatss for i in vec]\n",
    "\n",
    "len(set(ds) & set(ss))\n",
    "set(ds) & set(ss)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Extracting gene order sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">gi| Leuconostoc_phage_Lmd1_uid171974\n",
      "AAAAAAAAAQQCIIAEAAEEAEAAAAAAAAAAAAAAAAAA\n",
      ">gi| Leuconostoc_phage_phiLNTR3_uid253389\n",
      "AAAAAAAAAAQQCIIAEAAEEAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Leuconostoc_phage_phiLNTR3_uid253389\n",
      "AAAAAAAAAAQQCIIAEAAEEAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Leuconostoc_phage_phiLN12_uid253392\n",
      "AAAAAAAAQQCIIAEAAEEAAEAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Leuconostoc_phage_phiLNTR2_uid253393\n",
      "AAAAAAAAAAQQCIIAEAAEEAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Leuconostoc_phage_phiLN25_uid253397\n",
      "AAAAAAAAAAQAQCIIAEAAEEAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL111M01_uid219111\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL113M01_uid219107\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P105_uid177533\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P9_1_uid177529\n",
      "QQCAIIAAAAEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL112N00_uid219110\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P1_1_uid177537\n",
      "QQCAIIAAAAEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P105_uid177533\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL111M01_uid219111\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL111M01_uid219111\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL112N00_uid219110\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL111M01_uid219111\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P105_uid177533\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P100D_uid177534\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL111M01_uid219111\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL112N00_uid219110\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_ATCC29399B_C_uid177539\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_ATCC29399B_T_uid177538\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Erwinia_phage_ENT90_uid184166\n",
      "EEAEEEAAAEAAAAAAIQIAAIQAAAACAAAAAAAAAAAAAAAAAAAEAAAAEEAAAAEA\n",
      ">gi| Propionibacterium_phage_P101A_uid177531\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P101A_uid177531\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Salmonella_phage_FSL_SP_004_uid212714\n",
      "CQIIQIEAAAAEAEEEEEEEEAEAEAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_PsP3_uid14345\n",
      "CAQIIQAAAAAAAAAEEAAAEEAAAAEAAAAAAAAAAAAAAA\n",
      ">gi| Yersinia_phage_L_413C_uid14280\n",
      "CQIIQAAAAAAAAAEEAAAAEEAAAAEAAAAAAAAAAAAA\n",
      ">gi| Clostridium_phage_phiMMP04_uid179417\n",
      "QCAIIIAAEAAAEAAAAAEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Vibrio_phage_K139_uid14144\n",
      "AAAAAAAAAAAAAAAAAAACQIIQIAEAEEAAAAAEAAAEEAAA\n",
      ">gi| Vibrio_phage_kappa_uid28503\n",
      "AAAAAAAAAAAAAAAAAAAACQIIQIAEEEEAAAAAEAAAEEAAA\n",
      ">gi| Stenotrophomonas_phage_Smp131_uid240053\n",
      "AAAAACQIIQIEAAAAEEEEEAEEEEEEEAEEAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_P2_uid14035\n",
      "CQIIQAAAAAAAAAAAEEAAAAEEAAAAEAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_fiAA91_ss_uid226726\n",
      "IQIIQIEAAAAEEEEEEEEAEEEEEAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_Fels_2_uid32273\n",
      "AEEEAEEEAAAAAEEEAEEAAAAAAEAQIIQCAAAAAAAAAAAAAAA\n",
      ">gi| Aeromonas_phage_phiO18P_uid19769\n",
      "AAAAAAAAAAAAAAAAAACQQIIQIAEEEAAAAAAAEAAAEEAAA\n",
      ">gi| Salmonella_phage_RE_2010_uid181070\n",
      "AAAAAAAAAAAAAAIQIIQIEAAAEEAEEEEEEAAEAAEEEEEEAAA\n",
      ">gi| Mannheimia_phage_phiMHaA1_uid17103\n",
      "CQIIQIEAAAAEEAEAAEEEEEAEEEAEEAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Streptococcus_phage_2972_uid15254\n",
      "AQQQCIQIIAAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Mannheimia_phage_vB_MhM_1152AP_uid212715\n",
      "IQIIQIEAAAAEEEAAEEEEEAAAEEEAEEAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Idiomarinaceae_phage_1N2_2_uid266773\n",
      "QQAAACAIIIAACAAIEIEAEAAEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Streptococcus_phage_DT1_uid15124\n",
      "AAQQICAIQIEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Streptococcus_phage_Abc2_uid42791\n",
      "QQICAIQIEEEEEEEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Burkholderia_phage_ST79_uid206488\n",
      "AAAAAAAAAAAAAAAACQIIQIEAAAAAEAEEEEEAAEEEEEEAAAAAA\n",
      ">gi| Streptococcus_phage_ALQ13_2_uid42593\n",
      "QQCIQIIIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_phiCTX_uid14415\n",
      "ACQIIQIAAAAAAAEEAEEEAAAEEAAAAEAAAAAAAAAAAAAAAAA\n",
      ">gi| Psychrobacter_phage_Psymv2_uid240594\n",
      "AAAAAAAAAAAAAAAAAAAAAAQQCAIAIIAAAAAAEEEAEEEEEAAAA\n",
      ">gi| Burkholderia_phage_phiE202_uid19163\n",
      "QIIQCAAAAAAAAAAAAAAAAAAAAAEEEEAAEEEAAAAEAAAAAAAI\n",
      ">gi| Enterobacteria_phage_P88_uid271780\n",
      "AAAAAAAAAAAAAAAAAAAAAACQIIQIEAAAEEEEEEEEEEAEEEEEEAAAA\n",
      ">gi| Enterococcus_phage_phiFL2A_uid42795\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQCAIAAQAAAAAAEAAAEEEEEAAAAAAA\n",
      ">gi| Lactobacillus_phage_Lc_Nu_uid16114\n",
      "QQACAIQIIEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Bacillus_phage_Cherry_uid15784\n",
      "QQCIIAIAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Escherichia_phage_HK75_uid76733\n",
      "QQCIIAIAAEEEAEEEEAEAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_phiJL_1_uid15156\n",
      "AAAAAAAAAAQQIAQIAQAIAEAAEAAEAAAAAAAAAAAAAAAAAA\n",
      ">gi| Burkholderia_phage_phiE12_2_uid19161\n",
      "QIIQCAAAAAAAAAAAAAAAAAAAAAAAAEEEEAAEEEEAAAAAAAAAAI\n",
      ">gi| Enterobacteria_phage_Mu_uid14105\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAQIAAAAEAAEAEEAAAEEEEAAA\n",
      ">gi| Idiomarinaceae_phage_Phi1M2_2_uid267261\n",
      "QQAIAIIAAAAAAAEEEAAAAAAAAAEAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_SfV_uid14162\n",
      "QQCIIAAAAAEAAEEEEEEEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Haemophilus_phage_SuMu_uid181066\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAQCAAAAAAAAAAEEAAEAEEAEAEEAAAAA\n",
      ">gi| Streptococcus_phage_5093_uid38299\n",
      "AAAAAAAAAAAAAQQIIAIAAIIEAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Escherichia_phage_D108_uid42515\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAQIAAAAEEAAEAEEAAAEEEEAAAA\n",
      ">gi| Bacillus_phage_Gamma_uid15783\n",
      "QQCIIAIAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_gh_1_uid14265\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAEIIIEEAAAAEAAQAQAA\n",
      ">gi| Streptococcus_phage_Sfi19_uid14045\n",
      "AAQQICIQIEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Arthrobacter_phage_vB_ArS_ArV2_uid230590\n",
      "AQCIAIAAAAEAAAEEEEAEAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Yersinia_pestis_phage_phiA1122_uid14332\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAIIIIEEAAAAEAQAAAAAA\n",
      ">gi| Burkholderia_phage_phi52237_uid15422\n",
      "QIIQCAAAAAAAAAAAAAAAAAAAEAAEEEAAEEEEAAEAAAAAEAI\n",
      ">gi| Clostridium_phage_phiCD6356_uid64557\n",
      "QQCIIAIAIAEAAAEEAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Bacillus_phage_Fah_uid16382\n",
      "QQCIIQIIAEAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_ATCC_8014_B1_uid184150\n",
      "QQIIQIIQAIAEAAEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_T3_uid14336\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEIIIIEEAAAAEAQAAQAAA\n",
      ">gi| Lactobacillus_phage_KC5a_uid16663\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQQCIIIAAAAAEEAEAAAAEAAAAAAAA\n",
      ">gi| Salmonella_phage_Vi06_uid64609\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIIEEAAAAEAQAAAAA\n",
      ">gi| Yersinia_phage_Berlin_uid18481\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAEEIIEEAAAAEAQAAQAAA\n",
      ">gi| Yersinia_phage_vB_YenP_AP5_uid266772\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAEIIIIEEAAAAAEAQAQA\n",
      ">gi| Ralstonia_phage_phiRSA1_uid19481\n",
      "AAAACQIIQIEAAAAEEAEEEEEEEAEEAEAEAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Bacillus_phage_phiCM3_uid240022\n",
      "AAAAAQQCAIIIAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Cellulophaga_phage_phi17_1_uid212962\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCIAAAAAAAAAIIIIEEAEAAEAAEAAAA\n",
      ">gi| Lactobacillus_phage_Lv_1_uid33535\n",
      "QQCIIIIEEEEEIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Cronobacter_phage_Dev2_uid239722\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIAEEIIIIEAQAQA\n",
      ">gi| Cellulophaga_phage_phi12_1_uid212966\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAQCIAAAAAAAAAIIIIEAEAEAAEAAEAAAA\n",
      ">gi| Lactobacillus_phage_phiAT3_uid14475\n",
      "AQQACAIAIIEEEEEAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Cellulophaga_phage_phi18_1_uid212959\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCIAAAAAAAAAAAIIIIEEAEAAEAAAAAAA\n",
      ">gi| Citrobacter_phage_CR44b_uid239730\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIIEAEAAAAAEAAQAAAQAAA\n",
      ">gi| Halomonas_phage_phiHAP_1_uid28763\n",
      "QQACIAAAEEEEEAAEAEAEEAEEEAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Erwinia_phage_vB_EamP_L1_uid181229\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIIEAEAEAAAAQAAAQAAA\n",
      ">gi| Rhodobacter_phage_RcapMu_uid76743\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAACIAAAAIAAAAAAAAEAAEEEAE\n",
      ">gi| Kluyvera_phage_Kvp1_uid32673\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIIEEAAAAEEAQAAQAA\n",
      ">gi| Enterococcus_phage_phiFL3A_uid42787\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQQQCAIQIAAAAEAAAAEAAAAAA\n",
      ">gi| Yersinia_phage_phiYeO3_12_uid14591\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEIIIIEEAAAAAEAQAAQAAA\n",
      ">gi| Citrobacter_phage_CR8_uid239727\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIIEEAAAAEAAQAAAQAA\n",
      ">gi| Enterobacteria_phage_K1F_uid15880\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAEIIIEEAAAAAAQAAA\n",
      ">gi| Enterobacteria_phage_HK97_uid14592\n",
      "QQCAIAIAAEAEAEAAAAAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Shigella_phage_SfIV_uid227000\n",
      "QQCIIAAAAEAAEEEEEEEAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Burkholderia_phage_KS9_uid39771\n",
      "AAAAAAAAAAAAAAAAAAAAQCQIAIAAEEEEEAEEEEAAAAAAAAAAAA\n",
      ">gi| Cronobacter_phage_phiES15_uid173212\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAQQAAAIAAAAAEAAEAEEAEAAAAA\n",
      ">gi| Mycobacterium_phage_Leo_uid209361\n",
      "AQCAAAIAAAAAEAAEEEAEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_Lrm1_uid30879\n",
      "QQACIIQIIIEAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQ\n",
      ">gi| Staphylococcus_phage_StB27_uid184157\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCIQIIIAIEEAEEAAEAAEAAAA\n",
      ">gi| Salmonella_phage_ST64B_uid14228\n",
      "QAAAIIAAAAAAAAAEEAAEIEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_prophage_Lj965_uid14351\n",
      "AAAAAAAAAAAAAAAAAAAAAAQQCAIQAIAAAAEAAEAAAAAAAA\n",
      ">gi| Pseudomonas_phage_Phi_S1_uid197298\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIIEEAAAAEAQAQA\n",
      ">gi| Stenotrophomonas_phage_S1_uid32787\n",
      "AAQQAACIAAAAAAAEAEAEAEAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_phiPSA2_uid253322\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEIIIEEAAAAEAAQAQA\n",
      ">gi| Streptomyce_phage_TG1_uid177524\n",
      "QQCIIAAEAEEEAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterococcus_phage_IME_EF4_uid240030\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEAEIIIIEIICAQQ\n",
      ">gi| Acinetobacter_phage_Petty_uid240050\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIQIAEEAAAEAAQQAA\n",
      ">gi| Streptococcus_phage_Sfi21_uid14133\n",
      "QQICAIQIEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_HK022_uid14048\n",
      "QQIIIAIAAAAEAAEAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Bacillus_phage_WBeta_uid16329\n",
      "QQCIIQIAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_StB20_uid184156\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQACIIQIAAEAAAEEEEAAAEAAAEAAAAAA\n",
      ">gi| Enterococcus_phage_EfaCPT1_uid266785\n",
      "QACIIAIIAAEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQ\n",
      ">gi| Pseudomonas_phage_phiIBB_PF7A_uid64561\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAEIIIIEEAAAAEAQAQAAAAAAAA\n",
      ">gi| Clostridium_phage_phiCD38_2_uid67249\n",
      "QQIIAAIAAAAAAAAEEEAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Klebsiella_phage_KP32_uid42779\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAEIIIEEAAAAEAAQAAQA\n",
      ">gi| Listeria_phage_LP_030_3_uid253385\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAQQCIAAAIIIEAAAEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Bacillus_phage_PBC1_uid167662\n",
      "QQAACIAIAAIIAIAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Vibrio_phage_X29_uid253324\n",
      "AAQAAAACAAIAAAAAEEAAAEAAEAAEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterococcus_phage_IME_EF3_uid239731\n",
      "AAAQQACIIIAIIIIEEEEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_Sha1_uid181084\n",
      "AAAAAAAAAAAAQQICAIQIIAEAAEAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Temperate_phage_phiNIH1_1_uid14145\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAQIIAAAIAIIIEAAEAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_DW2_uid253383\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCIAAIAIIAEEEEEEAAEAAAEAA\n",
      ">gi| Bacillus_phage_BtCS33_uid169233\n",
      "QQCAIIIAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Salmonella_phage_vB_SenS_Ent2_uid239739\n",
      "QQCAIIAAQIIAAIAIAAEAAAAAAAEAAAEEAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_vB_SepiS_phiIPLA7_uid171982\n",
      "QQCIIIAIIAAEAAEAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_phig1e_uid14315\n",
      "AAAAAAAAAAAAAAAIIAAAAIAIIIIIIIAIIQQAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Escherichia_phage_ep3_uid266771\n",
      "AAAAAAAAAAAAAAAAAAAAAAQQAIIAAAAAAEEEEEAAEEAEEEAAAAA\n",
      ">gi| Salmonella_phage_vB_SenS_Ent1_uid181993\n",
      "QQCAIIAAQIIAAACCAAEAAAAAAAEAAAEEAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Cyanophage_PP_uid227004\n",
      "AAAAAAAAAAAAAAAAAAAAAAAEEEAAEAEAIACAQAAAA\n",
      ">gi| Staphylococcus_phage_phiSauS_IPLA88_uid33001\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCAIAQIAQIAEAAEAAAEAAAAEAA\n",
      ">gi| Enterobacteria_phage_phiP27_uid14599\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQACIIAAAAAAEAEEAAAEEEAA\n",
      ">gi| Enterococcus_phage_IME_EFm1_uid253326\n",
      "AQQACIIIIIIEAEEEAEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_LF1_uid181083\n",
      "QQACIIQIIIEAAEAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_SA13_uid213078\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCIAIAIIAAEEAEEAAEAAAEEAAAA\n",
      ">gi| Salmonella_phage_SETP13_uid226727\n",
      "QQCAIEAAAAIIEAACAEAEAAAAAAEAAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Salmonella_phage_SETP7_uid226725\n",
      "QQCAIEAAAQIIEAACAAAEAAAAAEAAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Salmonella_phage_vB_SenS_Ent3_uid249091\n",
      "QQCAIIAAQIIAAIAIAAEAAAAAAAEAAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Burkholderia_phage_KL1_uid171970\n",
      "AAAAAAQCIAIAAAIEAEEAEAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_vB_Pae_Kakheti25_uid167052\n",
      "AAAAAQQCIQIAAAAAEEEEEAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_SA12_uid212955\n",
      "AAAAAAAAAAAAAAAAAAAAAAQQCIAAIAIAAAEAAEEAAAAAAAEAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_MPK6_uid227001\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIEEAAAAEEEQQAAAAA\n",
      ">gi| Pseudomonas_phage_vB_Pae_TbilisiM32_uid167051\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACQIEEAAAIEEEAAAAAAAA\n",
      ">gi| Edwardsiella_phage_eiAU_uid240025\n",
      "EEEEEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAIAAAAAAEAAA\n",
      ">gi| Acinetobacter_phage_IME_AB3_uid240027\n",
      "AAAAAAAAAAAAAAAAAAAAAAEAAIAEEEEEEIAAIQICQQAAAAAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_vB_PaeS_SCH_Ab26_uid253340\n",
      "AAAAQQCIQIAAAAAEEEEEAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_phikF77_uid36373\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIEEAAAAEEEAQAAAAAA\n",
      ">gi| Phormidium_phage_Pf_WMP3_uid19801\n",
      "AAAAAAAAAAAAAAAAAAAAAAAEAAAAAEEIQCAQAAAAA\n",
      ">gi| Staphylococcus_phage_phiRS7_uid229900\n",
      "QQACIIAQIIAEAEAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_A2_uid14602\n",
      "QQCAIQIIIEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQ\n",
      ">gi| Enterobacteria_phage_JL1_uid179426\n",
      "QAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAEEEEEEAAEAAAAAAAAAIA\n",
      ">gi| Pseudomonas_phage_LUZ19_uid28741\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIEEAAAAEEEQQAAAAAAA\n",
      ">gi| Staphylococcus_phage_vB_SepiS_phiIPLA5_uid171981\n",
      "QQCIAAIAQIAAEEEEEACAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_phiBU01_uid271781\n",
      "AAAAAAAAAAEEAEAAAIIACQAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_SP6_uid14291\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIEEAAAEAQQAAAAAAAEAAA\n",
      ">gi| Staphylococcus_phage_TEM123_uid167573\n",
      "AAAAAAAAAAAAEAAAAAEEAAAAEAAIIQICQQAAAAAAAAA\n",
      ">gi| Phage_Paz_uid230593\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIQIAEEAAAEEEEAQQAAAAAA\n",
      ">gi| Phage_Prado_uid230587\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIQIAEEAAAEEEEAQQAAAAA\n",
      ">gi| Shigella_phage_EP23_uid80919\n",
      "AQAIAAAAAAAAAEAEAEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Synechococcus_phage_S_CBP4_uid266792\n",
      "AAAAAAAAAAAAAAAAAAAAIIAAAIAAEEAAAEAAAAAAAAAQAA\n",
      ">gi| Lelliottia_phage_phD2B_uid266769\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIEEAAAEAQQAQAAAAAA\n",
      ">gi| Bacillus_phage_phIS3501_uid181213\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCAIQIAAEAEEAAAAAAAAA\n",
      ">gi| Vibrio_phage_vB_VchM_138_uid181217\n",
      "AAAAAAAAAAAAAAAAAAAAQQCAAAAIAAAAIAIAAAAAAAAAAAAAAAAAAEAAAEAEAEEAAAA\n",
      ">gi| Enterobacteria_phage_SSL_2009a_uid34919\n",
      "QAIAAAAAAAAAEAEAAEEEEEEAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Salmonella_phage_E1_uid29079\n",
      "AQACIAAAEEAAAAAIIAAAAAAAAAAAEAAAEAAAEAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_StauST398_1_uid206473\n",
      "AAAAAAAAEEAAAAAAAEEEAEAEIIAIAICQQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_SMSAP5_uid181240\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAEEEAAEEAAAIAIQQ\n",
      ">gi| Streptococcus_phage_TP_J34_uid188154\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAQQCIQAIAAAAAEAAEEEAAAAAAAAAAAAAAA\n",
      ">gi| Mycobacterium_phage_BTCU_1_uid209077\n",
      "EEAAAEAAQCIQIIAAAAEEEAAAAEAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_RTP_uid16178\n",
      "AAAAAAAAAAAAAAAAAAQAQACIIAAAAAAAEAAAEEEAEEEAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_cdtI_uid19737\n",
      "QQACQAAEEEEEEEEEEEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Iodobacteriophage_phiPLPE_uid30965\n",
      "AAAAAAAAAAAAAAAAQQCAAAAAIAIAIAAAAAAEAEAAAAAAAAAAAAAAAAEAAAEAEAEAAAAAAAAAAAAAEAAAAAAA\n",
      ">gi| Enterobacteria_phage_lambda_uid14204\n",
      "QQIIIIIIQIEEEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_T1_uid14496\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAQQCAIAAAAAAAAEAAEEEEEEAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Cronobacter_phage_ESP2949_1_uid181234\n",
      "EEEEAEAAAAAAAAAIICQQAAAAAAAAAAAAAAAAAAEAAAAEE\n",
      ">gi| Burkholderia_phage_phiE125_uid14330\n",
      "QQACIIAAAAAEEAEEAEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Puniceispirillum_phage_HMO_2011_uid213071\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAEEEAEEEAAEEEEEAAAAAIAAACQAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Burkholderia_phage_BcepNazgul_uid14305\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEEEEEEAAAAAIAIACIQQAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_BP_4795_uid14287\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQIACAAIAAEEAEEEEEEAAEAAAAAAAAAA\n",
      ">gi| Burkholderia_phage_AH2_uid171971\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEEAAEEEAEAAAIAICIQQAAAAAAAAAA\n",
      ">gi| Salmonella_phage_FSL_SP_088_uid212711\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEAEAAEEEAAAAAIAICIQQAAAAAAAAAA\n",
      ">gi| Salmonella_phage_FSL_SP_030_uid212718\n",
      "AAAAAAAAAAQQICIAIAAAAAEEEAAEAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacter_phage_Enc34_uid181238\n",
      "AAAAAAAAAQQICIAIAAAAAEEEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Stx2_converting_phage_1717_uid32213\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAICCAIAEEAEEEAEEEAEAAAAA\n",
      ">gi| Acinetobacter_phage_133_uid64541\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAAEEEAEEEECCIEQQEEIIIIIIAIIAAAAAAAAEEEEEEEEAAAAAAAAAIAAAAAAAAAAAAAAAAAAEAEAAAAAAAAAAAAAAAAEEEEAAAAAAAAA\n",
      ">gi| Salmonella_phage_S16_uid191122\n",
      "AAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAAEEEEEEECCCEQQEAECIIAIIIIIAAAAAIIAAAAAAAEEEEEEEEAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAEAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Aeromonas_phage_Aes012_uid195532\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIAAEEAEEEEEEECCCEQQEEIIIIIIIAAAAAAAIAAAAEEEEEEEEAAAAAAIAAAAAAAAAEAAAAAAAAAAAAAEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_3S_uid266636\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAAIEEAAAEEEAAAEEEECCCEQQEECIIIIIAAAIAAAAAIAIAAAAEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Aeromonas_phage_phiAS4_uid59727\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEEEEEAAAAAIIAAAAAAIAAIIIIIIIAEEEQQEIIIIAEEEEEEEEEEEAEEAIAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_RB49_uid14301\n",
      "AAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAAIEEAAEEEAAAAEEEEACCEQQEECIIIIIAIAAAAIAIAAAAEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Acinetobacter_phage_Acj61_uid60117\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAAIEEAAEEEAEEEECCIEQQEAEIIIIIIIIIAAAAEEEEEEEEAAAAAAAAIAAAAAAAEAAAAAAAAAAAAAEEEEAAAAAAAA\n",
      ">gi| Enterobacteria_phage_Phi1_uid20789\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAAAIEEAAEEEAAAEEEECCACEQQEEIIIIIIAAIAAAAIAIAAAAEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Shigella_phage_Shfl2_uid66347\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAIIIAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAEAEAIEEAAEEEEEEECCIEQQEEIIIIIIIAAAIIAAAAAAEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_JSE_uid38263\n",
      "AAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAAAIEEAAEEEAAAEEEEACCAEQQEECIIIIIAAIAAAAIAIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Escherichia_phage_wV7_uid181232\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAIAEAEAIEEAAEEEEEEECCCEQQEEIIIIIIIAAAIIAAAAAEEEEEEEEAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_ime09_uid181233\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAEAEAIEEAAEEEEEEECCCEQQEEIIIIIIIAAAIIAAAAAEEEEEEEEAAAAAAAAAAAAAAIEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Serratia_phage_PS2_uid248461\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAIEEAAEEEEEEECCCEQQEECIIIIIIAAAAAAAAAAIIAAAAEEEEEEEEAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEEAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_vB_EcoM_ACG_C40_uid179415\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAAAEEEEEEECCCEQQEEIIIIIIIAAAAIIIAAAAAAAEEEEEEEEAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_RB69_uid15141\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAEAEAIEEAAEEEEEEEACCEQQEECIIIIIAAIAAAIIAAAAAEEEEEEEEAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_Bp7_uid181212\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAIAAAAAAAAAAAAAAAEEEEEEEEAAAAAAIIAAIIAIIIIIIIAEAEQQECCCEEEEEEEAEEIAEAEAI\n",
      ">gi| Escherichia_phage_e11_2_uid248462\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAEAEAAIEEAAEEEEEEEACCEQQEEIIIIIIIAAAIIAAAAAAEEAEEEEEAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Yersinia_phage_phiR1_RT_uid184143\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIAAAAAAAAAAAAAAAAAAAAAAAEAIAEEAAEEEEEAEAIIAEQQAQEEIIIIIIAIAAAIAAAIAAAAEEEEEEEEAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_T4_uid14044\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIAAAAAAAAAAAAAIAEAEAIEEAAAAAAEEEEEEEACCEQQQQQEECIIIIIIAIAAAIIAAAAAEEAEEEEEEAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_vB_EcoM_VR7_uid61099\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIAAAAAAAAAAAAAAAAAAAAAAAIAAIAAEAEAAIEEAAEEEEEEEACCEQQEAECIIIIIAIIAAAAAIIAAAAEEEEEEEEAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Acinetobacter_phage_Acj9_uid60121\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAIEEAAEEEEEEECCIAEQAQEEIIIIIIIIIAAAAAAAEEEAEEEEAAAAAAAAIAAAAAAAAAAAEAAAAAAAAAAAAAAAAEEEEAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_JS98_uid27983\n",
      "AAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAAEEEEEEEACCEQQEECIIIIIAIIAAIIAAAAEEEEEEEEAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Shigella_phage_SP18_uid56019\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAIIAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAEEEEAEEECCACEQQEEIIIIIIAIAAAIIAAAAAAEEEEEEEEAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAEEAEAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_JS10_uid38265\n",
      "AAAAAAAAAAAAAAAAAAAAAAAIIAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAAAEEEEEEEACCEQQEECIIIIIAIIAAAEIIIAAAAEEEEEEEEAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_IME08_uid50177\n",
      "AAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAAEEEEEEECCCEQQEEIIIIIIAAIIEAAIIIAAAAEEEEEEEEAAAAAAAAAAAAAAAIAAAAAAAAAAAAAEAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAA\n",
      ">gi| Klebsiella_phage_KP27_uid185314\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAIAAIEEAEEEEEEECIIEQQEEIIIIIIAIAAAAIAAIIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAA\n",
      ">gi| Klebsiella_phage_KP15_uid47333\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAAIEEAEEEEEEEACCEQQEECIIIIIAIAAAAIAAIIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAA\n",
      ">gi| Enterobacteria_phage_RB16_uid51699\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAIAAAIEEAAEEEEEEEACCAEQQEECIIIIIAIAAAAIAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAA\n",
      ">gi| T4likevirus_uid266638\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEAAAEEEEEEECCCEQQEAEIIIIIIIAAAIIAAAAAAAAEEEEEEAAAAAAAAAAAAAAAIEEAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEAAAEEEEEEECCCEQQEEIIIIIIIAAAIIAAAAAAAEEEEEEAAAAAAAAAAAAAAAIEEAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAA\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] \n",
    "          and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "genomeLen= [v['genomeLength'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage']  \n",
    "            and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "virusName = [v['folder'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage']  \n",
    "             and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "virusLineage = [v['englishLineage'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] \n",
    "                and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "hostLineage = [v['hostEnglishLineage'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage']  \n",
    "               and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "\n",
    "#ordering the name list and the protMat to be sorted based on the genome lengths. \n",
    "protMat = [protMat for (genomeLen, protMat) in sorted (zip(genomeLen, protMat),\n",
    "                                                       key=lambda pair:pair[0])]\n",
    "virusName = [virusName for (genomeLen, virusName) in sorted (zip(genomeLen, virusName),\n",
    "                                                             key=lambda pair:pair[0])]\n",
    "virusLineage = [virusLineage for (genomeLen, virusLineage) in sorted (zip(genomeLen, virusLineage), \n",
    "                                                                      key=lambda pair:pair[0])]\n",
    "hostLineage = [hostLineage for (genomeLen, hostLineage) in sorted (zip(genomeLen, hostLineage), \n",
    "                                                                   key=lambda pair:pair[0])]\n",
    "\n",
    "#here are some search terms that can be categorized into the functional groups\n",
    "capsid = ['coat', 'capsid', 'head']\n",
    "matrix = ['matrix', 'tegument']\n",
    "#env = ['envelope'] \n",
    "portal = ['portal','neck', 'head-tail connector', 'head-tail joining'] \n",
    "tail= [ 'tail','baseplate', 'tape measure']\n",
    "packing = ['terminase', 'TerL','TerS', 'scaffold protein', 'prohead protease', 'DNA packaging']\n",
    "lysis = ['holin', 'endolysin', 'lysozyme','lysis', 'lysin', 'N-acetylmuramoyl-L-alanine amidase', \n",
    "         'cell wall hydrolase']\n",
    "nuclease = ['nuclease']\n",
    "integrase =['integrase']\n",
    "polymerase =['polymerase']\n",
    "helicase =['helicase']\n",
    "replication =['Rep protein','rep protein','replicase', 'replication']\n",
    "regulationModification = ['transcriptional enhancer', 'transcriptional regulator', 'expression regulator',\n",
    "                          'promoter', 'antirepressor','anti-repressor', 'thioredoxin','DNA methylase',\n",
    "                          'methyltransferase','ADP-ribosylase','ADP-ribosyltransferase']\n",
    "metabolismSynth = ['thymidine kinase', 'deaminase', 'thymidylate synthase','dihydrofolate reductase', \n",
    "                   'ribonucleotide reductase', 'dNMP kinase','dCMP hydroxymethylase',\n",
    "                   'deoxynucleoside monophosphate kinase','amidase','phosphate starvation-inducible protein' ]\n",
    "entryInjection = ['movement protein','injection']\n",
    "trna = ['tRNA']\n",
    "\n",
    "#determing and storing the color-code information for each viral gene (looking at only structural genes here)\n",
    "colormat=[]\n",
    "virusNameFiltered =[]\n",
    "for vec in protMat:\n",
    "    colorvec=[]\n",
    "    for i in vec:          \n",
    "        if any(substring in i for substring in capsid):\n",
    "            colorvec.append('I')\n",
    "        elif any(substring in i for substring in portal):\n",
    "            colorvec.append('C')         \n",
    "        elif any(substring in i for substring in tail):\n",
    "            colorvec.append('E')          \n",
    "        elif any(substring in i for substring in packing):\n",
    "            colorvec.append('Q')         \n",
    "        else:       \n",
    "            colorvec.append('A')\n",
    "    #colormat.append(colorvec)\n",
    "    #Here we have some additional filters on gene order sequence to make sure it's not all unlabled\n",
    "    #genes or not too short\n",
    "    if colorvec.count('A')/len(colorvec) < 0.85 and len(colorvec)>39:                \n",
    "        colormat.append(colorvec)\n",
    "        #getting the index of the vectors in protMat that passed through the filter. \n",
    "        #Then I use the index to get the virus names from the virusName matrix \n",
    "        index = protMat.index(vec)\n",
    "        virusNameFiltered.append(virusName[index])\n",
    "        \n",
    "        print('>gi|',virusName[index])\n",
    "        print(str(colorvec).replace(\",\",\"\").replace(\"'\",\"\").replace(\" \", \"\").strip('[]'))\n",
    "        \n",
    "        \n",
    "        #note T4likevirus_uid266638 virus genome has been copied twice, which is why we have \n",
    "        #taken this virus out in our plots.\n",
    "        #originally we thought it's a genome with its gene order pattern repeated twice, until we looked \n",
    "        #further at the paper wherein the phage was first sequenced. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Plotting relative abundances of gene functional categories for different viral groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYEAAAD8CAYAAACRkhiPAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAH3BJREFUeJzt3X+c3VV95/FX7oRMMplMJomACQkBTHuQ0lXcaqW74Kp9\n+APXutv1IQirQBGCikKAFaFLFF1qWRoEMUktDSI/Wn5UsLSULpRFxR8glqpR2FOEOpm5M5loSGYy\nmUlicmf/OJ9vcufm/py5v8/7+XjkMTP3nu8553vO+Z7P+X7v93sza3JyEhERiVOq0RUQEZHGURAQ\nEYmYgoCISMQUBEREIqYgICISMQUBEZGIzS6VwDl3LPC3wI+AQWAHcBzQA6wBOoF1wHbgee/9Bufc\nldlpvPfba1F5ERGZmXLOBE4HhoBJ4PvAad77S4BNwEXAauAWe+0M51w3cLr9fbulERGRJlTyTAD4\nAfC4937YOfcE8LK9ngaWAUcA/fbaDqAXGLa/B4Cl1auuiIhUUzlB4BTCGQDABGHiB1hOCAQpYIX9\nvphwyWhJVprBQhlPTOybnD27o+JKd3TM4sCByp90nu521Sizr6+PMy/beNj7nfMXcef1Z7Fy5cqm\nqq/KrN22L7/8i8PGQrFxUI0yW6WNYilzJtt2dMwilUrNmlaheZQTBF4EbnTODQMPAV3OufWEFf/F\nQBdwk3PuXOBB733GOfdkTpq8xsb2TqvSvb1d7Nw5XrftqlHm6OhEwTSjoxN5825kfVVm7bYtNBYK\njYNqlNkqbRRLmTPZtre3i1QVb+kpGQS8988BZxZJsgs4J2ebW2dYLxERqQPdIioiEjEFARGRiCkI\niIhETEFARCRiCgIiIhFTEBARiZiCgIhIxBQEREQipiAgIhIxBQERkYgpCIiIRExBQEQkYgoCIiIR\nUxAQEYmYgoCISMQUBEREIqYgICISMQUBEZGIKQiIiERMQUBEJGIKAiIiEVMQEBGJmIKAiEjEFARE\nRCKmICAiEjEFARGRiCkIiIhETEFARCRiCgIiIhFTEBARiZiCgIhIxBQEREQipiAgIhIxBQERkYgp\nCIiIRExBQEQkYgoCIiIRUxAQEYmYgoCISMRml5PIOXc38HfACuA4oAdYA3QC64DtwPPe+w3OuSuz\n03jvt1e/2iIiUg0lzwScc2uAXfbnad77S4BNwEXAauAWe+0M51w3cLr9fbulERGRJlX0TMA5915g\nJ/A0IWBss7fSwDLgCKDfXtsB9ALD9vcAsLRY/t3dncye3VFxpTs6UvT2dtVtu2qUOTIyr2Canp55\nefNuZH1VZu227enJPxYKjYNqlNkqbRRLmTPZtqOjulfxS10OOgd4BTjR/k7OCJYTAkGKcIkoDSwG\nBoElWWkGi2U+Nra38hoDvb1d7Nw5XrftqlHm6OhEwTSjoxN5825kfVVm7bYtNBYKjYNqlNkqbRRL\nmTPZtre3i1Sq8sVzIUWDgPf+LADn3IeBPcDRzrn1hBX/xUAXcJNz7lzgQe99xjn3ZE4aEZGKZDIZ\n0ukBjjlmOamU7l+ppbI+GPbe31ngrV2Es4XstLfOtFIiErd0eoAL197BVz77YVKplIJBDalVRaQp\ndc5fxPDwVi5cewfp9ECjq9O2FAREpKl1zl/U6Cq0NQUBEZGIKQiIiERMQUBEJGIKAiIiEVMQEBGJ\nmIKAiEjEFARERCKmICAiEjEFARGRiCkIiIhETEFARCRiCgIiIhFTEBARiZiCgIhIxBQEREQipiAg\nIhIxBQERkYgpCIiIRExBQEQkYgoCIiIRUxAQEYmYgoCISMQUBEREIqYgICISMQUBEZGIKQiIiERM\nQUBEJGIKAiIiEVMQEBGJmIKAiEjEFARERCKmICAiEjEFARGRiCkIiIhETEFARCRiCgIiIhGbXSqB\nc24V8Hngl8A/A0cCxwE9wBqgE1gHbAee995vcM5dmZ3Ge7+9FpUXEZGZKedMoAf4NHA5cDZwmvf+\nEmATcBGwGrjFXjvDOdcNnG5/325pRESkCZUMAt7754BfA48ATwLb7K00sAw4Gui313YAvcCw/T0A\nLK1ifUVEpIrKuRx0CtDnvX+nc+5vgA57azkhEKSAFfb7YmAQWJKVZrBQ3t3dncye3VHo7YI6OlL0\n9nbVbbtqlDkyMq9gmp6eeXnzbmR9VWbttu3pyT8WCo2DapTZKm2Ue7x0d88FirdNtcqcjka1UTWV\nDAKW5i+cc/3AS8CAc249YcV/MdAF3OScOxd40Hufcc49mZMmr7GxvdOqdG9vFzt3jtdtu2qUOTo6\nUTDN6OhE3rwbWV+VWbttC42FQuOgGmW2ShvlHi9jY3uA4m1TrTKno1FtlEpVvngupGQQ8N4/C7y/\nSJJdwDk529w6w3qJiEgd6BZREZGIKQiIiERMQUBEJGIKAiIiEVMQEBGJmIKAiEjEFARERCKmICAi\nEjEFARGRiCkIiDSh/v4t9PdvaXQ1JAIKAiIiEVMQEBGJmIKAiEjEFARERCKmICAiEjEFAYlaJpOh\nv38LmUym0VURaQgFAYlaOj3AhWvvIJ0eaGg9JiczDA0NKhhJ3SkISPQ65y9qdBXYNz7C2o2PNTwY\nSXwUBESaRDMEI4mPgoCISMQUBEREIqYgICISMQUBEZGIKQiIiERMQUBEJGIKAiIiEVMQEBGJmIKA\niEjEFARERCKmICAiEjEFARFpevrK79pREBCRpjc8vLUpvvK7HSkIiEhL0Les1oaCgIhIxBQEREQi\npiAgIhIxBQERkYgpCIiIRExBQEQkYrOLvemcOxW4GBgFtgETwHFAD7AG6ATWAduB5733G5xzV2an\n8d5vr1XlRURkZkqdCfQCH/PefwL4j8Bp3vtLgE3ARcBq4BZ77QznXDdwuv19u6UREZEmVfRMwHv/\nKIBz7hrgbuB0eysNLAOOAPrttR2EoDFsfw8AS6tcX5GaSL6W4JhjlpNK6SqpxKPU5aBu4GbgHuDb\nwH+1t5YTAkEKWGG/LwYGgSVZaQaL5d/d3cns2R0VV7qjI0Vvb1fdtqtGmSMj8wqm6emZlzfvRtY3\nljKTftm9eyeX3/gw9938UVauXFnT+vb0lB4LSb2ScmLtl+7uuVN+FjpWqlHmdDSqjaqpaBAAbgFW\nAecDHwaedM6tJ6z4Lwa6gJucc+cCD3rvM8653DQFjY3tnVale3u72LlzvG7bVaPM0dGJgmlGRyfy\n5t3I+sZSZtIvY2N76Jy/qGBfVLO+5YyFJE1STsz9kv2zWP/MtMxMJkM6PVDR2WCj2iiVqnzxXEip\ny0EXlNh+F3BOzja3zrRSIiL1lk4PcOHaO7jtc+exYsWxja5O3ejip4iIifFL6hQEREQipiAg0iQm\nJzMMDQ3qP06RulIQEGkS+8ZHWLvxMf3HKVJXCgIiTSTGa9LSWAoCIiIRUxAQEYmYgoCISMQUBERE\nsiTfIxXLXVoKAiIiWYaHt3Lh2juiuUtLQUCaRmwrMGleMd2l1bJBQBNG+0m+uyWWFZhIM2jZIKAJ\noz3FtAITaQYtGwRAE4ZIPeisu721dBAQkdrTWXd7UxAQkZJ01t2+FARERCKmICB56TqwSBwUBCQv\nXQeWXFoYtCcFASlI14EbY2hokKGhwUZX4zCxPUkbCwUBESmbFgbtR0FARCRiCgIiIhFTEBARiZiC\ngEgEdGePFKIgIBIB3fIrhSgIiERCd/ZIPgoCIiIRUxAQEYmYgoBInWQyGfr6+vThbIUmJzNs2zbc\n6Gq0LQUBkTpJpwc487KNDA9vbXRVWsq+8RHW3fXtstP392+hv39LDWvUXhQEROpIH85OT2fXwkZX\noW0pCEhBk5MZhoYGdflCpI21RRDQ6V9t7BsfYe3Gx3RvuUgba4sgIPlV4ylRXb4QaW8NDwJ6nL12\nWvUpUY0JkfppeBBo1YmqVbTiSr6d//MS3e5YP1pMlKfhQQBac6JqJa14MLTrmKj0dkeZPi0wy9MU\nQUBqq51X1q1ItzvWT7suJqppdjmJnHOrgPu9929wzl0JHAf0AGuATmAdsB143nu/ITeN9357Deou\nFdDBICL5lDwTcM4dDVwAjDnnOoHTvfeXAJuAi4DVwC322hnOue6sNLdbGhFpAq14aVBqq+SZgPd+\nGLjaOfcPwCIg+VQrDSwDjgD67bUdQG9WmgFgaaG8u7s76emZB0BPzzx6e7vKqnRHR2rKdolS23d0\npMouo1rbJtuNjMwrmKbQvs+0vknbdHfPLVpOqfqW2zfVaqPp1LcVyiw2BrLlq0u1xsLu3Tu5/MaH\nue/mj7Jy5cqS2zVLvySKlV3oWCtV30bs50y27eio7lX8si4HZdkGLLHflxMCQQpYYb8vBgZz0gwW\nymxsbC+joxMAjI5OsHPneFmV6O3tmrJdotT2vb1dZZdRrW2T7bLrmavQvs+0vkmZY2N7ipZTqr75\ntstkMqTTAxxzzHJSqdSM6ptb5nTq2wplFhsD2fLVpZpjoXP+ommNhUb2S6JY2YWOtVL1bcR+zmTb\n3t4uUqmOaZWZT0UhxXufAZ50zq0HLgTWEy4LfcI5txF4sEAaKaBVv5qhXe+80OUSiU3ZZwLe+zPs\n5605b+0CzslJm5tGCki+muG2zy1jxYpjG12dirTjh83Dw1utP85ruf4QmQ7dItoE2nEybWXqD4F4\nzgoVBERE8ojl+ZpKPxgWkTpKvh23t/fEBtckTjGcFepMQEQkYk0TBGK5/iYi0kyaJgjEcv1NRKSZ\nNE0QgDiuv4mINJOmCgIiIlJfCgIiIhFTEGgCrfrVEbWmmwVam/qvNSgINIHkqyP0ofhU9b5ZQP/1\nY3W16/dLtRsFgSahD8Xzq2e76L9+LK3Ss1aN6+anICCSRf/1Y3E6a20/+tqICiWP8Sf0TZNST8l1\ndghjL/m/HLLfz/1/HqptOqv77DrXU3LmUsv2aHVqFYleK30WkE4PcN6nbmX1Z+/MuxrXdfipdOZS\nWlsFgf7+LYet1EVKabXPAjq7FhZdjes6/FRqj+LaKghI62vUqlyfBUisFASkqbTaqlyk1SkISNPR\nqry5xfBwYyt9TjRTCgIiUpEYPmyN6Yy0pYNADCuSatJj/FItMXzYGssZadMFgUomqmRFMjy8laGh\nQYaGButQw9al/7OhtdR6TOtuOoEmDAKVTlQxrEiqSe11SCaT0cJBotd0QQA0UdVCTB90lSudHuCq\ndQ8UTaNLaNLumjIISPXF9EFXOZKzgFLXfdvpCdx6LgRqcZbVjp8BNsMiQ0GgSdRjgE/ng65y65XJ\nZOjr62uZA7S/v7/kWUCiXc5M67kQqKR9y9WOdyU1wyKjKYJAu1+qKGdV1KwDvNx6pdMDnHnZxqar\nfzGFgmI7rjgT1brjpZw2qsXdNdMJyM3en53zFzX0jKApgkC7X6oo59ozNO+Ks9x6NWv9K9WsAbmZ\ntFIbzaSu9QogjbxzrymCALTmPbnhEsgv6Ov7RclB0or710iNPjtsl4BWytDQ4LRvE53uV0o34rbU\n6fZnuQEkWcnv379/2iv6Ro25pgkC9VStU6/h4a1Fv9a3ndT7dLXdzw6ldZQzOSfX9p977ocNv8Zf\nqaYKAvVa/SUdlqxKZjKxlfpa32Yz3dPbRpyu6uxp5vJ9YN+oa+St8EDnTOagZB5opfkAmiwI1HP1\n1zl/UZRP0M7k+mirDe52MZM7r5IP7IeHtx58LftJe5mq0WegjfiAuKmCANR/9RfjxFZsn5vhvmWZ\nanh464zuvMrX37GN+0rOfhp5BtqIhWnTBYFcrT4ptdpXE/T393Ph2juacpXY6mNhJmKbtKut3ncz\nzWSs1ruvmzIIZEft6T5MUW4n1Pr6aC0emqm1Zp1wio0FfRnazFSz/cJxm65KXtOV7/OHZFzXYzHR\nSpeamzII5Ebt6UxK5QaPeqwQGnF6WcszkHqe3eQG6WYNUO2ukg9M0+kBPn7dXTWu0fTV6yndVhmr\nTRkEoDoNOKdrYVmr/FbprHJlMhmefvrpmp2BlPvwWzUkQbq/f0vTXFbr6+try7OOYivkSj8wLbbw\naYbLerU85hv9jEulmjYITFf2wy/1ug7YbI+lJyuxWp6B1PPsJrmTq5zAk33bb7UnmuxLJjN5yGq6\naj3GSl3CqFaft9Klkulo9B1GlapJEHDOLXPO/bVz7svOuY9NN59Clx1KXY7Ifr+ciD85GT57KOfJ\n33xqGWymO5klB2wlq5LkWu7k5NSykjzyBbt6rOqS8ktNQtnXgat5yt8sQf6qdQ/U/AP7WqyQi12f\nb1eVBMzs4ys5Vuu5yKjVmcBq4Bbv/SXAe5xzHZVmMDmZYfPmH+dd/ZX6sLXcVWNi3/gIV3/x6zN6\n8reagzp7xTnTyaySVUk6PcDlNz7M3t07pwSOJI994yNcu+H/sHnzjw++l6zqajk5TXdlNZM+yZ74\nC91XPzQ0ePBrQ5IFRC2DYr3OvqrxFQjlllPu165Us8xKLytmj4VKPkAvd/GQfXw14gxido3yfTXQ\nb7/vABYCr+QmSq7z7h0fOSyDveMjfOG2nzJnbg97d++Y0nG7dr1yME0xyfvJtiMj8xgdnQhl7t7B\ntm3Dh+VRaICMjMxj4cIjC5ZRaNtdu14pWc985S9duqysehXKI7fMcuqQlDG2I80XbhtgztyevHl8\n4ba/P/heEiy2bRs+rJ8AVqw49rBycg+iKf1SpJ65/Zm9bb79yFefRKl+2Ts+wlXr+rjiQ29h7/jI\nwQl4aGiQXbvmHkw3PLyVNddvYs7cBdxwxQcAuGrd/dxwxQfy9mElY2FKfXbvYO/4yJQxm2/fkrGd\nvFdsrGdvk13O5s0/Zt1d3+La1e/g81957GAblNp2Sh1KHZtWThhLCw6213SP0XzjDA4f90mZCxav\nyD8vFJiLrlrXxw1XHH5cQtY8llPfUtuNjMybUubBn7t3TMk72b/k95GReaxadULe/Z2OWZOTk1XL\nLOGcuwZ4wnv/jHPuEeC93vvmuGAuIiIH1SoIHA3cBIwCP/Teb6p6ISIiMmM1CQIiItIa2u4WURER\nKZ+CgIhIxKIMAs65Wc65wz+ur01ZK+pRTk6Zc5xzR01ju7rXtVzF9mmm/Tnd9qog/yOdc3PKSFfV\n9nfOLa9mfs2g1n0Vo7oHAefcZ5xzv1vk/Uft541VKu9859xr7feH7OXTgA/Za5c6506vRlkFfMXK\nWeic+2pWvc51zt2f3RZZ9SsqN6+s179qB8hZwFunUddnK93AOfcW59xVOa89lPV73rpmvf9oifzf\n55z7MMX36WB/Zm13WL8WKess4K3Oudc7584uVp+c/K52zh2b7J9z7nXOuc/kpHkIuAHoLSPLr1ge\nL5ZbhzxlJb8fC/yx/Z4cU+9yzhUdF6XGYKnj19IUPHadcw855z7inHvC6vO2YnnlkXccJGU65x7P\nOc6+aG2Rmz45VpK/r3bOFeyj3PT2WjI2p83a4zPOucezXjvXOfe+AukrKrOcOaXuHwzbQfJaYILw\nLEEX4XmFhcAVwN3e+zNs4H4AWAfsAfYSDqYbgJ3ACcD7gf8JLACWAxcD7wTWAEcBS4CfAW8Afgkc\nCfwd4IHzgUeAdwNPA2PA5bb9O4FdVsaNwP8CfgA8Bfw+8B8s/Tdt+05gN/Ai4XmI/wwcAD4NbCBM\nrkfZvr5IuHPqNcB7gFcBxwFzgVHv/YnOuQ3APcC9QJ9t9/8sjy3Wfq8AfwZ8EOi29z9o+3sy0AH8\nwvJ/wdrge8CvgN8GftP64PvA24EfWb1PAz7rvX+3Hezvsnw+AhxhfZECTgIeA5YC3wZOtzaeBbwJ\nuBb4T1aXBcDdwD4rdxxYDHwMuA+4DFjjvf+oc+6PgX+yvj2N8MyJt/Zaanl839r3NUAayADzrQ0z\nhAn3l7bdK4Sx8VHgHhtbXwT2A78DbAdOsbboAxbZdnNtHHzX+nqVlXUS8BPCmLnI3n8jwRGWx2Jg\njuXRQRhLd1sdTwJ+bGVfB2y0/Pfa/j5t/fMAcCJhDGUIY3sS+A1gm9VtO/CPwB8R7sTrAV62Pp60\n/XoHYdz+PbAS+KK1993AZ4HHCeNvvpVzBHAq8AVr4znApcCThDE2mzBe7rN9eNTKWGB1ewp4vdXh\nBOA5648F1h5LrM1mAV8jjKdfcmgc/6Xtx+9bnX4E3AncYvv7Nsv7SODXhPH9O8BmwjNJr7Y26AFG\nrN9Os3RJGZsIx/VmwnHRDxxPGK9PE8b1a6zfjrBtjrK+vMXae77V4wBhfKXttecJY2g78LvAM/b6\nXsK88V3CXHG0vT9qeXTYewusvU4mzBVbLP2brH8GCXPls8Bc7/3HnXPXWrsutj59DWF+2wb8hvf+\nPRTRqMtBt3vvzycM8NcSJoUDwJtz0p0JPOy9/yTwV5bmq4SBtsz+HU/ojI2ExvoUYaJ4htCJ+4EB\nwqT0fStjB6ERH7Rt9ljepwIftLr9CaF9JoGnvPdXEAbXPYQD/QjCKv+3gZ8CF9j7b7ftvmP5prz3\n7yUEkn5L81bLF0JHnmB59Npqo9PSpq3e11k5q7z3fwQ8QTiILvXeXwScZ9tkgIcJB8BmwgQ4Ynm8\nRAhkb7dy9wB/TjgALyWssEa8989k1Q37fdLy/5bVfxZhcE4APyccIEcB11vb/hz4gNX1W8C/eu+v\nB/6H/X4p8HVs9e69/1fb9yMJE9crVsajwF3W5vsIA/unhAlwK4fOXBYTAtErhAnja7b/r+bQ2BgD\ncM6lrOzHgWMJk+h1VsYj1rfvI0y63yFMNA8QJj9PGEvfI0xG3wFuJkyijxDG17GEye5sq8evCBPh\nnVbXTd77PyH07/XW/n9o+/2s1eWA9/4y4POEwHw1YSyPACusXR4nTAivIkyUnyYE+1WE/r/A8nsz\nsM97f6aV323b/Rdgr/f+bGu/TxAmpVcB59hruy39ycAs7/0Ftu0jhMnrJMKi6ljbz+8SJthlhPGF\nvZcBvkRY1OwkBKRE7jjeyqHj7Du2//MJk/5zhODzA+vPfwPmAWdYPc8B/sXKvA34jLWZyyljiEOL\nlWuY+tDsJGGumWPl7QCuJASHfcAnCXPKPxCOzz2E8XU8cAxhDL6WEMAvBP6UEBxPsT57EngvYfH6\nT4TAus/yusP67y7gfmunWbZN8t3c91vfeGCTc+5EwkJt3PrkrcAl1q+fsroW1aggsDPr92e999cQ\nJveXc9JlX0c9hnDgnUk4gPoJDbSRELnPJ6xIDhAmyQ2EgZmx18YJB/Qs+5myf1sIAyYpPynzgKXN\nru+1hAE+YNtOWJrd2GUfwiTyAuFgeBmYdM7NBa6ybf6Fqe2+xyamDKGj7wH+wvJNWd7zmToxJ98f\ncIRzbpZtO0xYhawirETutX8vESb5bqvXPtvH5DHbZDWVXUZSvyVWj1mEwbSRcID9s7VbKqudPOFs\nIcn7gOWRycpvYdbrk1nti7XfXxEm3ANZ2+zOquesrLx7CZPhY1l1HLft9luabzB1bEAYR9cQDt5d\nWeVcTJjM01l1nG3177L0DxAmhjcSVn9w6CDbCHzZ2mGCsLJL6pG9r8lY+hLw760t91ualO1nMoHu\nt3xSwEY7TrzVeydhdZmU/0n7OYepYyUZ7xAmk0sJk0tyDGD7uMd+Jts8ZuXdS5j4k3ofTxgXb7PX\ndhEWTHcRAs4fWD5vJywsfkg4I1iQVd4+psoexydw6Dh7mkPH2acIE/J51iYdhLOVn3Lo2wk6LI8O\nq8crts2sPGVMEhZBqzl0rGVL2n2O7TO2zQ7CIuDPCMfWKkKAGiIEpS8RFgYT1i5HWj1+QjjLed7q\nMIsQyDKWb3IWloyTPfb7bxHG5a8I/X2AQ8f1Osv/ReurDZb//qx9LfmQbq2+NqKY7AH6TeB459zN\nhJXkRTnp7gFudc4ll2f+kdAg7yNMAksIK5pews6+ANxKiNynEhryWcLk9CHCBDZJWMlcRViBnUxY\nBf2aELnvc879ueWXTECJAcIp4+sIK7L/TWj0N9r2RxNWHqcCH7f8fgX8LWFQjxEO/A7CgJkkdOid\nhNXKZmCRPWm90ur7SULnDwETzrmfEVY/LwDrCQOrk3Dq+AZrp28TLjn9G2HVNEhYndxGCAbfszxP\nsf0829phvnPuXcAPnHObbJ8GrZ7fsDxXEsbNAqZOWNcQ+vN79t5659xfEgbxUXaJ63LgdXb9dgHh\nzOA9AN77bzrnbgAe8N5nnHPjHFrhfM1+3mllfN7q/O8Ik+i49cvL1o4nEybuhVn7+ILVc9za/12E\ngP0Rwmp8n7Vfh+3zl4HfI0z6/836+feAtxBWq39NmJD2WblP2O99hInpKkKgWGr9/2nCBILt70vO\nuT5rn+sJZ65vtt+zTRImu5ucc8OEs4zs926wNkmC+P8lXHK7nHBZ5k+z0r9AuMz2DsIl0/udc2sI\nZzDXEValEC4VfdU593rCeMn+LGW35b/Q2m2JtWGPlf+Mte1PbN+fIiwYLiKM027C6j6ZByaZOo43\nceg4eyOHgtznCGeYPyJczhomTHorORTQ7yVMij8jTNw3EY7T3jxlYPu7HVhLmOyvt/pNAn8D/HdC\nX68mHNtzCXPSWYRx20cYT79lP0+y9l5AWEy8iXDpaC8hoNxPmMO+bvtzArZQtPr83H5+kDDHPGvl\nO2vz4wlj8RjCQu/nhOPt/c659bbvlxIC0V3WPyXpYbEczrnLCI3cDTzkvX8sT5q1wDe89z+pYrmv\nJqxM1nnvn7IgcKn3/vJqldEI5bSVc2424eB6yHt/b90qV4Jz7kHv/R/WKO/1wEve+5uqkFdd29jK\nm++9v6pk4jLrN1O1KsMWRe8mBPht3vuq3LBSoKzZhIXlfsLZ+Vrvfc2/b1tBQEQkYlE+JyAiIoGC\ngIhIxBQEREQipiAgIhIxBQERkYgpCIiIROz/A6snXnL08x/kAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x116c287b8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(['helicased',\n",
       "  'rmd',\n",
       "  'integrased',\n",
       "  'portald',\n",
       "  'hypd',\n",
       "  'nucleard',\n",
       "  'structurald',\n",
       "  'envd',\n",
       "  'polyd',\n",
       "  'metd',\n",
       "  'translationd',\n",
       "  'capsidd',\n",
       "  'entryd',\n",
       "  'otherd',\n",
       "  'matrixd',\n",
       "  'polymerased',\n",
       "  'lysisd',\n",
       "  'replicased',\n",
       "  'taild',\n",
       "  'packingd',\n",
       "  'nucleased',\n",
       "  'recombd',\n",
       "  'ligased'],\n",
       " [[7,\n",
       "   0,\n",
       "   0,\n",
       "   0,\n",
       "   163,\n",
       "   0,\n",
       "   81,\n",
       "   170,\n",
       "   343,\n",
       "   2,\n",
       "   0,\n",
       "   360,\n",
       "   118,\n",
       "   1244,\n",
       "   61,\n",
       "   225,\n",
       "   8,\n",
       "   127,\n",
       "   0,\n",
       "   0,\n",
       "   0,\n",
       "   0,\n",
       "   0],\n",
       "  [0,\n",
       "   48,\n",
       "   2,\n",
       "   0,\n",
       "   302,\n",
       "   50,\n",
       "   110,\n",
       "   17,\n",
       "   0,\n",
       "   0,\n",
       "   0,\n",
       "   408,\n",
       "   80,\n",
       "   1105,\n",
       "   0,\n",
       "   0,\n",
       "   7,\n",
       "   353,\n",
       "   0,\n",
       "   7,\n",
       "   0,\n",
       "   0,\n",
       "   0],\n",
       "  [1120,\n",
       "   1659,\n",
       "   245,\n",
       "   617,\n",
       "   64428,\n",
       "   0,\n",
       "   1137,\n",
       "   1522,\n",
       "   1,\n",
       "   1668,\n",
       "   143,\n",
       "   2581,\n",
       "   35,\n",
       "   28851,\n",
       "   607,\n",
       "   1365,\n",
       "   1604,\n",
       "   693,\n",
       "   3939,\n",
       "   1262,\n",
       "   1947,\n",
       "   262,\n",
       "   352],\n",
       "  [310,\n",
       "   441,\n",
       "   4,\n",
       "   1,\n",
       "   11139,\n",
       "   0,\n",
       "   42,\n",
       "   1178,\n",
       "   0,\n",
       "   541,\n",
       "   53,\n",
       "   581,\n",
       "   11,\n",
       "   9847,\n",
       "   606,\n",
       "   363,\n",
       "   106,\n",
       "   74,\n",
       "   4,\n",
       "   152,\n",
       "   218,\n",
       "   5,\n",
       "   38],\n",
       "  [794,\n",
       "   1191,\n",
       "   228,\n",
       "   613,\n",
       "   51526,\n",
       "   0,\n",
       "   1062,\n",
       "   322,\n",
       "   1,\n",
       "   1111,\n",
       "   90,\n",
       "   1982,\n",
       "   24,\n",
       "   18782,\n",
       "   1,\n",
       "   997,\n",
       "   1496,\n",
       "   618,\n",
       "   3925,\n",
       "   1104,\n",
       "   1721,\n",
       "   255,\n",
       "   313],\n",
       "  [223,\n",
       "   454,\n",
       "   142,\n",
       "   273,\n",
       "   16642,\n",
       "   0,\n",
       "   374,\n",
       "   45,\n",
       "   0,\n",
       "   260,\n",
       "   13,\n",
       "   679,\n",
       "   2,\n",
       "   9589,\n",
       "   0,\n",
       "   225,\n",
       "   652,\n",
       "   234,\n",
       "   1252,\n",
       "   599,\n",
       "   486,\n",
       "   75,\n",
       "   33],\n",
       "  [416,\n",
       "   579,\n",
       "   39,\n",
       "   263,\n",
       "   25483,\n",
       "   0,\n",
       "   444,\n",
       "   237,\n",
       "   0,\n",
       "   694,\n",
       "   70,\n",
       "   932,\n",
       "   3,\n",
       "   6216,\n",
       "   1,\n",
       "   447,\n",
       "   492,\n",
       "   283,\n",
       "   2242,\n",
       "   299,\n",
       "   889,\n",
       "   148,\n",
       "   204],\n",
       "  [120,\n",
       "   91,\n",
       "   37,\n",
       "   61,\n",
       "   7175,\n",
       "   0,\n",
       "   225,\n",
       "   32,\n",
       "   1,\n",
       "   110,\n",
       "   2,\n",
       "   313,\n",
       "   19,\n",
       "   2550,\n",
       "   0,\n",
       "   291,\n",
       "   302,\n",
       "   84,\n",
       "   364,\n",
       "   168,\n",
       "   289,\n",
       "   23,\n",
       "   71],\n",
       "  [29,\n",
       "   42,\n",
       "   0,\n",
       "   0,\n",
       "   931,\n",
       "   0,\n",
       "   0,\n",
       "   18,\n",
       "   0,\n",
       "   31,\n",
       "   1,\n",
       "   29,\n",
       "   7,\n",
       "   1847,\n",
       "   0,\n",
       "   15,\n",
       "   3,\n",
       "   2,\n",
       "   0,\n",
       "   0,\n",
       "   15,\n",
       "   0,\n",
       "   7],\n",
       "  [31,\n",
       "   134,\n",
       "   0,\n",
       "   0,\n",
       "   689,\n",
       "   0,\n",
       "   2,\n",
       "   112,\n",
       "   0,\n",
       "   79,\n",
       "   2,\n",
       "   4,\n",
       "   0,\n",
       "   1498,\n",
       "   3,\n",
       "   148,\n",
       "   3,\n",
       "   22,\n",
       "   0,\n",
       "   9,\n",
       "   7,\n",
       "   1,\n",
       "   13],\n",
       "  [150,\n",
       "   77,\n",
       "   0,\n",
       "   1,\n",
       "   323,\n",
       "   0,\n",
       "   1,\n",
       "   970,\n",
       "   0,\n",
       "   157,\n",
       "   2,\n",
       "   371,\n",
       "   2,\n",
       "   2722,\n",
       "   601,\n",
       "   76,\n",
       "   1,\n",
       "   10,\n",
       "   0,\n",
       "   142,\n",
       "   40,\n",
       "   0,\n",
       "   1],\n",
       "  [0, 0, 0, 0, 19, 0, 1, 0, 0, 0, 0, 35, 0, 435, 0, 0, 0, 6, 0, 0, 0, 0, 0],\n",
       "  [0, 2, 0, 0, 42, 0, 0, 42, 0, 0, 0, 81, 0, 865, 0, 26, 0, 0, 0, 0, 0, 0, 0],\n",
       "  [0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 27, 0, 189, 0, 0, 0, 0, 0, 0, 0, 0, 0]])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# creating two functions for plotting the functional gene categories of each viral group\n",
    "def proteinBar (protMat):\n",
    "    protDict={} \n",
    "    capsid = ['coat', 'capsid', 'head', 'CP', 'Cap protein', 'cap protein']\n",
    "    matrix = ['matrix', 'tegument']\n",
    "    env =['envelope protein', 'envelope glycoprotein', 'membrane protein', 'membrane glycoprotein',\n",
    "          'glycoprotein', 'spike protein']\n",
    "    portal = ['portal','neck', 'head-tail connector', 'head-tail joining'] \n",
    "    tail= [ 'tail','baseplate', 'tape measure']\n",
    "    packing = ['terminase', 'TerL','TerS', 'scaffold protein', 'prohead protease', 'DNA packaging']\n",
    "    lysis = ['holin', 'endolysin', 'lysozyme','lysis', 'lysin', 'N-acetylmuramoyl-L-alanine amidase',\n",
    "             'cell wall hydrolase','hydrolase']\n",
    "    nuclease = ['nuclease', 'DNAse', 'DNase','RNAse', 'RNase']\n",
    "    integrase =['integrase']\n",
    "    polymerase =['polymerase']\n",
    "    helicase =['helicase']\n",
    "    replicase =['topisomerase', 'primase','Rep protein','rep protein','replicase', 'replication', \n",
    "                'holliday junction resolvase','clamp loader', 'sliding clamp']\n",
    "    regulationModification = ['transcriptional enhancer', 'transcriptional repressor',\n",
    "                              'transcriptional activation protein', 'transcription regulator', 'transcription factor',\n",
    "                              'transcriptional regulator','transcriptional activator','expression regulator','promoter',\n",
    "                              'antirepressor','anti-repressor', 'thioredoxin','DNA methylase','methyltransferase',\n",
    "                              'ADP-ribosylase','ADP-ribosyltransferase', 'sigma factor', 'phosphatase', \n",
    "                              'apoptosis regulator','growth factor', 'ubiquitin']\n",
    "    metabolismSynth = ['deaminase', 'thymidylate synthase','dihydrofolate reductase', 'ribonucleotide reductase',\n",
    "                       'ribonucleotide-reductase', 'ribonucleoside-diphosphate reductase','anaerobic NTP reductase',\n",
    "                       'oxidoreductase', 'kinase','dUTPase', 'ATPase','dCMP hydroxymethylase','amidase',\n",
    "                       'phosphate starvation-inducible protein', 'deoxyuridine triphosphatase', 'dCTP pyrophosphatase',\n",
    "                       'uracil DNA glycosylase' ]\n",
    "    entryInjection = ['movement protein','injection', 'fusion protein']\n",
    "    hyp=['hypothetical protein', 'putative uncharacterised protein']\n",
    "    ligase =['ligase']\n",
    "    structural=['structural protein']\n",
    "    recomb = ['recombination','recombinase']\n",
    "    translation = ['translation initiation', 'translation elongation', 'tRNA', 'translation factor']\n",
    "    nuclear=['nuclear shuttle protein', 'nuclear transporter']\n",
    "    poly=['polyprotein', 'readthrough protein', 'read-through protein']\n",
    "\n",
    "    #initiating several lists that will hold genes labels with \n",
    "    capsidVec=[]\n",
    "    matrixVec=[]\n",
    "    envVec=[]\n",
    "    portalVec=[]        \n",
    "    tailVec=[]\n",
    "    polymeraseVec=[]\n",
    "    helicaseVec=[]\n",
    "    replicaseVec=[]\n",
    "    integraseVec=[]       \n",
    "    nucleaseVec=[]           \n",
    "    packingVec=[]\n",
    "    lysisVec=[]    \n",
    "    rmVec=[]\n",
    "    translationVec=[]        \n",
    "    metVec=[]      \n",
    "    entryVec=[]  \n",
    "    hypVec=[]\n",
    "    otherVec=[]\n",
    "    ligaseVec =[]\n",
    "    structuralVec=[]\n",
    "    recombVec=[]\n",
    "    nuclearVec=[]\n",
    "    polyVec=[]\n",
    "    #Checking to see which functional category each gene belongs to\n",
    "    for vec in protMat:\n",
    "        for i in vec:          \n",
    "            if any(substring in i for substring in capsid):\n",
    "                capsidVec.append(i)\n",
    "            elif any(substring in i for substring in matrix):\n",
    "                matrixVec.append(i)\n",
    "            elif any(substring in i for substring in env):\n",
    "                envVec.append(i)  \n",
    "            elif any(substring in i for substring in portal):\n",
    "                portalVec.append(i)        \n",
    "            elif any(substring in i for substring in tail):\n",
    "                tailVec.append(i)\n",
    "            elif any(substring in i for substring in polymerase):\n",
    "                polymeraseVec.append(i)\n",
    "            elif any(substring in i for substring in helicase):\n",
    "                helicaseVec.append(i)\n",
    "            elif any(substring in i for substring in replicase):\n",
    "                replicaseVec.append(i)\n",
    "            elif any(substring in i for substring in integrase):\n",
    "                integraseVec.append(i)       \n",
    "            elif any(substring in i for substring in nuclease):\n",
    "                nucleaseVec.append(i)           \n",
    "            elif any(substring in i for substring in packing):\n",
    "                packingVec.append(i)\n",
    "            elif any(substring in i for substring in lysis):\n",
    "                lysisVec.append(i)    \n",
    "            elif any(substring in i for substring in regulationModification):\n",
    "                rmVec.append(i)\n",
    "            elif any(substring in i for substring in translation):\n",
    "                translationVec.append(i)        \n",
    "            elif any(substring in i for substring in metabolismSynth):\n",
    "                metVec.append(i)      \n",
    "            elif any(substring in i for substring in entryInjection):\n",
    "                entryVec.append(i)            \n",
    "            elif any(substring in i for substring in ligase):\n",
    "                ligaseVec.append(i)\n",
    "            elif any(substring in i for substring in structural):\n",
    "                structuralVec.append(i)\n",
    "            elif any(substring in i for substring in recomb):\n",
    "                recombVec.append(i)\n",
    "            elif any(substring in i for substring in nuclear):\n",
    "                nuclearVec.append(i)\n",
    "            elif any(substring in i for substring in poly):\n",
    "                polyVec.append(i)\n",
    "            elif any(substring in i for substring in hyp):\n",
    "                hypVec.append(i)\n",
    "            else:       \n",
    "                otherVec.append(i)\n",
    "    protDict={'capsidd':capsidVec, 'matrixd': matrixVec, 'envd': envVec, 'portald': portalVec, \n",
    "              'taild': tailVec, 'structurald': structuralVec,'polymerased':polymeraseVec, \n",
    "              'replicased':replicaseVec, 'rmd':rmVec,'translationd':translationVec, 'helicased':helicaseVec,\n",
    "              'integrased': integraseVec,'ligased':ligaseVec,'nucleased': nucleaseVec, 'recombd': recombVec,\n",
    "              'metd':metVec,'entryd': entryVec,'nucleard': nuclearVec,'packingd': packingVec, 'lysisd':lysisVec,\n",
    "              'polyd': polyVec,'hypd': hypVec, 'otherd': otherVec}\n",
    "\n",
    "    return(protDict)\n",
    "\n",
    "#putting all dictionaries in a list\n",
    "protDictList=[]\n",
    "\n",
    "#indices of proteDictList = 0 RNA, 1 ssDNA, 2 dsDNA, 3 euk dsDNA, 4 bact dsDNA, 5 bact sipho, 6 bact myo,\n",
    "#7 bact podo, and then a bunch of eukaryotic dsDNA viruses\n",
    "#going through different viral categories\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'dsRNA viruses' in  v['englishLineage'] \n",
    "          or 'ssRNA positive-strand viruses, no DNA stage' in v['englishLineage']\n",
    "         or 'ssRNA negative-strand viruses' in v['englishLineage'] or 'Metaviridae' in v['englishLineage'] \n",
    "          or 'Retroviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'ssDNA viruses' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] \n",
    "          and 'dsDNA viruses, no RNA stage' in v['englishLineage']] \n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] \n",
    "          and 'dsDNA viruses, no RNA stage' in v['englishLineage']] \n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] \n",
    "          and 'Siphoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] \n",
    "          and 'Myoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] \n",
    "          and 'Podoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] \n",
    "          and 'Baculoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] \n",
    "          and 'Poxviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] \n",
    "          and 'Herpesvirales' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] \n",
    "          and 'Papillomaviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] \n",
    "          and 'Adenoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] \n",
    "          and 'Polyomaviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "#this function was used for plotting abundances of each gene functional cateogry, but actually ended up using excel to \n",
    "#create bar plots. Both methods deal with the same data, just asthetically different.\n",
    "def getXYs(ind):     \n",
    "    #ind is a vector containing the list of indices (viral groups) we're interested in plotting together\n",
    "    #labels is a vector of strings that describe the viral groups being plotted\n",
    "    bar_width = 0.17\n",
    "    #m is the number of groups we're plotting \n",
    "    m = len(ind)\n",
    "    #the multiplier for bar_width so the bars can be placed next to each other\n",
    "    n= 0\n",
    "    #starting and updating the index of labels that we need to pull out. \n",
    "    #Corresponds to the labe of the viral group we're plotting\n",
    "    n2 =-1\n",
    "    #yvec will contain the number of proteins associated with each protein category, and all viral groups \n",
    "    yvec =[]\n",
    "    \n",
    "    for i in ind:\n",
    "    #i corresponds to the indices of the protein groups we want to plot from protDictList\n",
    "        n=n+1\n",
    "        n2 =n2 +1\n",
    "        #initializing the x and y coordinates for each bar plot\n",
    "        y=[]\n",
    "        x=[]\n",
    "    \n",
    "        for key in protDictList[i]: \n",
    "            #going through different protein categories (keys) to build the x vector labels\n",
    "            #the x values are just the number of categories, multiplied by n (which keeps getting updated)\n",
    "            #so the bar plots don't overlap\n",
    "            x.append(key)\n",
    "            #counting the number of items associated with each protein category (or key) to build the y vector\n",
    "            y.append(len(protDictList[i].get(key)))    \n",
    "        \n",
    "        xvec = list(range(len(y)))\n",
    "        xvalnew = [n + xval * m for xval in xvec] \n",
    "        yvec.append(y)\n",
    "        plt.bar(xvalnew, y, align='center')\n",
    "        plt.xticks(xvalnew, x)\n",
    "    plt.ylim(0,5000)\n",
    "    plt.show()\n",
    "    return(x, yvec)\n",
    "getXYs(list(range(14)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
