{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Gene order and gene functional category abundances\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import sklearn\n",
    "import matplotlib.pyplot as plt\n",
    "import ete3 as ete\n",
    "import os\n",
    "import Bio\n",
    "import random\n",
    "import scipy.stats as stats\n",
    "from functools import *\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "import matplotlib.patches as mpatches\n",
    "import seaborn as sns\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "\n",
    "#as part of quality control the following viruses had to be omitted from our analysis\n",
    "#they contain misannotations or mislabeled hosts, etc. \n",
    "badFolders = ['Acyrthosiphon_pisum_secondary_endosymbiont_phage_1_uid14047', 'Aureococcus_anophagefferens_virus_MM_2014_uid258005', 'Dragonfly_associated_microphage_1_uid177547', 'Enterobacter_phage_PG7_uid240014', 'Escherichia_phage_vB_EcoM_JS09_uid248321', 'Lactobacillus_phage_LL_H_uid19803', 'Providencia_phage_Redjac_uid177540', 'Pseudoalteromonas_phage_RIO_1_uid206039', 'Sputnik_virophage_2_uid243067', 'Sputnik_virophage_3_uid243065', 'Staphylococcus_phage_6ec_uid253318', 'Streptococcus_phage_DCC1738_uid253219', 'Streptococcus_phage_IC1_uid253220', 'Streptococcus_phage_K13_uid253223', 'Streptococcus_phage_SMP_uid18529', 'Vibrio_phage_CP_T1_uid181062', 'Vibrio_phage_vpms1_uid212709', 'Vibrio_phage_VPUSM_8_uid227006',  'Abaca_bunchy_top_virus_uid28697','Shallot_yellow_stripe_virus_uid15745','Equine_rhinitis_A_virus_uid15205','African_pouched_rat_arterivirus_uid274797','Spilanthes_yellow_vein_virus_uid19779', 'Velvet_bean_severe_mosaic_virus_uid41175','Paramecium_bursaria_Chlorella_virus_FR483_uid18305','Human_endogenous_retrovirus_K113_uid222261','Acholeplasma_phage_MV_L1_uid14573']\n",
    "#viruses whose lineage is improperly annotated (they don't belong to any Baltimore group)\n",
    "unknowns= ['Bacillus_phage_phBC6A51_uid15021', 'Bacillus_phage_phBC6A52_uid15022', 'Badger_feces_associated_gemycircularvirus_uid281125', 'Chaetoceros_DNA_virus_7_uid237558', 'Chaetoceros_lorenzianus_DNA_Virus_uid63565', 'Chaetoceros_tenuissimus_DNA_virus_uid60753', 'Cladosporium_cladosporioides_virus_1_uid258308', 'Enterobacteria_phage_YYZ_2008_uid32231', 'European_mountain_ash_ringspot_associated_virus_uid39973', 'Faecal_associated_gemycircularvirus_3_uid268557', 'Faecal_associated_gemycircularvirus_4_uid268556', 'Faecal_associated_gemycircularvirus_5_uid268555', 'Faecal_associated_gemycircularvirus_6_uid268554', 'Faecal_associated_gemycircularvirus_7_uid268553', 'Gentian_ovary_ring_spot_virus_uid256090', 'Geobacillus_virus_E2_uid19797', 'Glossina_pallidipes_salivary_gland_hypertrophy_virus_uid28839', 'Haloarcula_hispanica_icosahedral_virus_2_uid109269', 'Haloarcula_phage_SH1_uid15535', 'Halovirus_PH1_uid196975', 'Halovirus_VNH_1_uid262927', 'Human_genital_associated_circular_DNA_virus_1_uid281399', 'Lactobacillus_johnsonii_prophage_Lj771_uid28145', 'Magnaporthe_oryzae_RNA_virus_uid272442', 'Mongoose_feces_associated_gemycircularvirus_a_uid281407', 'Mongoose_feces_associated_gemycircularvirus_b_uid281406', 'Mongoose_feces_associated_gemycircularvirus_c_uid281404', 'Mongoose_feces_associated_gemycircularvirus_d_uid281405', 'Mycoplasma_phage_phiMFV1_uid14387', 'Penicillium_roqueforti_dsRNA_mycovirus_1_uid258162', 'Phage_Gifsy_1_uid32269', 'Phage_Gifsy_2_uid32271', 'Pigeonpea_cryptic_virus_uid244664', 'Planaria_asexual_strain_specific_virus_like_element_type_1_uid14140', 'Pseudomonas_phage_phi_Pto_bp6g_uid240724', 'Rhizoctonia_fumigata_mycovirus_uid283068', 'Rhodococcus_phage_REQ2_uid81171', 'Rhodococcus_phage_REQ3_uid81175', 'Rose_rosette_virus_uid64937', 'Rosellinia_necatrix_fusarivirus_1_uid255787', 'Rosellinia_necatrix_megabirnavirus_1_W779_uid41609', 'Salisaeta_icosahedral_phage_1_uid167575', 'Salmonella_phage_Fels_1_uid29267', 'Sodalis_phage_phiSG1_uid16583', 'Staphylococcus_phage_phi2958PVL_uid32173', 'Staphylococcus_phage_tp310_1_uid20659', 'Staphylococcus_phage_tp310_3_uid20663', 'Stenotrophomonas_phage_phiSMA9_uid15493', 'Streptococcus_phage_20617_uid239271', 'Streptococcus_phage_phiBHN167_uid227353', 'Streptococcus_pyogenes_phage_315_1_uid14533', 'Streptococcus_pyogenes_phage_315_2_uid14528', 'Streptococcus_pyogenes_phage_315_3_uid14529', 'Streptococcus_pyogenes_phage_315_4_uid14530', 'Streptococcus_pyogenes_phage_315_5_uid14531', 'Streptococcus_pyogenes_phage_315_6_uid14532', 'Tanay_virus_uid246129', 'Thermococcus_prieurii_virus_1_uid84407', 'Thermus_phage_IN93_uid14235', 'Thermus_phage_P23_77_uid40235']\n",
    "badFolders=badFolders+unknowns\n",
    "#these are satellite viruses \n",
    "sats = ['Ageratum_conyzoides_associated_symptomless_alphasatellite_uid259293', 'Ageratum_enation_alphasatellite_uid181994', 'Ageratum_leaf_curl_betasatellite_uid195929', 'Ageratum_leaf_curl_Cameroon_betasatellite_uid36669', 'Ageratum_yellow_leaf_curl_betasatellite_uid14439', 'Ageratum_yellow_vein_China_alphasatellite_uid237561', 'Ageratum_yellow_vein_China_virus_associated_DNA_beta_uid15515', 'Ageratum_yellow_vein_Singapore_alphasatellite_uid14232', 'Ageratum_yellow_vein_virus_satellite_DNA_beta_uid14444', 'Alternanthera_yellow_vein_virus_satellite_DNA_beta_uid19833', 'Andrographis_yellow_vein_leaf_curl_betasatellite_uid243492', 'Bhendi_yellow_vein_India_betasatellite__India_Aurangabad_OY164_2006__uid61557', 'Bhendi_yellow_vein_mosaic_betasatellite__India_Coimbator_OYCO1_2005__uid61777', 'Bhendi_yellow_vein_mosaic_virus_associated_alphasatellite_uid174781', 'Bhendi_yellow_vein_mosaic_virus_satellite_DNA_beta_uid14445', 'Black_medic_leafroll_alphasatellite_1_uid243500', 'Cardamom_bushy_dwarf_virus_satellite_uid230064', 'Cassava_mosaic_Madagascar_alphasatellite_uid175666', 'Chili_leaf_curl_Bhatinda_betasatellite_uid206467', 'Chilli_leaf_curl_alphasatellite_uid253224', 'Chilli_leaf_curl_Multan_alphasatellite_uid39933', 'Chilli_leaf_curl_virus_satellite_DNA_beta_uid14441', 'Cleome_leaf_crumple_virus_associated_DNA_1_uid60045', 'Corchorus_yellow_vein_mosaic_betasatellite_uid192608', 'Cotton_leaf_curl_Burewala_alphasatellite_uid45935', 'Cotton_leaf_curl_Burewala_betasatellite_uid45933', 'Cotton_leaf_curl_Gezira_alphasatellite_uid42507', 'Cotton_leaf_curl_Gezira_virus_satellite_DNA_beta_uid15166', 'Cotton_leaf_curl_Multan_betasatellite_uid15780', 'Cotton_leaf_curl_virus_associated_DNA_1_isolate_Lucknow_uid65305', 'Cotton_leaf_curl_virus_associated_DNA_beta_uid14438', 'Cotton_leaf_curl_virus_betasatellite_uid162497', 'Cowpea_severe_leaf_curl_associated_DNA_beta_uid15157', 'Croton_yellow_vein_mosaic_alphasatellite_uid45931', 'Croton_yellow_vein_mosaic_betasatellite_uid18249', 'Cuban_alphasatellite_1_uid210798', 'Dragonfly_associated_alphasatellite_uid181244', 'Emilia_yellow_vein_virus_associated_DNA_beta_uid37893', 'Erectites_yellow_mosaic_virus_satellite_DNA_beta_uid19827', 'Eupatorium_yellow_vein_virus_satellite_DNA_beta_uid14447', 'Faba_bean_necrotic_stunt_alphasatellite_1_uid243499', 'Faba_bean_necrotic_stunt_alphasatellite_2_uid243498', 'French_bean_leaf_curl_betasatellite_Kanpur_uid169556', 'Gossypium_darwinii_symptomless_alphasatellite_uid39593', 'Gossypium_davidsonii_symptomless_alphasatellite_uid39589', 'Gossypium_mustilinum_symptomless_alphasatellite_uid39591', 'Grapevine_satellite_virus_uid208539', 'Guar_leaf_curl_alphasatellite_uid193981', 'Hedyotis_uncinella_yellow_mosaic_betasatellite_uid230991', 'Honeysuckle_yellow_vein_mosaic_disease_associated_satellite_DNA_beta_uid19863', 'Honeysuckle_yellow_vein_mosaic_virus_satellite_DNA_beta_uid14620', 'Leucas_zeylanica_yellow_vein_virus_satellite_DNA_beta_uid41305', 'Ludwigia_leaf_distortion_betasatellite__India_Amadalavalasa_Hibiscus_2007__uid29233', 'Ludwigia_yellow_vein_virus_associated_DNA_beta_uid15561', 'Luffa_puckering_and_leaf_distortion_associated_betasatellite__India_Gurdaspur_Okra_2013__uid253218', 'Luffa_puckering_and_leaf_distortion_associated_DNA_beta_uid15779', 'Malachra_yellow_vein_mosaic_virus_associated_satellite_DNA_beta_uid28727', 'Malvastrum_leaf_curl_betasatellite_uid16301', 'Malvastrum_leaf_curl_Guangdong_betasatellite_uid243827', 'Malvastrum_yellow_vein_Yunnan_virus_satellite_DNA_beta_uid14567', 'Melon_chlorotic_mosaic_virus_associated_alphasatellite_uid51413', 'Mesta_yellow_vein_mosaic_virus_alphasatellite__India_Hoshiarpur_Okra_2013__uid253217', 'Mesta_yellow_vein_mosaic_virus_associated_alphasatellite_uid174780', 'Mesta_yellow_vein_mosaic_virus_associated_DNA_beta_uid21015', 'Milk_vetch_dwarf_C10_alphasatellite_uid240682', 'Mirabilis_leaf_curl_India_virus_associated_betasatellite_uid255800', 'Mungbean_yellow_mosaic_India_virus_associated_betasatellite__India__Faizabad__Cow_Pea_2012__uid177773', 'Nanovirus_like_particle_uid14386', 'Okra_enation_leaf_curl_betasatellite__India_Sonipat_EL10_2006__uid61781', 'Okra_leaf_curl_Mali_virus_satellite_DNA_beta_uid20323', 'Okra_yellow_vein_disease_associated_sequence_uid14443', 'Olive_viral_satellite_RNA_uid230268', 'Papaya_leaf_curl_alphasatellite_uid233414', 'Papaya_leaf_curl_China_virus_satellite_DNA_beta_uid19819', 'Pepper_leaf_curl_virus_satellite_DNA_beta_uid28283', 'Pepper_leaf_curl_Yunnan_virus_satellite_DNA_beta_uid29415', 'Potato_apical_leaf_curl_disease_associated_satellite_DNA_beta_uid18323', 'Radish_leaf_curl_virus_betasatellite_uid28281', 'Sida_leaf_curl_virus_satellite_DNA_beta_uid19823', 'Sida_yellow_mosaic_virus_China_associated_DNA_beta_uid15514', 'Sida_yellow_vein_China_alphasatellite_uid238950', 'Siegesbeckia_yellow_vein_virus_associated_DNA_beta_uid17269', 'Subterranean_clover_stunt_C6_alphasatellite_uid14180', 'Synedrella_leaf_curl_alphasatellite_uid263034', 'Tobacco_leaf_chlorosis_betasatellite_uid178075', 'Tobacco_leaf_curl_disease_associated_sequence_uid14442', 'Tobacco_leaf_curl_PUSA_alphasatellite_uid56023', 'Tobacco_leaf_curl_virus_associated_DNA_beta_uid45925', 'Tomato_leaf_curl_Bangladesh_betasatellite_uid56017', 'Tomato_leaf_curl_Gandhinagar_betasatellite_uid231683', 'Tomato_leaf_curl_Hajipur_betasatellite_uid175587', 'Tomato_leaf_curl_Joydebpur_virus_satellite_DNA_beta_uid28273', 'Tomato_leaf_curl_Karnataka_alphasatellite_uid181995', 'Tomato_leaf_curl_Pakistan_virus_associated_DNA_1_uid38463', 'Tomato_leaf_curl_Patna_betasatellite_uid36541', 'Tomato_leaf_curl_Togo_betasatellite_Togo_2006__uid60629', 'Tomato_leaf_curl_Yemen_betasatellite_uid177643', 'Tomato_yellow_leaf_curl_China_betasatellite_uid181248', 'Tomato_yellow_leaf_curl_Vietnam_virus_satellite_DNA_beta_uid19829', 'TYLCCNV_Y322__satellite_DNA_beta_uid16338', 'Vernonia_yellow_vein_betasatellite_uid41303', 'Vernonia_yellow_vein_Fujian_virus_alphasatellite_uid72145', 'Vernonia_yellow_vein_Fujian_virus_betasatellite_uid72143', 'Zinnia_leaf_curl_disease_associated_sequence_uid14440', 'Zinnia_leaf_curl_virus_associated_DNA_beta_uid14538']\n",
    "\n",
    "\n",
    "def getTaxidFromGenomeFolder(genome_folder):\n",
    "    #Takes an rpt file for a genome and returns the taxid in it.\n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        with open(os.path.join('all_rpt', folder, os.listdir(os.path.join('all_rpt', folder))[0])) as rpt_file:\n",
    "            for line in rpt_file:\n",
    "                if line.startswith('Taxid: '):\n",
    "                    return int(line[7:-1])\n",
    "                \n",
    "\n",
    "def getHostTaxidFromGenomeFolder(genome_folder):\n",
    "    #Takes a folder containing an asn file and returns the taxid of the host of the virus described by that asn file. \n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        with open('folderToHostTaxid.txt', 'r') as taxid2host:\n",
    "            for line in taxid2host:\n",
    "                if genome_folder in line:\n",
    "                    return int(line.split('\\t')[1][:-1])\n",
    "        return 0\n",
    "    \n",
    "\n",
    "def getGenomeLengthFromGenomeFolder(genome_folder):\n",
    "    total = 0\n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        for file in os.listdir(os.path.join('all_ptt', folder)):\n",
    "            with open(os.path.join('all_ptt', folder, file)) as ptt_file:\n",
    "                line = ptt_file.readline()\n",
    "                index = line.find('- 1..')\n",
    "                if index == -1:\n",
    "                    return -1\n",
    "                num = line[index + 5 : -1]\n",
    "                total += int(num)\n",
    "        return total\n",
    "    \n",
    "\n",
    "def getCodingNumberAndOverlapNumberFromGenomeFolder(folder):\n",
    "    totalCoding = 0\n",
    "    totalOverlap = 0\n",
    "    for file in os.listdir(os.path.join('all_ptt', folder)):\n",
    "        #these folders contain genomes with misannotation of gene overlap,non-coding DNA, or host information from NCBI. \n",
    "        if folder not in badFolders and folder not in sats:\n",
    "            with open(os.path.join('all_ptt', folder, file)) as ptt_file:\n",
    "                line = ptt_file.readline()\n",
    "                index = line.find('- 1..')\n",
    "                repliconLength = int(line[index + 5 : -1])\n",
    "                indicesThatOverlap = [0] * repliconLength\n",
    "                line = ptt_file.readline()\n",
    "                line = ptt_file.readline()\n",
    "\n",
    "                uniqueIndices = set([])\n",
    "                for line in ptt_file:                                     \n",
    "                    #this part is to collect beginning and end indices for genes\n",
    "                    line = line.split('\\t')[0]\n",
    "                    [start, end] = [int(x) for x in line.split('..')]\n",
    "                    #this is put in place to distinguish linear from circular genomes. Here, we're looking for linear genomes\n",
    "                    if start < end:                    \n",
    "                        for i in range(start, end):\n",
    "                            if i in uniqueIndices:\n",
    "                                indicesThatOverlap[i] = 1\n",
    "                            else:\n",
    "                                uniqueIndices.add(i)\n",
    "                    #here were are looking for circular genomes to be able to properly count noncoding bases\n",
    "                    elif start > end:\n",
    "                        for i in range(start, repliconLength):\n",
    "                            if i in uniqueIndices:\n",
    "                                indicesThatOverlap[i] = 1\n",
    "                            else:\n",
    "                                uniqueIndices.add(i)\n",
    "                        for i in range(1, end):\n",
    "                            if i in uniqueIndices:\n",
    "                                indicesThatOverlap[i] = 1\n",
    "                            else:\n",
    "                                uniqueIndices.add(i)          \n",
    "                \n",
    "                totalCoding += len(uniqueIndices)\n",
    "                #totalOverlap += len(bigList) - len(uniqueIndices)\n",
    "                totalOverlap += sum(indicesThatOverlap)\n",
    "    return (totalCoding, totalOverlap)\n",
    "\n",
    "\n",
    "def getProtienNamesFromGenomeFolder(folder):\n",
    "    protVec=[]\n",
    "    for file in os.listdir(os.path.join('all_ptt', folder)):\n",
    "        #these folders contain genomes with misannotation of gene overlap, non-coding DNA, or host information from NCBI. \n",
    "        if folder not in badFolders and folder not in sats:\n",
    "            with open(os.path.join('all_ptt', folder, file)) as ptt_file:\n",
    "                line = ptt_file.readline()\n",
    "                line = ptt_file.readline()\n",
    "                line = ptt_file.readline()               \n",
    "                for line in ptt_file:                    \n",
    "                    #this is to collect the name of proteins from ptt files                   \n",
    "                    protVec.append(line.split('\\t')[-1].rstrip())                                    \n",
    "    return (protVec)\n",
    "\n",
    "\n",
    "genomeLengths = []\n",
    "lengthData = []\n",
    "counter = 0\n",
    "length_lists = []\n",
    "hostTaxidList = []\n",
    "taxidList = []\n",
    "folders = []\n",
    "proteinNames=[]\n",
    "dbObject = ete.NCBITaxa()\n",
    "\n",
    "for folder in os.listdir('all_faa'):\n",
    "    #these entries had problems with their gene overlap, protein, and noncoding annotation so they're excluded\n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        length_lists.append([])\n",
    "        for fileName in os.listdir(os.path.join('all_faa', folder)):\n",
    "            with open(os.path.join('all_faa', folder, fileName), \"r\") as faa_file:\n",
    "                proteins = SeqIO.parse(faa_file, 'fasta')\n",
    "                for protein in proteins:\n",
    "                    length_lists[counter].append(float(len(protein.seq)))\n",
    "\n",
    "        hostTaxid = getHostTaxidFromGenomeFolder(folder)\n",
    "        # These taxids come viruses with unidentified hosts \n",
    "        if hostTaxid and hostTaxid != 11086 and hostTaxid != 10036 and hostTaxid != 37965 and hostTaxid != 10640  and hostTaxid != 32644 and hostTaxid != 1 and hostTaxid != 212035 and hostTaxid != 1051671:\n",
    "            taxidList.append(getTaxidFromGenomeFolder(folder))\n",
    "            hostTaxidList.append(hostTaxid)\n",
    "            genomeLengths.append(getGenomeLengthFromGenomeFolder(folder))\n",
    "            folders.append(folder)\n",
    "            proteinNames.append(getProtienNamesFromGenomeFolder(folder))\n",
    "            counter += 1\n",
    "        else:\n",
    "            length_lists = length_lists[:-1]\n",
    "        \n",
    "        #Uncomment the following lines to see the code's progress while its running\n",
    "        \"\"\"\n",
    "        if counter % 100 == 0:\n",
    "            print(str(int(np.round(float(counter) * 100.0 / 2701.0))) + '% done.')\n",
    "        \"\"\"\n",
    "\n",
    "taxidLineages = [dbObject.get_lineage(t) for t in taxidList]\n",
    "englishLineages = [[str(rank) for rank in dbObject.translate_to_names(lin)]\n",
    "                   for lin in taxidLineages]\n",
    "hostTaxidLineages = [dbObject.get_lineage(t) for t in hostTaxidList]\n",
    "hostEnglishLineages = [[str(rank) for rank in dbObject.translate_to_names(lin)] \n",
    "                       for lin in hostTaxidLineages]\n",
    "\n",
    "for i in range(len(length_lists)):\n",
    "    lengthData.append({'length_list' : length_lists[i],\n",
    "                       'taxid' : taxidList[i],\n",
    "                       'hostTaxid' : hostTaxidList[i],\n",
    "                       'englishLineage' : englishLineages[i],\n",
    "                       'taxidLineage' : taxidLineages[i],\n",
    "                       'hostEnglishLineage' : hostEnglishLineages[i],\n",
    "                       'hostTaxidLineage' : hostTaxidLineages[i],\n",
    "                       'mean' : np.mean(length_lists[i]),\n",
    "                       'std' : np.std(length_lists[i]),\n",
    "                       'median' : np.median(length_lists[i]),\n",
    "                       'numProteins' : len(length_lists[i]),\n",
    "                       'genomeLength' : genomeLengths[i],\n",
    "                       'proteinNames': proteinNames[i],\n",
    "                       'folder' : folders[i]})\n",
    "                       \n",
    "lengthData = [virus for virus in lengthData if virus['genomeLength'] != -1 and virus['median'] != -1]\n",
    "\n",
    "for (i, virus) in enumerate(lengthData):\n",
    "    #print(str(i) + '/' + str(len(lengthData)))\n",
    "    (totalCoding, totalOverlap) = getCodingNumberAndOverlapNumberFromGenomeFolder(virus['folder'])\n",
    "    virus['percentNonCoding'] = 100 * (virus['genomeLength'] - totalCoding) / virus['genomeLength']\n",
    "    virus['percentOverlap'] = 100 * totalOverlap / virus['genomeLength']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#the shared protein coding genes between dsDNA, ssDNA and RNA viruses\n",
    "protMatds= [v['proteinNames'] for v in lengthData if 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "protMatss= [v['proteinNames'] for v in lengthData if 'ssDNA viruses' in v['englishLineage']]\n",
    "protMatRNA= [v['proteinNames'] for v in lengthData if any('dsRNA viruses','ssRNA positive-strand viruses, no DNA stage','ssRNA negative-strand viruses' in v['hostEnglishLineage']\n",
    "\n",
    "\n",
    "ds = [i for vec in protMatds for i in vec]\n",
    "ss = [i for vec in protMatss for i in vec]\n",
    "#rna = [i for vec in protMatRNA for i in vec]\n",
    "\n",
    "#len(set(a) & set(b) & set(e))\n",
    "len(set(ds) & set(ss))\n",
    "#set(e) & set(a)\n",
    "#set(e) & set(a) & set(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#want to produce gene order sequences\n",
    "from collections import Counter\n",
    "\n",
    "\"\"\"\n",
    "protMat= [v['proteinNames'] for v in lengthData]\n",
    "genomeLen= [v['genomeLength'] for v in lengthData]\n",
    "virusName = [v['folder'] for v in lengthData]\n",
    "virusLineage = [v['englishLineage'] for v in lengthData]\n",
    "hostLineage = [v['hostEnglishLineage'] for v in lengthData]\n",
    "\"\"\"\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "genomeLen= [v['genomeLength'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage']  and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "virusName = [v['folder'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage']  and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "virusLineage = [v['englishLineage'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage']  and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "hostLineage = [v['hostEnglishLineage'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage']  and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "\n",
    "#ordering the name list and the protMat to be sorted based on the genome lengths. You can used virusName instead to order based on folder names. \n",
    "protMat = [protMat for (genomeLen, protMat) in sorted (zip(genomeLen, protMat), key=lambda pair:pair[0])]\n",
    "virusName = [virusName for (genomeLen, virusName) in sorted (zip(genomeLen, virusName), key=lambda pair:pair[0])]\n",
    "virusLineage = [virusLineage for (genomeLen, virusLineage) in sorted (zip(genomeLen, virusLineage), key=lambda pair:pair[0])]\n",
    "hostLineage = [hostLineage for (genomeLen, hostLineage) in sorted (zip(genomeLen, hostLineage), key=lambda pair:pair[0])]\n",
    "\n",
    "capsid = ['coat', 'capsid', 'head']\n",
    "matrix = ['matrix', 'tegument']\n",
    "#env = ['envelope'] \n",
    "portal = ['portal','neck', 'head-tail connector', 'head-tail joining'] \n",
    "tail= [ 'tail','baseplate', 'tape measure']\n",
    "packing = ['terminase', 'TerL','TerS', 'scaffold protein', 'prohead protease', 'DNA packaging']\n",
    "lysis = ['holin', 'endolysin', 'lysozyme','lysis', 'lysin', 'N-acetylmuramoyl-L-alanine amidase', 'cell wall hydrolase']\n",
    "nuclease = ['nuclease']\n",
    "integrase =['integrase']\n",
    "polymerase =['polymerase']\n",
    "helicase =['helicase']\n",
    "replication =['Rep protein','rep protein','replicase', 'replication']\n",
    "regulationModification = ['transcriptional enhancer', 'transcriptional regulator', 'expression regulator','promoter', 'antirepressor','anti-repressor', 'thioredoxin','DNA methylase','methyltransferase','ADP-ribosylase','ADP-ribosyltransferase']\n",
    "metabolismSynth = ['thymidine kinase', 'deaminase', 'thymidylate synthase','dihydrofolate reductase', 'ribonucleotide reductase', 'dNMP kinase','dCMP hydroxymethylase','deoxynucleoside monophosphate kinase','amidase','phosphate starvation-inducible protein' ]\n",
    "entryInjection = ['movement protein','injection']\n",
    "trna = ['tRNA']\n",
    "#determing and storing the color-code information for each viral gene\n",
    "colormat=[]\n",
    "virusNameFiltered =[]\n",
    "for vec in protMat:\n",
    "    colorvec=[]\n",
    "    for i in vec:          \n",
    "        if any(substring in i for substring in capsid):\n",
    "            colorvec.append('I')\n",
    "        elif any(substring in i for substring in portal):\n",
    "            colorvec.append('C')         \n",
    "        elif any(substring in i for substring in tail):\n",
    "            colorvec.append('E')          \n",
    "        elif any(substring in i for substring in packing):\n",
    "            colorvec.append('Q')         \n",
    "        else:       \n",
    "            colorvec.append('A')\n",
    "    #colormat.append(colorvec)\n",
    "                  \n",
    "    if colorvec.count('A')/len(colorvec) < 0.85 and len(colorvec)>39:                \n",
    "        colormat.append(colorvec)\n",
    "        #getting the index of the vectors in protMat that passed through the filter. Then I use the index to get the virus names from the virusName matrix \n",
    "        index = protMat.index(vec)\n",
    "        virusNameFiltered.append(virusName[index])\n",
    "        #uncomment the next 2 lines to display the gene order sequences\n",
    "        \"\"\"\n",
    "        print('>gi|',virusName[index])#,'_',virusLineage[index][4],'_',hostLineage[index][3:6])\n",
    "        print(str(colorvec).replace(\",\",\"\").replace(\"'\",\"\").replace(\" \", \"\").strip('[]'))\n",
    "        \"\"\"\n",
    "        #note T4likevirus_uid266638 virus genome has been copied twice, which is why we have taken this virus out in our plots.\n",
    "        #originally we thought it's a genome with its gene order pattern repeated twice, until we looked further at the paper wherein the phage was first sequenced. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nvirusList =['Pseudomonas_phage_gh_1_uid14265','Yersinia_pestis_phage_phiA1122_uid14332','Enterobacteria_phage_T3_uid14336','Salmonella_phage_Vi06_uid64609','Yersinia_phage_Berlin_uid18481','Yersinia_phage_vB_YenP_AP5_uid266772','Synechococcus_phage_S_CBP4_uid266792','Cronobacter_phage_Dev2_uid239722','Citrobacter_phage_CR44b_uid239730','Erwinia_phage_vB_EamP_L1_uid181229','Kluyvera_phage_Kvp1_uid32673','Yersinia_phage_phiYeO3_12_uid14591','Pseudomonas_phage_Phi_S1_uid197298','Pseudomonas_phage_phiPSA2_uid253322','Citrobacter_phage_CR8_uid239727','Pseudomonas_phage_phiIBB_PF7A_uid64561','Klebsiella_phage_KP32_uid42779','Enterobacteria_phage_K1F_uid15880','Pseudomonas_phage_phikF77_uid36373','Pseudomonas_phage_MPK6_uid227001','Pseudomonas_phage_LUZ19_uid28741','Lelliottia_phage_phD2B_uid266769','Phage_Paz_uid230593','Phage_Prado_uid230587','Acinetobacter_phage_Petty_uid240050']\\nfor a in virusList:\\n    printAminoAcidSeq(a)\\n\""
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#defining a function that gives us both the gene order sequence, as well as the COG labels of the proteins that the genes encode. \n",
    "from collections import Counter\n",
    "def printAminoAcidSeq(a):\n",
    "    protMat= [v['proteinNames'] for v in lengthData if a in v['folder']]\n",
    "    #print(protMat)\n",
    "    genomeLen= [v['genomeLength'] for v in lengthData if a in v['folder']]\n",
    "    virusName = [v['folder'] for v in lengthData if a in v['folder']]\n",
    "    virusLineage = [v['englishLineage'] for v in lengthData if a in v['folder']]\n",
    "    hostLineage = [v['hostEnglishLineage'] for v in lengthData if a in v['folder']]\n",
    "    \n",
    "    capsid = ['coat', 'capsid', 'head']\n",
    "    matrix = ['matrix', 'tegument']\n",
    "    neck = ['neck', 'head-tail connector', 'head-tail joining']\n",
    "    env = ['envelope'] \n",
    "    portal = ['portal'] \n",
    "    tail= [ 'tail','baseplate', 'tape measure']\n",
    "    packing = ['terminase', 'TerL','TerS', 'scaffold protein', 'prohead protease', 'DNA packaging']\n",
    "    lysis = ['holin', 'endolysin', 'lysozyme','lysis', 'lysin', 'N-acetylmuramoyl-L-alanine amidase', 'cell wall hydrolase']\n",
    "    nuclease = ['nuclease']\n",
    "    integrase =['integrase']\n",
    "    polymerase =['polymerase']\n",
    "    helicase =['helicase']\n",
    "    replicase =['Rep protein','rep protein','replicase', 'replication']\n",
    "    regulationModification = ['transcriptional enhancer', 'transcriptional regulator', 'expression regulator','promoter', 'antirepressor','anti-repressor', 'thioredoxin','DNA methylase','methyltransferase','ADP-ribosylase','ADP-ribosyltransferase']\n",
    "    metabolismSynth = ['thymidine kinase', 'deaminase', 'thymidylate synthase','dihydrofolate reductase', 'ribonucleotide reductase', 'dNMP kinase','dCMP hydroxymethylase','deoxynucleoside monophosphate kinase','amidase','phosphate starvation-inducible protein' ]\n",
    "    entryInjection = ['movement protein','injection']\n",
    "\n",
    "    #determing and storing the color-code information for each viral protein\n",
    "    for vec in protMat:\n",
    "        colorvec=[]\n",
    "        for i in vec:          \n",
    "            # make sure the search is not case sensitive        \n",
    "            if any(substring in i for substring in capsid):\n",
    "                colorvec.append('I')\n",
    "            elif any(substring in i for substring in matrix):\n",
    "                colorvec.append('R')\n",
    "            elif any(substring in i for substring in neck):\n",
    "                colorvec.append('N')\n",
    "            elif any(substring in i for substring in env):\n",
    "                colorvec.append('D')\n",
    "            elif any(substring in i for substring in portal):\n",
    "                colorvec.append('C')         \n",
    "            elif any(substring in i for substring in tail):\n",
    "                colorvec.append('E')\n",
    "            elif any(substring in i for substring in polymerase):\n",
    "                colorvec.append('T')\n",
    "            elif any(substring in i for substring in helicase):\n",
    "                colorvec.append('L')\n",
    "            elif any(substring in i for substring in replicase):\n",
    "                colorvec.append('S')\n",
    "            elif any(substring in i for substring in integrase):\n",
    "                colorvec.append('K')            \n",
    "            elif any(substring in i for substring in nuclease):\n",
    "                colorvec.append('P')            \n",
    "            elif any(substring in i for substring in packing):\n",
    "                colorvec.append('Q')\n",
    "            elif any(substring in i for substring in lysis):\n",
    "                colorvec.append('G')           \n",
    "            elif any(substring in i for substring in regulationModification):\n",
    "                colorvec.append('Y')\n",
    "            elif any(substring in i for substring in metabolismSynth):\n",
    "                colorvec.append('burlywood')       \n",
    "            elif any(substring in i for substring in entryInjection):\n",
    "                colorvec.append('H')           \n",
    "            else:       \n",
    "                colorvec.append('A') \n",
    "        \n",
    "        #uncomment the next few lines to print out gene order sequences, as well as the COG labels of the proteins that the genes encode.        \n",
    "        \"\"\"\n",
    "        print('>gi|',virusName[0],'_',virusLineage[0][4],'_',hostLineage[0][3])\n",
    "        print(str(colorvec).replace(\",\",\"\").replace(\"'\",\"\").replace(\" \", \"\").strip('[]'))\n",
    "\n",
    "    print(len(colorvec))\n",
    "    \"\"\"\n",
    "    return()\n",
    "\n",
    "#getting this information for the podoviruses in pattern C (uncomment the next few lines)\n",
    "\"\"\"\n",
    "virusList =['Pseudomonas_phage_gh_1_uid14265','Yersinia_pestis_phage_phiA1122_uid14332','Enterobacteria_phage_T3_uid14336','Salmonella_phage_Vi06_uid64609','Yersinia_phage_Berlin_uid18481','Yersinia_phage_vB_YenP_AP5_uid266772','Synechococcus_phage_S_CBP4_uid266792','Cronobacter_phage_Dev2_uid239722','Citrobacter_phage_CR44b_uid239730','Erwinia_phage_vB_EamP_L1_uid181229','Kluyvera_phage_Kvp1_uid32673','Yersinia_phage_phiYeO3_12_uid14591','Pseudomonas_phage_Phi_S1_uid197298','Pseudomonas_phage_phiPSA2_uid253322','Citrobacter_phage_CR8_uid239727','Pseudomonas_phage_phiIBB_PF7A_uid64561','Klebsiella_phage_KP32_uid42779','Enterobacteria_phage_K1F_uid15880','Pseudomonas_phage_phikF77_uid36373','Pseudomonas_phage_MPK6_uid227001','Pseudomonas_phage_LUZ19_uid28741','Lelliottia_phage_phD2B_uid266769','Phage_Paz_uid230593','Phage_Prado_uid230587','Acinetobacter_phage_Petty_uid240050']\n",
    "for a in virusList:\n",
    "    printAminoAcidSeq(a)\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# creating two functions for plotting the functional gene categories of each viral group\n",
    "def proteinBar (protMat):\n",
    "    protDict={} \n",
    "    capsid = ['coat', 'capsid', 'head', 'CP', 'Cap protein', 'cap protein']\n",
    "    matrix = ['matrix', 'tegument']\n",
    "    env =['envelope protein', 'envelope glycoprotein', 'membrane protein', 'membrane glycoprotein', 'glycoprotein', 'spike protein']\n",
    "    portal = ['portal','neck', 'head-tail connector', 'head-tail joining'] \n",
    "    tail= [ 'tail','baseplate', 'tape measure']\n",
    "    packing = ['terminase', 'TerL','TerS', 'scaffold protein', 'prohead protease', 'DNA packaging']\n",
    "    lysis = ['holin', 'endolysin', 'lysozyme','lysis', 'lysin', 'N-acetylmuramoyl-L-alanine amidase', 'cell wall hydrolase','hydrolase']\n",
    "    nuclease = ['nuclease', 'DNAse', 'DNase','RNAse', 'RNase']\n",
    "    integrase =['integrase']\n",
    "    polymerase =['polymerase']\n",
    "    helicase =['helicase']\n",
    "    replicase =['topisomerase', 'primase','Rep protein','rep protein','replicase', 'replication', 'holliday junction resolvase','clamp loader', 'sliding clamp']\n",
    "    regulationModification = ['transcriptional enhancer', 'transcriptional repressor','transcriptional activation protein', 'transcription regulator', 'transcription factor', 'transcriptional regulator','transcriptional activator','expression regulator','promoter', 'antirepressor','anti-repressor', 'thioredoxin','DNA methylase','methyltransferase','ADP-ribosylase','ADP-ribosyltransferase', 'sigma factor', 'phosphatase', 'apoptosis regulator'\n",
    "                             'growth factor', 'ubiquitin']\n",
    "    metabolismSynth = ['deaminase', 'thymidylate synthase','dihydrofolate reductase', 'ribonucleotide reductase','ribonucleotide-reductase', 'ribonucleoside-diphosphate reductase','anaerobic NTP reductase','oxidoreductase', 'kinase','dUTPase', 'ATPase','dCMP hydroxymethylase','amidase','phosphate starvation-inducible protein', 'deoxyuridine triphosphatase', 'dCTP pyrophosphatase', 'uracil DNA glycosylase' ]\n",
    "    entryInjection = ['movement protein','injection', 'fusion protein']\n",
    "    hyp=['hypothetical protein', 'putative uncharacterised protein']\n",
    "    ligase =['ligase']\n",
    "    structural=['structural protein']\n",
    "    recomb = ['recombination','recombinase']\n",
    "    translation = ['translation initiation', 'translation elongation', 'tRNA', 'translation factor']\n",
    "    nuclear=['nuclear shuttle protein', 'nuclear transporter']\n",
    "    poly=['polyprotein', 'readthrough protein', 'read-through protein']\n",
    "\n",
    "    capsidVec=[]\n",
    "    matrixVec=[]\n",
    "    envVec=[]\n",
    "    portalVec=[]        \n",
    "    tailVec=[]\n",
    "    polymeraseVec=[]\n",
    "    helicaseVec=[]\n",
    "    replicaseVec=[]\n",
    "    integraseVec=[]       \n",
    "    nucleaseVec=[]           \n",
    "    packingVec=[]\n",
    "    lysisVec=[]    \n",
    "    rmVec=[]\n",
    "    translationVec=[]        \n",
    "    metVec=[]      \n",
    "    entryVec=[]  \n",
    "    hypVec=[]\n",
    "    otherVec=[]\n",
    "    ligaseVec =[]\n",
    "    structuralVec=[]\n",
    "    recombVec=[]\n",
    "    nuclearVec=[]\n",
    "    polyVec=[]\n",
    "\n",
    "    for vec in protMat:\n",
    "        for i in vec:          \n",
    "            if any(substring in i for substring in capsid):\n",
    "                capsidVec.append(i)\n",
    "            elif any(substring in i for substring in matrix):\n",
    "                matrixVec.append(i)\n",
    "            elif any(substring in i for substring in env):\n",
    "                envVec.append(i)  \n",
    "            elif any(substring in i for substring in portal):\n",
    "                portalVec.append(i)        \n",
    "            elif any(substring in i for substring in tail):\n",
    "                tailVec.append(i)\n",
    "            elif any(substring in i for substring in polymerase):\n",
    "                polymeraseVec.append(i)\n",
    "            elif any(substring in i for substring in helicase):\n",
    "                helicaseVec.append(i)\n",
    "            elif any(substring in i for substring in replicase):\n",
    "                replicaseVec.append(i)\n",
    "            elif any(substring in i for substring in integrase):\n",
    "                integraseVec.append(i)       \n",
    "            elif any(substring in i for substring in nuclease):\n",
    "                nucleaseVec.append(i)           \n",
    "            elif any(substring in i for substring in packing):\n",
    "                packingVec.append(i)\n",
    "            elif any(substring in i for substring in lysis):\n",
    "                lysisVec.append(i)    \n",
    "            elif any(substring in i for substring in regulationModification):\n",
    "                rmVec.append(i)\n",
    "            elif any(substring in i for substring in translation):\n",
    "                translationVec.append(i)        \n",
    "            elif any(substring in i for substring in metabolismSynth):\n",
    "                metVec.append(i)      \n",
    "            elif any(substring in i for substring in entryInjection):\n",
    "                entryVec.append(i)            \n",
    "            elif any(substring in i for substring in ligase):\n",
    "                ligaseVec.append(i)\n",
    "            elif any(substring in i for substring in structural):\n",
    "                structuralVec.append(i)\n",
    "            elif any(substring in i for substring in recomb):\n",
    "                recombVec.append(i)\n",
    "            elif any(substring in i for substring in nuclear):\n",
    "                nuclearVec.append(i)\n",
    "            elif any(substring in i for substring in poly):\n",
    "                polyVec.append(i)\n",
    "            elif any(substring in i for substring in hyp):\n",
    "                hypVec.append(i)\n",
    "            else:       \n",
    "                otherVec.append(i)\n",
    "    protDict={'capsidd':capsidVec, 'matrixd': matrixVec, 'envd': envVec, 'portald': portalVec, 'taild': tailVec, 'structurald': structuralVec,'polymerased':polymeraseVec, \n",
    "              'replicased':replicaseVec, 'rmd':rmVec,'translationd':translationVec, 'helicased':helicaseVec,'integrased': integraseVec,'ligased':ligaseVec,'nucleased': nucleaseVec, 'recombd': recombVec,'metd':metVec,'entryd': entryVec,'nucleard': nuclearVec,'packingd': packingVec, 'lysisd':lysisVec, 'polyd': polyVec,'hypd': hypVec, 'otherd': otherVec}\n",
    "\n",
    "    return(protDict)\n",
    "\n",
    "#putting all dictionaries in a list\n",
    "protDictList=[]\n",
    "\n",
    "#indices of proteDictList = 0 RNA, 1 ssDNA, 2 dsDNA, 3 euk dsDNA, 4 bact dsDNA, 5 bact sipho, 6 bact myo, 7 bact podo, and then a bunch of eukaryotic dsDNA VIRUSES\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'dsRNA viruses' in  v['englishLineage'] or 'ssRNA positive-strand viruses, no DNA stage' in v['englishLineage']\n",
    "         or 'ssRNA negative-strand viruses' in v['englishLineage'] or 'Metaviridae' in v['englishLineage'] or 'Retroviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'ssDNA viruses' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'dsDNA viruses, no RNA stage' in v['englishLineage']] \n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] and 'dsDNA viruses, no RNA stage' in v['englishLineage']] \n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] and 'Siphoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] and 'Myoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] and 'Podoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'Baculoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'Poxviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'Herpesvirales' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'Papillomaviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'Adenoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'Polyomaviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "def getXYs(ind):     \n",
    "    #ind is a vector containing the list of indices (viral groups) we're interested in plotting together\n",
    "    #labels is a vector of strings that describe the viral groups being plotted\n",
    "    bar_width = 0.17\n",
    "    #m is the number of groups we're plotting \n",
    "    m = len(ind)\n",
    "    #the multiplier for bar_width so the bars can be placed next to each other\n",
    "    n= 0\n",
    "    #starting and updating the index of labels that we need to pull out. Corresponds to the labe of the viral group we're plotting\n",
    "    n2 =-1\n",
    "    #yvec will contain the number of proteins associated with each protein category, and all viral groups \n",
    "    yvec =[]\n",
    "    \n",
    "    for i in ind:\n",
    "    #i corresponds to the indices of the protein groups we want to plot from protDictList\n",
    "        n=n+1\n",
    "        n2 =n2 +1\n",
    "        #initializing the x and y coordinates for each bar plot\n",
    "        y=[]\n",
    "        x=[]\n",
    "    \n",
    "        for key in protDictList[i]: \n",
    "            #going through different protein categories (keys) to build the x vector labels\n",
    "            #the x values are just the number of categories, multiplied by n (which keeps getting updated) so the bar plots don't overlap\n",
    "            x.append(key)\n",
    "            #counting the number of items associated with each protein category (or key) to build the y vector\n",
    "            y.append(len(protDictList[i].get(key)))    \n",
    "        \n",
    "        xvec = list(range(len(y)))\n",
    "        xvalnew = [n + xval * m for xval in xvec] \n",
    "        yvec.append(y)\n",
    "\n",
    "\n",
    "        plt.bar(xvalnew, y, align='center')#,color=colvec[n2], label =labvec[n2])\n",
    "        plt.xticks(xvalnew, x)\n",
    "    plt.ylim(0,5000)\n",
    "    plt.show()\n",
    "    return(x, yvec)\n",
    "\n",
    "\n",
    "#Uncomment the next line to call the function to output x and y values for plotting\n",
    "#getXYs(list(range(14)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
