{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Gene order and gene functional category abundances\n",
    "###May 2017 | Gita Mahmoudabadi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4% done.\n",
      "4% done.\n",
      "7% done.\n",
      "11% done.\n",
      "15% done.\n",
      "19% done.\n",
      "22% done.\n",
      "26% done.\n",
      "30% done.\n",
      "33% done.\n",
      "33% done.\n",
      "37% done.\n",
      "41% done.\n",
      "44% done.\n",
      "48% done.\n",
      "52% done.\n",
      "56% done.\n",
      "59% done.\n",
      "59% done.\n",
      "63% done.\n",
      "67% done.\n",
      "67% done.\n",
      "70% done.\n",
      "70% done.\n",
      "74% done.\n",
      "78% done.\n",
      "78% done.\n",
      "81% done.\n",
      "81% done.\n",
      "81% done.\n",
      "81% done.\n",
      "81% done.\n",
      "81% done.\n",
      "81% done.\n",
      "81% done.\n",
      "81% done.\n",
      "81% done.\n",
      "81% done.\n",
      "81% done.\n",
      "81% done.\n",
      "81% done.\n",
      "85% done.\n",
      "89% done.\n",
      "89% done.\n",
      "0/2399\n",
      "1/2399\n",
      "2/2399\n",
      "3/2399\n",
      "4/2399\n",
      "5/2399\n",
      "6/2399\n",
      "7/2399\n",
      "8/2399\n",
      "9/2399\n",
      "10/2399\n",
      "11/2399\n",
      "12/2399\n",
      "13/2399\n",
      "14/2399\n",
      "15/2399\n",
      "16/2399\n",
      "17/2399\n",
      "18/2399\n",
      "19/2399\n",
      "20/2399\n",
      "21/2399\n",
      "22/2399\n",
      "23/2399\n",
      "24/2399\n",
      "25/2399\n",
      "26/2399\n",
      "27/2399\n",
      "28/2399\n",
      "29/2399\n",
      "30/2399\n",
      "31/2399\n",
      "32/2399\n",
      "33/2399\n",
      "34/2399\n",
      "35/2399\n",
      "36/2399\n",
      "37/2399\n",
      "38/2399\n",
      "39/2399\n",
      "40/2399\n",
      "41/2399\n",
      "42/2399\n",
      "43/2399\n",
      "44/2399\n",
      "45/2399\n",
      "46/2399\n",
      "47/2399\n",
      "48/2399\n",
      "49/2399\n",
      "50/2399\n",
      "51/2399\n",
      "52/2399\n",
      "53/2399\n",
      "54/2399\n",
      "55/2399\n",
      "56/2399\n",
      "57/2399\n",
      "58/2399\n",
      "59/2399\n",
      "60/2399\n",
      "61/2399\n",
      "62/2399\n",
      "63/2399\n",
      "64/2399\n",
      "65/2399\n",
      "66/2399\n",
      "67/2399\n",
      "68/2399\n",
      "69/2399\n",
      "70/2399\n",
      "71/2399\n",
      "72/2399\n",
      "73/2399\n",
      "74/2399\n",
      "75/2399\n",
      "76/2399\n",
      "77/2399\n",
      "78/2399\n",
      "79/2399\n",
      "80/2399\n",
      "81/2399\n",
      "82/2399\n",
      "83/2399\n",
      "84/2399\n",
      "85/2399\n",
      "86/2399\n",
      "87/2399\n",
      "88/2399\n",
      "89/2399\n",
      "90/2399\n",
      "91/2399\n",
      "92/2399\n",
      "93/2399\n",
      "94/2399\n",
      "95/2399\n",
      "96/2399\n",
      "97/2399\n",
      "98/2399\n",
      "99/2399\n",
      "100/2399\n",
      "101/2399\n",
      "102/2399\n",
      "103/2399\n",
      "104/2399\n",
      "105/2399\n",
      "106/2399\n",
      "107/2399\n",
      "108/2399\n",
      "109/2399\n",
      "110/2399\n",
      "111/2399\n",
      "112/2399\n",
      "113/2399\n",
      "114/2399\n",
      "115/2399\n",
      "116/2399\n",
      "117/2399\n",
      "118/2399\n",
      "119/2399\n",
      "120/2399\n",
      "121/2399\n",
      "122/2399\n",
      "123/2399\n",
      "124/2399\n",
      "125/2399\n",
      "126/2399\n",
      "127/2399\n",
      "128/2399\n",
      "129/2399\n",
      "130/2399\n",
      "131/2399\n",
      "132/2399\n",
      "133/2399\n",
      "134/2399\n",
      "135/2399\n",
      "136/2399\n",
      "137/2399\n",
      "138/2399\n",
      "139/2399\n",
      "140/2399\n",
      "141/2399\n",
      "142/2399\n",
      "143/2399\n",
      "144/2399\n",
      "145/2399\n",
      "146/2399\n",
      "147/2399\n",
      "148/2399\n",
      "149/2399\n",
      "150/2399\n",
      "151/2399\n",
      "152/2399\n",
      "153/2399\n",
      "154/2399\n",
      "155/2399\n",
      "156/2399\n",
      "157/2399\n",
      "158/2399\n",
      "159/2399\n",
      "160/2399\n",
      "161/2399\n",
      "162/2399\n",
      "163/2399\n",
      "164/2399\n",
      "165/2399\n",
      "166/2399\n",
      "167/2399\n",
      "168/2399\n",
      "169/2399\n",
      "170/2399\n",
      "171/2399\n",
      "172/2399\n",
      "173/2399\n",
      "174/2399\n",
      "175/2399\n",
      "176/2399\n",
      "177/2399\n",
      "178/2399\n",
      "179/2399\n",
      "180/2399\n",
      "181/2399\n",
      "182/2399\n",
      "183/2399\n",
      "184/2399\n",
      "185/2399\n",
      "186/2399\n",
      "187/2399\n",
      "188/2399\n",
      "189/2399\n",
      "190/2399\n",
      "191/2399\n",
      "192/2399\n",
      "193/2399\n",
      "194/2399\n",
      "195/2399\n",
      "196/2399\n",
      "197/2399\n",
      "198/2399\n",
      "199/2399\n",
      "200/2399\n",
      "201/2399\n",
      "202/2399\n",
      "203/2399\n",
      "204/2399\n",
      "205/2399\n",
      "206/2399\n",
      "207/2399\n",
      "208/2399\n",
      "209/2399\n",
      "210/2399\n",
      "211/2399\n",
      "212/2399\n",
      "213/2399\n",
      "214/2399\n",
      "215/2399\n",
      "216/2399\n",
      "217/2399\n",
      "218/2399\n",
      "219/2399\n",
      "220/2399\n",
      "221/2399\n",
      "222/2399\n",
      "223/2399\n",
      "224/2399\n",
      "225/2399\n",
      "226/2399\n",
      "227/2399\n",
      "228/2399\n",
      "229/2399\n",
      "230/2399\n",
      "231/2399\n",
      "232/2399\n",
      "233/2399\n",
      "234/2399\n",
      "235/2399\n",
      "236/2399\n",
      "237/2399\n",
      "238/2399\n",
      "239/2399\n",
      "240/2399\n",
      "241/2399\n",
      "242/2399\n",
      "243/2399\n",
      "244/2399\n",
      "245/2399\n",
      "246/2399\n",
      "247/2399\n",
      "248/2399\n",
      "249/2399\n",
      "250/2399\n",
      "251/2399\n",
      "252/2399\n",
      "253/2399\n",
      "254/2399\n",
      "255/2399\n",
      "256/2399\n",
      "257/2399\n",
      "258/2399\n",
      "259/2399\n",
      "260/2399\n",
      "261/2399\n",
      "262/2399\n",
      "263/2399\n",
      "264/2399\n",
      "265/2399\n",
      "266/2399\n",
      "267/2399\n",
      "268/2399\n",
      "269/2399\n",
      "270/2399\n",
      "271/2399\n",
      "272/2399\n",
      "273/2399\n",
      "274/2399\n",
      "275/2399\n",
      "276/2399\n",
      "277/2399\n",
      "278/2399\n",
      "279/2399\n",
      "280/2399\n",
      "281/2399\n",
      "282/2399\n",
      "283/2399\n",
      "284/2399\n",
      "285/2399\n",
      "286/2399\n",
      "287/2399\n",
      "288/2399\n",
      "289/2399\n",
      "290/2399\n",
      "291/2399\n",
      "292/2399\n",
      "293/2399\n",
      "294/2399\n",
      "295/2399\n",
      "296/2399\n",
      "297/2399\n",
      "298/2399\n",
      "299/2399\n",
      "300/2399\n",
      "301/2399\n",
      "302/2399\n",
      "303/2399\n",
      "304/2399\n",
      "305/2399\n",
      "306/2399\n",
      "307/2399\n",
      "308/2399\n",
      "309/2399\n",
      "310/2399\n",
      "311/2399\n",
      "312/2399\n",
      "313/2399\n",
      "314/2399\n",
      "315/2399\n",
      "316/2399\n",
      "317/2399\n",
      "318/2399\n",
      "319/2399\n",
      "320/2399\n",
      "321/2399\n",
      "322/2399\n",
      "323/2399\n",
      "324/2399\n",
      "325/2399\n",
      "326/2399\n",
      "327/2399\n",
      "328/2399\n",
      "329/2399\n",
      "330/2399\n",
      "331/2399\n",
      "332/2399\n",
      "333/2399\n",
      "334/2399\n",
      "335/2399\n",
      "336/2399\n",
      "337/2399\n",
      "338/2399\n",
      "339/2399\n",
      "340/2399\n",
      "341/2399\n",
      "342/2399\n",
      "343/2399\n",
      "344/2399\n",
      "345/2399\n",
      "346/2399\n",
      "347/2399\n",
      "348/2399\n",
      "349/2399\n",
      "350/2399\n",
      "351/2399\n",
      "352/2399\n",
      "353/2399\n",
      "354/2399\n",
      "355/2399\n",
      "356/2399\n",
      "357/2399\n",
      "358/2399\n",
      "359/2399\n",
      "360/2399\n",
      "361/2399\n",
      "362/2399\n",
      "363/2399\n",
      "364/2399\n",
      "365/2399\n",
      "366/2399\n",
      "367/2399\n",
      "368/2399\n",
      "369/2399\n",
      "370/2399\n",
      "371/2399\n",
      "372/2399\n",
      "373/2399\n",
      "374/2399\n",
      "375/2399\n",
      "376/2399\n",
      "377/2399\n",
      "378/2399\n",
      "379/2399\n",
      "380/2399\n",
      "381/2399\n",
      "382/2399\n",
      "383/2399\n",
      "384/2399\n",
      "385/2399\n",
      "386/2399\n",
      "387/2399\n",
      "388/2399\n",
      "389/2399\n",
      "390/2399\n",
      "391/2399\n",
      "392/2399\n",
      "393/2399\n",
      "394/2399\n",
      "395/2399\n",
      "396/2399\n",
      "397/2399\n",
      "398/2399\n",
      "399/2399\n",
      "400/2399\n",
      "401/2399\n",
      "402/2399\n",
      "403/2399\n",
      "404/2399\n",
      "405/2399\n",
      "406/2399\n",
      "407/2399\n",
      "408/2399\n",
      "409/2399\n",
      "410/2399\n",
      "411/2399\n",
      "412/2399\n",
      "413/2399\n",
      "414/2399\n",
      "415/2399\n",
      "416/2399\n",
      "417/2399\n",
      "418/2399\n",
      "419/2399\n",
      "420/2399\n",
      "421/2399\n",
      "422/2399\n",
      "423/2399\n",
      "424/2399\n",
      "425/2399\n",
      "426/2399\n",
      "427/2399\n",
      "428/2399\n",
      "429/2399\n",
      "430/2399\n",
      "431/2399\n",
      "432/2399\n",
      "433/2399\n",
      "434/2399\n",
      "435/2399\n",
      "436/2399\n",
      "437/2399\n",
      "438/2399\n",
      "439/2399\n",
      "440/2399\n",
      "441/2399\n",
      "442/2399\n",
      "443/2399\n",
      "444/2399\n",
      "445/2399\n",
      "446/2399\n",
      "447/2399\n",
      "448/2399\n",
      "449/2399\n",
      "450/2399\n",
      "451/2399\n",
      "452/2399\n",
      "453/2399\n",
      "454/2399\n",
      "455/2399\n",
      "456/2399\n",
      "457/2399\n",
      "458/2399\n",
      "459/2399\n",
      "460/2399\n",
      "461/2399\n",
      "462/2399\n",
      "463/2399\n",
      "464/2399\n",
      "465/2399\n",
      "466/2399\n",
      "467/2399\n",
      "468/2399\n",
      "469/2399\n",
      "470/2399\n",
      "471/2399\n",
      "472/2399\n",
      "473/2399\n",
      "474/2399\n",
      "475/2399\n",
      "476/2399\n",
      "477/2399\n",
      "478/2399\n",
      "479/2399\n",
      "480/2399\n",
      "481/2399\n",
      "482/2399\n",
      "483/2399\n",
      "484/2399\n",
      "485/2399\n",
      "486/2399\n",
      "487/2399\n",
      "488/2399\n",
      "489/2399\n",
      "490/2399\n",
      "491/2399\n",
      "492/2399\n",
      "493/2399\n",
      "494/2399\n",
      "495/2399\n",
      "496/2399\n",
      "497/2399\n",
      "498/2399\n",
      "499/2399\n",
      "500/2399\n",
      "501/2399\n",
      "502/2399\n",
      "503/2399\n",
      "504/2399\n",
      "505/2399\n",
      "506/2399\n",
      "507/2399\n",
      "508/2399\n",
      "509/2399\n",
      "510/2399\n",
      "511/2399\n",
      "512/2399\n",
      "513/2399\n",
      "514/2399\n",
      "515/2399\n",
      "516/2399\n",
      "517/2399\n",
      "518/2399\n",
      "519/2399\n",
      "520/2399\n",
      "521/2399\n",
      "522/2399\n",
      "523/2399\n",
      "524/2399\n",
      "525/2399\n",
      "526/2399\n",
      "527/2399\n",
      "528/2399\n",
      "529/2399\n",
      "530/2399\n",
      "531/2399\n",
      "532/2399\n",
      "533/2399\n",
      "534/2399\n",
      "535/2399\n",
      "536/2399\n",
      "537/2399\n",
      "538/2399\n",
      "539/2399\n",
      "540/2399\n",
      "541/2399\n",
      "542/2399\n",
      "543/2399\n",
      "544/2399\n",
      "545/2399\n",
      "546/2399\n",
      "547/2399\n",
      "548/2399\n",
      "549/2399\n",
      "550/2399\n",
      "551/2399\n",
      "552/2399\n",
      "553/2399\n",
      "554/2399\n",
      "555/2399\n",
      "556/2399\n",
      "557/2399\n",
      "558/2399\n",
      "559/2399\n",
      "560/2399\n",
      "561/2399\n",
      "562/2399\n",
      "563/2399\n",
      "564/2399\n",
      "565/2399\n",
      "566/2399\n",
      "567/2399\n",
      "568/2399\n",
      "569/2399\n",
      "570/2399\n",
      "571/2399\n",
      "572/2399\n",
      "573/2399\n",
      "574/2399\n",
      "575/2399\n",
      "576/2399\n",
      "577/2399\n",
      "578/2399\n",
      "579/2399\n",
      "580/2399\n",
      "581/2399\n",
      "582/2399\n",
      "583/2399\n",
      "584/2399\n",
      "585/2399\n",
      "586/2399\n",
      "587/2399\n",
      "588/2399\n",
      "589/2399\n",
      "590/2399\n",
      "591/2399\n",
      "592/2399\n",
      "593/2399\n",
      "594/2399\n",
      "595/2399\n",
      "596/2399\n",
      "597/2399\n",
      "598/2399\n",
      "599/2399\n",
      "600/2399\n",
      "601/2399\n",
      "602/2399\n",
      "603/2399\n",
      "604/2399\n",
      "605/2399\n",
      "606/2399\n",
      "607/2399\n",
      "608/2399\n",
      "609/2399\n",
      "610/2399\n",
      "611/2399\n",
      "612/2399\n",
      "613/2399\n",
      "614/2399\n",
      "615/2399\n",
      "616/2399\n",
      "617/2399\n",
      "618/2399\n",
      "619/2399\n",
      "620/2399\n",
      "621/2399\n",
      "622/2399\n",
      "623/2399\n",
      "624/2399\n",
      "625/2399\n",
      "626/2399\n",
      "627/2399\n",
      "628/2399\n",
      "629/2399\n",
      "630/2399\n",
      "631/2399\n",
      "632/2399\n",
      "633/2399\n",
      "634/2399\n",
      "635/2399\n",
      "636/2399\n",
      "637/2399\n",
      "638/2399\n",
      "639/2399\n",
      "640/2399\n",
      "641/2399\n",
      "642/2399\n",
      "643/2399\n",
      "644/2399\n",
      "645/2399\n",
      "646/2399\n",
      "647/2399\n",
      "648/2399\n",
      "649/2399\n",
      "650/2399\n",
      "651/2399\n",
      "652/2399\n",
      "653/2399\n",
      "654/2399\n",
      "655/2399\n",
      "656/2399\n",
      "657/2399\n",
      "658/2399\n",
      "659/2399\n",
      "660/2399\n",
      "661/2399\n",
      "662/2399\n",
      "663/2399\n",
      "664/2399\n",
      "665/2399\n",
      "666/2399\n",
      "667/2399\n",
      "668/2399\n",
      "669/2399\n",
      "670/2399\n",
      "671/2399\n",
      "672/2399\n",
      "673/2399\n",
      "674/2399\n",
      "675/2399\n",
      "676/2399\n",
      "677/2399\n",
      "678/2399\n",
      "679/2399\n",
      "680/2399\n",
      "681/2399\n",
      "682/2399\n",
      "683/2399\n",
      "684/2399\n",
      "685/2399\n",
      "686/2399\n",
      "687/2399\n",
      "688/2399\n",
      "689/2399\n",
      "690/2399\n",
      "691/2399\n",
      "692/2399\n",
      "693/2399\n",
      "694/2399\n",
      "695/2399\n",
      "696/2399\n",
      "697/2399\n",
      "698/2399\n",
      "699/2399\n",
      "700/2399\n",
      "701/2399\n",
      "702/2399\n",
      "703/2399\n",
      "704/2399\n",
      "705/2399\n",
      "706/2399\n",
      "707/2399\n",
      "708/2399\n",
      "709/2399\n",
      "710/2399\n",
      "711/2399\n",
      "712/2399\n",
      "713/2399\n",
      "714/2399\n",
      "715/2399\n",
      "716/2399\n",
      "717/2399\n",
      "718/2399\n",
      "719/2399\n",
      "720/2399\n",
      "721/2399\n",
      "722/2399\n",
      "723/2399\n",
      "724/2399\n",
      "725/2399\n",
      "726/2399\n",
      "727/2399\n",
      "728/2399\n",
      "729/2399\n",
      "730/2399\n",
      "731/2399\n",
      "732/2399\n",
      "733/2399\n",
      "734/2399\n",
      "735/2399\n",
      "736/2399\n",
      "737/2399\n",
      "738/2399\n",
      "739/2399\n",
      "740/2399\n",
      "741/2399\n",
      "742/2399\n",
      "743/2399\n",
      "744/2399\n",
      "745/2399\n",
      "746/2399\n",
      "747/2399\n",
      "748/2399\n",
      "749/2399\n",
      "750/2399\n",
      "751/2399\n",
      "752/2399\n",
      "753/2399\n",
      "754/2399\n",
      "755/2399\n",
      "756/2399\n",
      "757/2399\n",
      "758/2399\n",
      "759/2399\n",
      "760/2399\n",
      "761/2399\n",
      "762/2399\n",
      "763/2399\n",
      "764/2399\n",
      "765/2399\n",
      "766/2399\n",
      "767/2399\n",
      "768/2399\n",
      "769/2399\n",
      "770/2399\n",
      "771/2399\n",
      "772/2399\n",
      "773/2399\n",
      "774/2399\n",
      "775/2399\n",
      "776/2399\n",
      "777/2399\n",
      "778/2399\n",
      "779/2399\n",
      "780/2399\n",
      "781/2399\n",
      "782/2399\n",
      "783/2399\n",
      "784/2399\n",
      "785/2399\n",
      "786/2399\n",
      "787/2399\n",
      "788/2399\n",
      "789/2399\n",
      "790/2399\n",
      "791/2399\n",
      "792/2399\n",
      "793/2399\n",
      "794/2399\n",
      "795/2399\n",
      "796/2399\n",
      "797/2399\n",
      "798/2399\n",
      "799/2399\n",
      "800/2399\n",
      "801/2399\n",
      "802/2399\n",
      "803/2399\n",
      "804/2399\n",
      "805/2399\n",
      "806/2399\n",
      "807/2399\n",
      "808/2399\n",
      "809/2399\n",
      "810/2399\n",
      "811/2399\n",
      "812/2399\n",
      "813/2399\n",
      "814/2399\n",
      "815/2399\n",
      "816/2399\n",
      "817/2399\n",
      "818/2399\n",
      "819/2399\n",
      "820/2399\n",
      "821/2399\n",
      "822/2399\n",
      "823/2399\n",
      "824/2399\n",
      "825/2399\n",
      "826/2399\n",
      "827/2399\n",
      "828/2399\n",
      "829/2399\n",
      "830/2399\n",
      "831/2399\n",
      "832/2399\n",
      "833/2399\n",
      "834/2399\n",
      "835/2399\n",
      "836/2399\n",
      "837/2399\n",
      "838/2399\n",
      "839/2399\n",
      "840/2399\n",
      "841/2399\n",
      "842/2399\n",
      "843/2399\n",
      "844/2399\n",
      "845/2399\n",
      "846/2399\n",
      "847/2399\n",
      "848/2399\n",
      "849/2399\n",
      "850/2399\n",
      "851/2399\n",
      "852/2399\n",
      "853/2399\n",
      "854/2399\n",
      "855/2399\n",
      "856/2399\n",
      "857/2399\n",
      "858/2399\n",
      "859/2399\n",
      "860/2399\n",
      "861/2399\n",
      "862/2399\n",
      "863/2399\n",
      "864/2399\n",
      "865/2399\n",
      "866/2399\n",
      "867/2399\n",
      "868/2399\n",
      "869/2399\n",
      "870/2399\n",
      "871/2399\n",
      "872/2399\n",
      "873/2399\n",
      "874/2399\n",
      "875/2399\n",
      "876/2399\n",
      "877/2399\n",
      "878/2399\n",
      "879/2399\n",
      "880/2399\n",
      "881/2399\n",
      "882/2399\n",
      "883/2399\n",
      "884/2399\n",
      "885/2399\n",
      "886/2399\n",
      "887/2399\n",
      "888/2399\n",
      "889/2399\n",
      "890/2399\n",
      "891/2399\n",
      "892/2399\n",
      "893/2399\n",
      "894/2399\n",
      "895/2399\n",
      "896/2399\n",
      "897/2399\n",
      "898/2399\n",
      "899/2399\n",
      "900/2399\n",
      "901/2399\n",
      "902/2399\n",
      "903/2399\n",
      "904/2399\n",
      "905/2399\n",
      "906/2399\n",
      "907/2399\n",
      "908/2399\n",
      "909/2399\n",
      "910/2399\n",
      "911/2399\n",
      "912/2399\n",
      "913/2399\n",
      "914/2399\n",
      "915/2399\n",
      "916/2399\n",
      "917/2399\n",
      "918/2399\n",
      "919/2399\n",
      "920/2399\n",
      "921/2399\n",
      "922/2399\n",
      "923/2399\n",
      "924/2399\n",
      "925/2399\n",
      "926/2399\n",
      "927/2399\n",
      "928/2399\n",
      "929/2399\n",
      "930/2399\n",
      "931/2399\n",
      "932/2399\n",
      "933/2399\n",
      "934/2399\n",
      "935/2399\n",
      "936/2399\n",
      "937/2399\n",
      "938/2399\n",
      "939/2399\n",
      "940/2399\n",
      "941/2399\n",
      "942/2399\n",
      "943/2399\n",
      "944/2399\n",
      "945/2399\n",
      "946/2399\n",
      "947/2399\n",
      "948/2399\n",
      "949/2399\n",
      "950/2399\n",
      "951/2399\n",
      "952/2399\n",
      "953/2399\n",
      "954/2399\n",
      "955/2399\n",
      "956/2399\n",
      "957/2399\n",
      "958/2399\n",
      "959/2399\n",
      "960/2399\n",
      "961/2399\n",
      "962/2399\n",
      "963/2399\n",
      "964/2399\n",
      "965/2399\n",
      "966/2399\n",
      "967/2399\n",
      "968/2399\n",
      "969/2399\n",
      "970/2399\n",
      "971/2399\n",
      "972/2399\n",
      "973/2399\n",
      "974/2399\n",
      "975/2399\n",
      "976/2399\n",
      "977/2399\n",
      "978/2399\n",
      "979/2399\n",
      "980/2399\n",
      "981/2399\n",
      "982/2399\n",
      "983/2399\n",
      "984/2399\n",
      "985/2399\n",
      "986/2399\n",
      "987/2399\n",
      "988/2399\n",
      "989/2399\n",
      "990/2399\n",
      "991/2399\n",
      "992/2399\n",
      "993/2399\n",
      "994/2399\n",
      "995/2399\n",
      "996/2399\n",
      "997/2399\n",
      "998/2399\n",
      "999/2399\n",
      "1000/2399\n",
      "1001/2399\n",
      "1002/2399\n",
      "1003/2399\n",
      "1004/2399\n",
      "1005/2399\n",
      "1006/2399\n",
      "1007/2399\n",
      "1008/2399\n",
      "1009/2399\n",
      "1010/2399\n",
      "1011/2399\n",
      "1012/2399\n",
      "1013/2399\n",
      "1014/2399\n",
      "1015/2399\n",
      "1016/2399\n",
      "1017/2399\n",
      "1018/2399\n",
      "1019/2399\n",
      "1020/2399\n",
      "1021/2399\n",
      "1022/2399\n",
      "1023/2399\n",
      "1024/2399\n",
      "1025/2399\n",
      "1026/2399\n",
      "1027/2399\n",
      "1028/2399\n",
      "1029/2399\n",
      "1030/2399\n",
      "1031/2399\n",
      "1032/2399\n",
      "1033/2399\n",
      "1034/2399\n",
      "1035/2399\n",
      "1036/2399\n",
      "1037/2399\n",
      "1038/2399\n",
      "1039/2399\n",
      "1040/2399\n",
      "1041/2399\n",
      "1042/2399\n",
      "1043/2399\n",
      "1044/2399\n",
      "1045/2399\n",
      "1046/2399\n",
      "1047/2399\n",
      "1048/2399\n",
      "1049/2399\n",
      "1050/2399\n",
      "1051/2399\n",
      "1052/2399\n",
      "1053/2399\n",
      "1054/2399\n",
      "1055/2399\n",
      "1056/2399\n",
      "1057/2399\n",
      "1058/2399\n",
      "1059/2399\n",
      "1060/2399\n",
      "1061/2399\n",
      "1062/2399\n",
      "1063/2399\n",
      "1064/2399\n",
      "1065/2399\n",
      "1066/2399\n",
      "1067/2399\n",
      "1068/2399\n",
      "1069/2399\n",
      "1070/2399\n",
      "1071/2399\n",
      "1072/2399\n",
      "1073/2399\n",
      "1074/2399\n",
      "1075/2399\n",
      "1076/2399\n",
      "1077/2399\n",
      "1078/2399\n",
      "1079/2399\n",
      "1080/2399\n",
      "1081/2399\n",
      "1082/2399\n",
      "1083/2399\n",
      "1084/2399\n",
      "1085/2399\n",
      "1086/2399\n",
      "1087/2399\n",
      "1088/2399\n",
      "1089/2399\n",
      "1090/2399\n",
      "1091/2399\n",
      "1092/2399\n",
      "1093/2399\n",
      "1094/2399\n",
      "1095/2399\n",
      "1096/2399\n",
      "1097/2399\n",
      "1098/2399\n",
      "1099/2399\n",
      "1100/2399\n",
      "1101/2399\n",
      "1102/2399\n",
      "1103/2399\n",
      "1104/2399\n",
      "1105/2399\n",
      "1106/2399\n",
      "1107/2399\n",
      "1108/2399\n",
      "1109/2399\n",
      "1110/2399\n",
      "1111/2399\n",
      "1112/2399\n",
      "1113/2399\n",
      "1114/2399\n",
      "1115/2399\n",
      "1116/2399\n",
      "1117/2399\n",
      "1118/2399\n",
      "1119/2399\n",
      "1120/2399\n",
      "1121/2399\n",
      "1122/2399\n",
      "1123/2399\n",
      "1124/2399\n",
      "1125/2399\n",
      "1126/2399\n",
      "1127/2399\n",
      "1128/2399\n",
      "1129/2399\n",
      "1130/2399\n",
      "1131/2399\n",
      "1132/2399\n",
      "1133/2399\n",
      "1134/2399\n",
      "1135/2399\n",
      "1136/2399\n",
      "1137/2399\n",
      "1138/2399\n",
      "1139/2399\n",
      "1140/2399\n",
      "1141/2399\n",
      "1142/2399\n",
      "1143/2399\n",
      "1144/2399\n",
      "1145/2399\n",
      "1146/2399\n",
      "1147/2399\n",
      "1148/2399\n",
      "1149/2399\n",
      "1150/2399\n",
      "1151/2399\n",
      "1152/2399\n",
      "1153/2399\n",
      "1154/2399\n",
      "1155/2399\n",
      "1156/2399\n",
      "1157/2399\n",
      "1158/2399\n",
      "1159/2399\n",
      "1160/2399\n",
      "1161/2399\n",
      "1162/2399\n",
      "1163/2399\n",
      "1164/2399\n",
      "1165/2399\n",
      "1166/2399\n",
      "1167/2399\n",
      "1168/2399\n",
      "1169/2399\n",
      "1170/2399\n",
      "1171/2399\n",
      "1172/2399\n",
      "1173/2399\n",
      "1174/2399\n",
      "1175/2399\n",
      "1176/2399\n",
      "1177/2399\n",
      "1178/2399\n",
      "1179/2399\n",
      "1180/2399\n",
      "1181/2399\n",
      "1182/2399\n",
      "1183/2399\n",
      "1184/2399\n",
      "1185/2399\n",
      "1186/2399\n",
      "1187/2399\n",
      "1188/2399\n",
      "1189/2399\n",
      "1190/2399\n",
      "1191/2399\n",
      "1192/2399\n",
      "1193/2399\n",
      "1194/2399\n",
      "1195/2399\n",
      "1196/2399\n",
      "1197/2399\n",
      "1198/2399\n",
      "1199/2399\n",
      "1200/2399\n",
      "1201/2399\n",
      "1202/2399\n",
      "1203/2399\n",
      "1204/2399\n",
      "1205/2399\n",
      "1206/2399\n",
      "1207/2399\n",
      "1208/2399\n",
      "1209/2399\n",
      "1210/2399\n",
      "1211/2399\n",
      "1212/2399\n",
      "1213/2399\n",
      "1214/2399\n",
      "1215/2399\n",
      "1216/2399\n",
      "1217/2399\n",
      "1218/2399\n",
      "1219/2399\n",
      "1220/2399\n",
      "1221/2399\n",
      "1222/2399\n",
      "1223/2399\n",
      "1224/2399\n",
      "1225/2399\n",
      "1226/2399\n",
      "1227/2399\n",
      "1228/2399\n",
      "1229/2399\n",
      "1230/2399\n",
      "1231/2399\n",
      "1232/2399\n",
      "1233/2399\n",
      "1234/2399\n",
      "1235/2399\n",
      "1236/2399\n",
      "1237/2399\n",
      "1238/2399\n",
      "1239/2399\n",
      "1240/2399\n",
      "1241/2399\n",
      "1242/2399\n",
      "1243/2399\n",
      "1244/2399\n",
      "1245/2399\n",
      "1246/2399\n",
      "1247/2399\n",
      "1248/2399\n",
      "1249/2399\n",
      "1250/2399\n",
      "1251/2399\n",
      "1252/2399\n",
      "1253/2399\n",
      "1254/2399\n",
      "1255/2399\n",
      "1256/2399\n",
      "1257/2399\n",
      "1258/2399\n",
      "1259/2399\n",
      "1260/2399\n",
      "1261/2399\n",
      "1262/2399\n",
      "1263/2399\n",
      "1264/2399\n",
      "1265/2399\n",
      "1266/2399\n",
      "1267/2399\n",
      "1268/2399\n",
      "1269/2399\n",
      "1270/2399\n",
      "1271/2399\n",
      "1272/2399\n",
      "1273/2399\n",
      "1274/2399\n",
      "1275/2399\n",
      "1276/2399\n",
      "1277/2399\n",
      "1278/2399\n",
      "1279/2399\n",
      "1280/2399\n",
      "1281/2399\n",
      "1282/2399\n",
      "1283/2399\n",
      "1284/2399\n",
      "1285/2399\n",
      "1286/2399\n",
      "1287/2399\n",
      "1288/2399\n",
      "1289/2399\n",
      "1290/2399\n",
      "1291/2399\n",
      "1292/2399\n",
      "1293/2399\n",
      "1294/2399\n",
      "1295/2399\n",
      "1296/2399\n",
      "1297/2399\n",
      "1298/2399\n",
      "1299/2399\n",
      "1300/2399\n",
      "1301/2399\n",
      "1302/2399\n",
      "1303/2399\n",
      "1304/2399\n",
      "1305/2399\n",
      "1306/2399\n",
      "1307/2399\n",
      "1308/2399\n",
      "1309/2399\n",
      "1310/2399\n",
      "1311/2399\n",
      "1312/2399\n",
      "1313/2399\n",
      "1314/2399\n",
      "1315/2399\n",
      "1316/2399\n",
      "1317/2399\n",
      "1318/2399\n",
      "1319/2399\n",
      "1320/2399\n",
      "1321/2399\n",
      "1322/2399\n",
      "1323/2399\n",
      "1324/2399\n",
      "1325/2399\n",
      "1326/2399\n",
      "1327/2399\n",
      "1328/2399\n",
      "1329/2399\n",
      "1330/2399\n",
      "1331/2399\n",
      "1332/2399\n",
      "1333/2399\n",
      "1334/2399\n",
      "1335/2399\n",
      "1336/2399\n",
      "1337/2399\n",
      "1338/2399\n",
      "1339/2399\n",
      "1340/2399\n",
      "1341/2399\n",
      "1342/2399\n",
      "1343/2399\n",
      "1344/2399\n",
      "1345/2399\n",
      "1346/2399\n",
      "1347/2399\n",
      "1348/2399\n",
      "1349/2399\n",
      "1350/2399\n",
      "1351/2399\n",
      "1352/2399\n",
      "1353/2399\n",
      "1354/2399\n",
      "1355/2399\n",
      "1356/2399\n",
      "1357/2399\n",
      "1358/2399\n",
      "1359/2399\n",
      "1360/2399\n",
      "1361/2399\n",
      "1362/2399\n",
      "1363/2399\n",
      "1364/2399\n",
      "1365/2399\n",
      "1366/2399\n",
      "1367/2399\n",
      "1368/2399\n",
      "1369/2399\n",
      "1370/2399\n",
      "1371/2399\n",
      "1372/2399\n",
      "1373/2399\n",
      "1374/2399\n",
      "1375/2399\n",
      "1376/2399\n",
      "1377/2399\n",
      "1378/2399\n",
      "1379/2399\n",
      "1380/2399\n",
      "1381/2399\n",
      "1382/2399\n",
      "1383/2399\n",
      "1384/2399\n",
      "1385/2399\n",
      "1386/2399\n",
      "1387/2399\n",
      "1388/2399\n",
      "1389/2399\n",
      "1390/2399\n",
      "1391/2399\n",
      "1392/2399\n",
      "1393/2399\n",
      "1394/2399\n",
      "1395/2399\n",
      "1396/2399\n",
      "1397/2399\n",
      "1398/2399\n",
      "1399/2399\n",
      "1400/2399\n",
      "1401/2399\n",
      "1402/2399\n",
      "1403/2399\n",
      "1404/2399\n",
      "1405/2399\n",
      "1406/2399\n",
      "1407/2399\n",
      "1408/2399\n",
      "1409/2399\n",
      "1410/2399\n",
      "1411/2399\n",
      "1412/2399\n",
      "1413/2399\n",
      "1414/2399\n",
      "1415/2399\n",
      "1416/2399\n",
      "1417/2399\n",
      "1418/2399\n",
      "1419/2399\n",
      "1420/2399\n",
      "1421/2399\n",
      "1422/2399\n",
      "1423/2399\n",
      "1424/2399\n",
      "1425/2399\n",
      "1426/2399\n",
      "1427/2399\n",
      "1428/2399\n",
      "1429/2399\n",
      "1430/2399\n",
      "1431/2399\n",
      "1432/2399\n",
      "1433/2399\n",
      "1434/2399\n",
      "1435/2399\n",
      "1436/2399\n",
      "1437/2399\n",
      "1438/2399\n",
      "1439/2399\n",
      "1440/2399\n",
      "1441/2399\n",
      "1442/2399\n",
      "1443/2399\n",
      "1444/2399\n",
      "1445/2399\n",
      "1446/2399\n",
      "1447/2399\n",
      "1448/2399\n",
      "1449/2399\n",
      "1450/2399\n",
      "1451/2399\n",
      "1452/2399\n",
      "1453/2399\n",
      "1454/2399\n",
      "1455/2399\n",
      "1456/2399\n",
      "1457/2399\n",
      "1458/2399\n",
      "1459/2399\n",
      "1460/2399\n",
      "1461/2399\n",
      "1462/2399\n",
      "1463/2399\n",
      "1464/2399\n",
      "1465/2399\n",
      "1466/2399\n",
      "1467/2399\n",
      "1468/2399\n",
      "1469/2399\n",
      "1470/2399\n",
      "1471/2399\n",
      "1472/2399\n",
      "1473/2399\n",
      "1474/2399\n",
      "1475/2399\n",
      "1476/2399\n",
      "1477/2399\n",
      "1478/2399\n",
      "1479/2399\n",
      "1480/2399\n",
      "1481/2399\n",
      "1482/2399\n",
      "1483/2399\n",
      "1484/2399\n",
      "1485/2399\n",
      "1486/2399\n",
      "1487/2399\n",
      "1488/2399\n",
      "1489/2399\n",
      "1490/2399\n",
      "1491/2399\n",
      "1492/2399\n",
      "1493/2399\n",
      "1494/2399\n",
      "1495/2399\n",
      "1496/2399\n",
      "1497/2399\n",
      "1498/2399\n",
      "1499/2399\n",
      "1500/2399\n",
      "1501/2399\n",
      "1502/2399\n",
      "1503/2399\n",
      "1504/2399\n",
      "1505/2399\n",
      "1506/2399\n",
      "1507/2399\n",
      "1508/2399\n",
      "1509/2399\n",
      "1510/2399\n",
      "1511/2399\n",
      "1512/2399\n",
      "1513/2399\n",
      "1514/2399\n",
      "1515/2399\n",
      "1516/2399\n",
      "1517/2399\n",
      "1518/2399\n",
      "1519/2399\n",
      "1520/2399\n",
      "1521/2399\n",
      "1522/2399\n",
      "1523/2399\n",
      "1524/2399\n",
      "1525/2399\n",
      "1526/2399\n",
      "1527/2399\n",
      "1528/2399\n",
      "1529/2399\n",
      "1530/2399\n",
      "1531/2399\n",
      "1532/2399\n",
      "1533/2399\n",
      "1534/2399\n",
      "1535/2399\n",
      "1536/2399\n",
      "1537/2399\n",
      "1538/2399\n",
      "1539/2399\n",
      "1540/2399\n",
      "1541/2399\n",
      "1542/2399\n",
      "1543/2399\n",
      "1544/2399\n",
      "1545/2399\n",
      "1546/2399\n",
      "1547/2399\n",
      "1548/2399\n",
      "1549/2399\n",
      "1550/2399\n",
      "1551/2399\n",
      "1552/2399\n",
      "1553/2399\n",
      "1554/2399\n",
      "1555/2399\n",
      "1556/2399\n",
      "1557/2399\n",
      "1558/2399\n",
      "1559/2399\n",
      "1560/2399\n",
      "1561/2399\n",
      "1562/2399\n",
      "1563/2399\n",
      "1564/2399\n",
      "1565/2399\n",
      "1566/2399\n",
      "1567/2399\n",
      "1568/2399\n",
      "1569/2399\n",
      "1570/2399\n",
      "1571/2399\n",
      "1572/2399\n",
      "1573/2399\n",
      "1574/2399\n",
      "1575/2399\n",
      "1576/2399\n",
      "1577/2399\n",
      "1578/2399\n",
      "1579/2399\n",
      "1580/2399\n",
      "1581/2399\n",
      "1582/2399\n",
      "1583/2399\n",
      "1584/2399\n",
      "1585/2399\n",
      "1586/2399\n",
      "1587/2399\n",
      "1588/2399\n",
      "1589/2399\n",
      "1590/2399\n",
      "1591/2399\n",
      "1592/2399\n",
      "1593/2399\n",
      "1594/2399\n",
      "1595/2399\n",
      "1596/2399\n",
      "1597/2399\n",
      "1598/2399\n",
      "1599/2399\n",
      "1600/2399\n",
      "1601/2399\n",
      "1602/2399\n",
      "1603/2399\n",
      "1604/2399\n",
      "1605/2399\n",
      "1606/2399\n",
      "1607/2399\n",
      "1608/2399\n",
      "1609/2399\n",
      "1610/2399\n",
      "1611/2399\n",
      "1612/2399\n",
      "1613/2399\n",
      "1614/2399\n",
      "1615/2399\n",
      "1616/2399\n",
      "1617/2399\n",
      "1618/2399\n",
      "1619/2399\n",
      "1620/2399\n",
      "1621/2399\n",
      "1622/2399\n",
      "1623/2399\n",
      "1624/2399\n",
      "1625/2399\n",
      "1626/2399\n",
      "1627/2399\n",
      "1628/2399\n",
      "1629/2399\n",
      "1630/2399\n",
      "1631/2399\n",
      "1632/2399\n",
      "1633/2399\n",
      "1634/2399\n",
      "1635/2399\n",
      "1636/2399\n",
      "1637/2399\n",
      "1638/2399\n",
      "1639/2399\n",
      "1640/2399\n",
      "1641/2399\n",
      "1642/2399\n",
      "1643/2399\n",
      "1644/2399\n",
      "1645/2399\n",
      "1646/2399\n",
      "1647/2399\n",
      "1648/2399\n",
      "1649/2399\n",
      "1650/2399\n",
      "1651/2399\n",
      "1652/2399\n",
      "1653/2399\n",
      "1654/2399\n",
      "1655/2399\n",
      "1656/2399\n",
      "1657/2399\n",
      "1658/2399\n",
      "1659/2399\n",
      "1660/2399\n",
      "1661/2399\n",
      "1662/2399\n",
      "1663/2399\n",
      "1664/2399\n",
      "1665/2399\n",
      "1666/2399\n",
      "1667/2399\n",
      "1668/2399\n",
      "1669/2399\n",
      "1670/2399\n",
      "1671/2399\n",
      "1672/2399\n",
      "1673/2399\n",
      "1674/2399\n",
      "1675/2399\n",
      "1676/2399\n",
      "1677/2399\n",
      "1678/2399\n",
      "1679/2399\n",
      "1680/2399\n",
      "1681/2399\n",
      "1682/2399\n",
      "1683/2399\n",
      "1684/2399\n",
      "1685/2399\n",
      "1686/2399\n",
      "1687/2399\n",
      "1688/2399\n",
      "1689/2399\n",
      "1690/2399\n",
      "1691/2399\n",
      "1692/2399\n",
      "1693/2399\n",
      "1694/2399\n",
      "1695/2399\n",
      "1696/2399\n",
      "1697/2399\n",
      "1698/2399\n",
      "1699/2399\n",
      "1700/2399\n",
      "1701/2399\n",
      "1702/2399\n",
      "1703/2399\n",
      "1704/2399\n",
      "1705/2399\n",
      "1706/2399\n",
      "1707/2399\n",
      "1708/2399\n",
      "1709/2399\n",
      "1710/2399\n",
      "1711/2399\n",
      "1712/2399\n",
      "1713/2399\n",
      "1714/2399\n",
      "1715/2399\n",
      "1716/2399\n",
      "1717/2399\n",
      "1718/2399\n",
      "1719/2399\n",
      "1720/2399\n",
      "1721/2399\n",
      "1722/2399\n",
      "1723/2399\n",
      "1724/2399\n",
      "1725/2399\n",
      "1726/2399\n",
      "1727/2399\n",
      "1728/2399\n",
      "1729/2399\n",
      "1730/2399\n",
      "1731/2399\n",
      "1732/2399\n",
      "1733/2399\n",
      "1734/2399\n",
      "1735/2399\n",
      "1736/2399\n",
      "1737/2399\n",
      "1738/2399\n",
      "1739/2399\n",
      "1740/2399\n",
      "1741/2399\n",
      "1742/2399\n",
      "1743/2399\n",
      "1744/2399\n",
      "1745/2399\n",
      "1746/2399\n",
      "1747/2399\n",
      "1748/2399\n",
      "1749/2399\n",
      "1750/2399\n",
      "1751/2399\n",
      "1752/2399\n",
      "1753/2399\n",
      "1754/2399\n",
      "1755/2399\n",
      "1756/2399\n",
      "1757/2399\n",
      "1758/2399\n",
      "1759/2399\n",
      "1760/2399\n",
      "1761/2399\n",
      "1762/2399\n",
      "1763/2399\n",
      "1764/2399\n",
      "1765/2399\n",
      "1766/2399\n",
      "1767/2399\n",
      "1768/2399\n",
      "1769/2399\n",
      "1770/2399\n",
      "1771/2399\n",
      "1772/2399\n",
      "1773/2399\n",
      "1774/2399\n",
      "1775/2399\n",
      "1776/2399\n",
      "1777/2399\n",
      "1778/2399\n",
      "1779/2399\n",
      "1780/2399\n",
      "1781/2399\n",
      "1782/2399\n",
      "1783/2399\n",
      "1784/2399\n",
      "1785/2399\n",
      "1786/2399\n",
      "1787/2399\n",
      "1788/2399\n",
      "1789/2399\n",
      "1790/2399\n",
      "1791/2399\n",
      "1792/2399\n",
      "1793/2399\n",
      "1794/2399\n",
      "1795/2399\n",
      "1796/2399\n",
      "1797/2399\n",
      "1798/2399\n",
      "1799/2399\n",
      "1800/2399\n",
      "1801/2399\n",
      "1802/2399\n",
      "1803/2399\n",
      "1804/2399\n",
      "1805/2399\n",
      "1806/2399\n",
      "1807/2399\n",
      "1808/2399\n",
      "1809/2399\n",
      "1810/2399\n",
      "1811/2399\n",
      "1812/2399\n",
      "1813/2399\n",
      "1814/2399\n",
      "1815/2399\n",
      "1816/2399\n",
      "1817/2399\n",
      "1818/2399\n",
      "1819/2399\n",
      "1820/2399\n",
      "1821/2399\n",
      "1822/2399\n",
      "1823/2399\n",
      "1824/2399\n",
      "1825/2399\n",
      "1826/2399\n",
      "1827/2399\n",
      "1828/2399\n",
      "1829/2399\n",
      "1830/2399\n",
      "1831/2399\n",
      "1832/2399\n",
      "1833/2399\n",
      "1834/2399\n",
      "1835/2399\n",
      "1836/2399\n",
      "1837/2399\n",
      "1838/2399\n",
      "1839/2399\n",
      "1840/2399\n",
      "1841/2399\n",
      "1842/2399\n",
      "1843/2399\n",
      "1844/2399\n",
      "1845/2399\n",
      "1846/2399\n",
      "1847/2399\n",
      "1848/2399\n",
      "1849/2399\n",
      "1850/2399\n",
      "1851/2399\n",
      "1852/2399\n",
      "1853/2399\n",
      "1854/2399\n",
      "1855/2399\n",
      "1856/2399\n",
      "1857/2399\n",
      "1858/2399\n",
      "1859/2399\n",
      "1860/2399\n",
      "1861/2399\n",
      "1862/2399\n",
      "1863/2399\n",
      "1864/2399\n",
      "1865/2399\n",
      "1866/2399\n",
      "1867/2399\n",
      "1868/2399\n",
      "1869/2399\n",
      "1870/2399\n",
      "1871/2399\n",
      "1872/2399\n",
      "1873/2399\n",
      "1874/2399\n",
      "1875/2399\n",
      "1876/2399\n",
      "1877/2399\n",
      "1878/2399\n",
      "1879/2399\n",
      "1880/2399\n",
      "1881/2399\n",
      "1882/2399\n",
      "1883/2399\n",
      "1884/2399\n",
      "1885/2399\n",
      "1886/2399\n",
      "1887/2399\n",
      "1888/2399\n",
      "1889/2399\n",
      "1890/2399\n",
      "1891/2399\n",
      "1892/2399\n",
      "1893/2399\n",
      "1894/2399\n",
      "1895/2399\n",
      "1896/2399\n",
      "1897/2399\n",
      "1898/2399\n",
      "1899/2399\n",
      "1900/2399\n",
      "1901/2399\n",
      "1902/2399\n",
      "1903/2399\n",
      "1904/2399\n",
      "1905/2399\n",
      "1906/2399\n",
      "1907/2399\n",
      "1908/2399\n",
      "1909/2399\n",
      "1910/2399\n",
      "1911/2399\n",
      "1912/2399\n",
      "1913/2399\n",
      "1914/2399\n",
      "1915/2399\n",
      "1916/2399\n",
      "1917/2399\n",
      "1918/2399\n",
      "1919/2399\n",
      "1920/2399\n",
      "1921/2399\n",
      "1922/2399\n",
      "1923/2399\n",
      "1924/2399\n",
      "1925/2399\n",
      "1926/2399\n",
      "1927/2399\n",
      "1928/2399\n",
      "1929/2399\n",
      "1930/2399\n",
      "1931/2399\n",
      "1932/2399\n",
      "1933/2399\n",
      "1934/2399\n",
      "1935/2399\n",
      "1936/2399\n",
      "1937/2399\n",
      "1938/2399\n",
      "1939/2399\n",
      "1940/2399\n",
      "1941/2399\n",
      "1942/2399\n",
      "1943/2399\n",
      "1944/2399\n",
      "1945/2399\n",
      "1946/2399\n",
      "1947/2399\n",
      "1948/2399\n",
      "1949/2399\n",
      "1950/2399\n",
      "1951/2399\n",
      "1952/2399\n",
      "1953/2399\n",
      "1954/2399\n",
      "1955/2399\n",
      "1956/2399\n",
      "1957/2399\n",
      "1958/2399\n",
      "1959/2399\n",
      "1960/2399\n",
      "1961/2399\n",
      "1962/2399\n",
      "1963/2399\n",
      "1964/2399\n",
      "1965/2399\n",
      "1966/2399\n",
      "1967/2399\n",
      "1968/2399\n",
      "1969/2399\n",
      "1970/2399\n",
      "1971/2399\n",
      "1972/2399\n",
      "1973/2399\n",
      "1974/2399\n",
      "1975/2399\n",
      "1976/2399\n",
      "1977/2399\n",
      "1978/2399\n",
      "1979/2399\n",
      "1980/2399\n",
      "1981/2399\n",
      "1982/2399\n",
      "1983/2399\n",
      "1984/2399\n",
      "1985/2399\n",
      "1986/2399\n",
      "1987/2399\n",
      "1988/2399\n",
      "1989/2399\n",
      "1990/2399\n",
      "1991/2399\n",
      "1992/2399\n",
      "1993/2399\n",
      "1994/2399\n",
      "1995/2399\n",
      "1996/2399\n",
      "1997/2399\n",
      "1998/2399\n",
      "1999/2399\n",
      "2000/2399\n",
      "2001/2399\n",
      "2002/2399\n",
      "2003/2399\n",
      "2004/2399\n",
      "2005/2399\n",
      "2006/2399\n",
      "2007/2399\n",
      "2008/2399\n",
      "2009/2399\n",
      "2010/2399\n",
      "2011/2399\n",
      "2012/2399\n",
      "2013/2399\n",
      "2014/2399\n",
      "2015/2399\n",
      "2016/2399\n",
      "2017/2399\n",
      "2018/2399\n",
      "2019/2399\n",
      "2020/2399\n",
      "2021/2399\n",
      "2022/2399\n",
      "2023/2399\n",
      "2024/2399\n",
      "2025/2399\n",
      "2026/2399\n",
      "2027/2399\n",
      "2028/2399\n",
      "2029/2399\n",
      "2030/2399\n",
      "2031/2399\n",
      "2032/2399\n",
      "2033/2399\n",
      "2034/2399\n",
      "2035/2399\n",
      "2036/2399\n",
      "2037/2399\n",
      "2038/2399\n",
      "2039/2399\n",
      "2040/2399\n",
      "2041/2399\n",
      "2042/2399\n",
      "2043/2399\n",
      "2044/2399\n",
      "2045/2399\n",
      "2046/2399\n",
      "2047/2399\n",
      "2048/2399\n",
      "2049/2399\n",
      "2050/2399\n",
      "2051/2399\n",
      "2052/2399\n",
      "2053/2399\n",
      "2054/2399\n",
      "2055/2399\n",
      "2056/2399\n",
      "2057/2399\n",
      "2058/2399\n",
      "2059/2399\n",
      "2060/2399\n",
      "2061/2399\n",
      "2062/2399\n",
      "2063/2399\n",
      "2064/2399\n",
      "2065/2399\n",
      "2066/2399\n",
      "2067/2399\n",
      "2068/2399\n",
      "2069/2399\n",
      "2070/2399\n",
      "2071/2399\n",
      "2072/2399\n",
      "2073/2399\n",
      "2074/2399\n",
      "2075/2399\n",
      "2076/2399\n",
      "2077/2399\n",
      "2078/2399\n",
      "2079/2399\n",
      "2080/2399\n",
      "2081/2399\n",
      "2082/2399\n",
      "2083/2399\n",
      "2084/2399\n",
      "2085/2399\n",
      "2086/2399\n",
      "2087/2399\n",
      "2088/2399\n",
      "2089/2399\n",
      "2090/2399\n",
      "2091/2399\n",
      "2092/2399\n",
      "2093/2399\n",
      "2094/2399\n",
      "2095/2399\n",
      "2096/2399\n",
      "2097/2399\n",
      "2098/2399\n",
      "2099/2399\n",
      "2100/2399\n",
      "2101/2399\n",
      "2102/2399\n",
      "2103/2399\n",
      "2104/2399\n",
      "2105/2399\n",
      "2106/2399\n",
      "2107/2399\n",
      "2108/2399\n",
      "2109/2399\n",
      "2110/2399\n",
      "2111/2399\n",
      "2112/2399\n",
      "2113/2399\n",
      "2114/2399\n",
      "2115/2399\n",
      "2116/2399\n",
      "2117/2399\n",
      "2118/2399\n",
      "2119/2399\n",
      "2120/2399\n",
      "2121/2399\n",
      "2122/2399\n",
      "2123/2399\n",
      "2124/2399\n",
      "2125/2399\n",
      "2126/2399\n",
      "2127/2399\n",
      "2128/2399\n",
      "2129/2399\n",
      "2130/2399\n",
      "2131/2399\n",
      "2132/2399\n",
      "2133/2399\n",
      "2134/2399\n",
      "2135/2399\n",
      "2136/2399\n",
      "2137/2399\n",
      "2138/2399\n",
      "2139/2399\n",
      "2140/2399\n",
      "2141/2399\n",
      "2142/2399\n",
      "2143/2399\n",
      "2144/2399\n",
      "2145/2399\n",
      "2146/2399\n",
      "2147/2399\n",
      "2148/2399\n",
      "2149/2399\n",
      "2150/2399\n",
      "2151/2399\n",
      "2152/2399\n",
      "2153/2399\n",
      "2154/2399\n",
      "2155/2399\n",
      "2156/2399\n",
      "2157/2399\n",
      "2158/2399\n",
      "2159/2399\n",
      "2160/2399\n",
      "2161/2399\n",
      "2162/2399\n",
      "2163/2399\n",
      "2164/2399\n",
      "2165/2399\n",
      "2166/2399\n",
      "2167/2399\n",
      "2168/2399\n",
      "2169/2399\n",
      "2170/2399\n",
      "2171/2399\n",
      "2172/2399\n",
      "2173/2399\n",
      "2174/2399\n",
      "2175/2399\n",
      "2176/2399\n",
      "2177/2399\n",
      "2178/2399\n",
      "2179/2399\n",
      "2180/2399\n",
      "2181/2399\n",
      "2182/2399\n",
      "2183/2399\n",
      "2184/2399\n",
      "2185/2399\n",
      "2186/2399\n",
      "2187/2399\n",
      "2188/2399\n",
      "2189/2399\n",
      "2190/2399\n",
      "2191/2399\n",
      "2192/2399\n",
      "2193/2399\n",
      "2194/2399\n",
      "2195/2399\n",
      "2196/2399\n",
      "2197/2399\n",
      "2198/2399\n",
      "2199/2399\n",
      "2200/2399\n",
      "2201/2399\n",
      "2202/2399\n",
      "2203/2399\n",
      "2204/2399\n",
      "2205/2399\n",
      "2206/2399\n",
      "2207/2399\n",
      "2208/2399\n",
      "2209/2399\n",
      "2210/2399\n",
      "2211/2399\n",
      "2212/2399\n",
      "2213/2399\n",
      "2214/2399\n",
      "2215/2399\n",
      "2216/2399\n",
      "2217/2399\n",
      "2218/2399\n",
      "2219/2399\n",
      "2220/2399\n",
      "2221/2399\n",
      "2222/2399\n",
      "2223/2399\n",
      "2224/2399\n",
      "2225/2399\n",
      "2226/2399\n",
      "2227/2399\n",
      "2228/2399\n",
      "2229/2399\n",
      "2230/2399\n",
      "2231/2399\n",
      "2232/2399\n",
      "2233/2399\n",
      "2234/2399\n",
      "2235/2399\n",
      "2236/2399\n",
      "2237/2399\n",
      "2238/2399\n",
      "2239/2399\n",
      "2240/2399\n",
      "2241/2399\n",
      "2242/2399\n",
      "2243/2399\n",
      "2244/2399\n",
      "2245/2399\n",
      "2246/2399\n",
      "2247/2399\n",
      "2248/2399\n",
      "2249/2399\n",
      "2250/2399\n",
      "2251/2399\n",
      "2252/2399\n",
      "2253/2399\n",
      "2254/2399\n",
      "2255/2399\n",
      "2256/2399\n",
      "2257/2399\n",
      "2258/2399\n",
      "2259/2399\n",
      "2260/2399\n",
      "2261/2399\n",
      "2262/2399\n",
      "2263/2399\n",
      "2264/2399\n",
      "2265/2399\n",
      "2266/2399\n",
      "2267/2399\n",
      "2268/2399\n",
      "2269/2399\n",
      "2270/2399\n",
      "2271/2399\n",
      "2272/2399\n",
      "2273/2399\n",
      "2274/2399\n",
      "2275/2399\n",
      "2276/2399\n",
      "2277/2399\n",
      "2278/2399\n",
      "2279/2399\n",
      "2280/2399\n",
      "2281/2399\n",
      "2282/2399\n",
      "2283/2399\n",
      "2284/2399\n",
      "2285/2399\n",
      "2286/2399\n",
      "2287/2399\n",
      "2288/2399\n",
      "2289/2399\n",
      "2290/2399\n",
      "2291/2399\n",
      "2292/2399\n",
      "2293/2399\n",
      "2294/2399\n",
      "2295/2399\n",
      "2296/2399\n",
      "2297/2399\n",
      "2298/2399\n",
      "2299/2399\n",
      "2300/2399\n",
      "2301/2399\n",
      "2302/2399\n",
      "2303/2399\n",
      "2304/2399\n",
      "2305/2399\n",
      "2306/2399\n",
      "2307/2399\n",
      "2308/2399\n",
      "2309/2399\n",
      "2310/2399\n",
      "2311/2399\n",
      "2312/2399\n",
      "2313/2399\n",
      "2314/2399\n",
      "2315/2399\n",
      "2316/2399\n",
      "2317/2399\n",
      "2318/2399\n",
      "2319/2399\n",
      "2320/2399\n",
      "2321/2399\n",
      "2322/2399\n",
      "2323/2399\n",
      "2324/2399\n",
      "2325/2399\n",
      "2326/2399\n",
      "2327/2399\n",
      "2328/2399\n",
      "2329/2399\n",
      "2330/2399\n",
      "2331/2399\n",
      "2332/2399\n",
      "2333/2399\n",
      "2334/2399\n",
      "2335/2399\n",
      "2336/2399\n",
      "2337/2399\n",
      "2338/2399\n",
      "2339/2399\n",
      "2340/2399\n",
      "2341/2399\n",
      "2342/2399\n",
      "2343/2399\n",
      "2344/2399\n",
      "2345/2399\n",
      "2346/2399\n",
      "2347/2399\n",
      "2348/2399\n",
      "2349/2399\n",
      "2350/2399\n",
      "2351/2399\n",
      "2352/2399\n",
      "2353/2399\n",
      "2354/2399\n",
      "2355/2399\n",
      "2356/2399\n",
      "2357/2399\n",
      "2358/2399\n",
      "2359/2399\n",
      "2360/2399\n",
      "2361/2399\n",
      "2362/2399\n",
      "2363/2399\n",
      "2364/2399\n",
      "2365/2399\n",
      "2366/2399\n",
      "2367/2399\n",
      "2368/2399\n",
      "2369/2399\n",
      "2370/2399\n",
      "2371/2399\n",
      "2372/2399\n",
      "2373/2399\n",
      "2374/2399\n",
      "2375/2399\n",
      "2376/2399\n",
      "2377/2399\n",
      "2378/2399\n",
      "2379/2399\n",
      "2380/2399\n",
      "2381/2399\n",
      "2382/2399\n",
      "2383/2399\n",
      "2384/2399\n",
      "2385/2399\n",
      "2386/2399\n",
      "2387/2399\n",
      "2388/2399\n",
      "2389/2399\n",
      "2390/2399\n",
      "2391/2399\n",
      "2392/2399\n",
      "2393/2399\n",
      "2394/2399\n",
      "2395/2399\n",
      "2396/2399\n",
      "2397/2399\n",
      "2398/2399\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Gita/anaconda/lib/python3.4/site-packages/matplotlib/__init__.py:872: UserWarning: axes.color_cycle is deprecated and replaced with axes.prop_cycle; please use the latter.\n",
      "  warnings.warn(self.msg_depr % (key, alt_key))\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import sklearn\n",
    "import matplotlib.pyplot as plt\n",
    "import ete3 as ete\n",
    "import os\n",
    "import Bio\n",
    "import random\n",
    "import scipy.stats as stats\n",
    "from functools import *\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "import matplotlib.patches as mpatches\n",
    "import seaborn as sns\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "\n",
    "#as part of quality control the following viruses had to be omitted from our analysis\n",
    "#they contain misannotations or mislabeled hosts, etc. \n",
    "badFolders = ['Acyrthosiphon_pisum_secondary_endosymbiont_phage_1_uid14047', 'Aureococcus_anophagefferens_virus_MM_2014_uid258005', 'Dragonfly_associated_microphage_1_uid177547', 'Enterobacter_phage_PG7_uid240014', 'Escherichia_phage_vB_EcoM_JS09_uid248321', 'Lactobacillus_phage_LL_H_uid19803', 'Providencia_phage_Redjac_uid177540', 'Pseudoalteromonas_phage_RIO_1_uid206039', 'Sputnik_virophage_2_uid243067', 'Sputnik_virophage_3_uid243065', 'Staphylococcus_phage_6ec_uid253318', 'Streptococcus_phage_DCC1738_uid253219', 'Streptococcus_phage_IC1_uid253220', 'Streptococcus_phage_K13_uid253223', 'Streptococcus_phage_SMP_uid18529', 'Vibrio_phage_CP_T1_uid181062', 'Vibrio_phage_vpms1_uid212709', 'Vibrio_phage_VPUSM_8_uid227006',  'Abaca_bunchy_top_virus_uid28697','Shallot_yellow_stripe_virus_uid15745','Equine_rhinitis_A_virus_uid15205','African_pouched_rat_arterivirus_uid274797','Spilanthes_yellow_vein_virus_uid19779', 'Velvet_bean_severe_mosaic_virus_uid41175','Paramecium_bursaria_Chlorella_virus_FR483_uid18305','Human_endogenous_retrovirus_K113_uid222261','Acholeplasma_phage_MV_L1_uid14573']\n",
    "#viruses whose lineage is improperly annotated (they don't belong to any Baltimore group)\n",
    "unknowns= ['Bacillus_phage_phBC6A51_uid15021', 'Bacillus_phage_phBC6A52_uid15022', 'Badger_feces_associated_gemycircularvirus_uid281125', 'Chaetoceros_DNA_virus_7_uid237558', 'Chaetoceros_lorenzianus_DNA_Virus_uid63565', 'Chaetoceros_tenuissimus_DNA_virus_uid60753', 'Cladosporium_cladosporioides_virus_1_uid258308', 'Enterobacteria_phage_YYZ_2008_uid32231', 'European_mountain_ash_ringspot_associated_virus_uid39973', 'Faecal_associated_gemycircularvirus_3_uid268557', 'Faecal_associated_gemycircularvirus_4_uid268556', 'Faecal_associated_gemycircularvirus_5_uid268555', 'Faecal_associated_gemycircularvirus_6_uid268554', 'Faecal_associated_gemycircularvirus_7_uid268553', 'Gentian_ovary_ring_spot_virus_uid256090', 'Geobacillus_virus_E2_uid19797', 'Glossina_pallidipes_salivary_gland_hypertrophy_virus_uid28839', 'Haloarcula_hispanica_icosahedral_virus_2_uid109269', 'Haloarcula_phage_SH1_uid15535', 'Halovirus_PH1_uid196975', 'Halovirus_VNH_1_uid262927', 'Human_genital_associated_circular_DNA_virus_1_uid281399', 'Lactobacillus_johnsonii_prophage_Lj771_uid28145', 'Magnaporthe_oryzae_RNA_virus_uid272442', 'Mongoose_feces_associated_gemycircularvirus_a_uid281407', 'Mongoose_feces_associated_gemycircularvirus_b_uid281406', 'Mongoose_feces_associated_gemycircularvirus_c_uid281404', 'Mongoose_feces_associated_gemycircularvirus_d_uid281405', 'Mycoplasma_phage_phiMFV1_uid14387', 'Penicillium_roqueforti_dsRNA_mycovirus_1_uid258162', 'Phage_Gifsy_1_uid32269', 'Phage_Gifsy_2_uid32271', 'Pigeonpea_cryptic_virus_uid244664', 'Planaria_asexual_strain_specific_virus_like_element_type_1_uid14140', 'Pseudomonas_phage_phi_Pto_bp6g_uid240724', 'Rhizoctonia_fumigata_mycovirus_uid283068', 'Rhodococcus_phage_REQ2_uid81171', 'Rhodococcus_phage_REQ3_uid81175', 'Rose_rosette_virus_uid64937', 'Rosellinia_necatrix_fusarivirus_1_uid255787', 'Rosellinia_necatrix_megabirnavirus_1_W779_uid41609', 'Salisaeta_icosahedral_phage_1_uid167575', 'Salmonella_phage_Fels_1_uid29267', 'Sodalis_phage_phiSG1_uid16583', 'Staphylococcus_phage_phi2958PVL_uid32173', 'Staphylococcus_phage_tp310_1_uid20659', 'Staphylococcus_phage_tp310_3_uid20663', 'Stenotrophomonas_phage_phiSMA9_uid15493', 'Streptococcus_phage_20617_uid239271', 'Streptococcus_phage_phiBHN167_uid227353', 'Streptococcus_pyogenes_phage_315_1_uid14533', 'Streptococcus_pyogenes_phage_315_2_uid14528', 'Streptococcus_pyogenes_phage_315_3_uid14529', 'Streptococcus_pyogenes_phage_315_4_uid14530', 'Streptococcus_pyogenes_phage_315_5_uid14531', 'Streptococcus_pyogenes_phage_315_6_uid14532', 'Tanay_virus_uid246129', 'Thermococcus_prieurii_virus_1_uid84407', 'Thermus_phage_IN93_uid14235', 'Thermus_phage_P23_77_uid40235']\n",
    "badFolders=badFolders+unknowns\n",
    "#these are satellite viruses \n",
    "sats = ['Ageratum_conyzoides_associated_symptomless_alphasatellite_uid259293', 'Ageratum_enation_alphasatellite_uid181994', 'Ageratum_leaf_curl_betasatellite_uid195929', 'Ageratum_leaf_curl_Cameroon_betasatellite_uid36669', 'Ageratum_yellow_leaf_curl_betasatellite_uid14439', 'Ageratum_yellow_vein_China_alphasatellite_uid237561', 'Ageratum_yellow_vein_China_virus_associated_DNA_beta_uid15515', 'Ageratum_yellow_vein_Singapore_alphasatellite_uid14232', 'Ageratum_yellow_vein_virus_satellite_DNA_beta_uid14444', 'Alternanthera_yellow_vein_virus_satellite_DNA_beta_uid19833', 'Andrographis_yellow_vein_leaf_curl_betasatellite_uid243492', 'Bhendi_yellow_vein_India_betasatellite__India_Aurangabad_OY164_2006__uid61557', 'Bhendi_yellow_vein_mosaic_betasatellite__India_Coimbator_OYCO1_2005__uid61777', 'Bhendi_yellow_vein_mosaic_virus_associated_alphasatellite_uid174781', 'Bhendi_yellow_vein_mosaic_virus_satellite_DNA_beta_uid14445', 'Black_medic_leafroll_alphasatellite_1_uid243500', 'Cardamom_bushy_dwarf_virus_satellite_uid230064', 'Cassava_mosaic_Madagascar_alphasatellite_uid175666', 'Chili_leaf_curl_Bhatinda_betasatellite_uid206467', 'Chilli_leaf_curl_alphasatellite_uid253224', 'Chilli_leaf_curl_Multan_alphasatellite_uid39933', 'Chilli_leaf_curl_virus_satellite_DNA_beta_uid14441', 'Cleome_leaf_crumple_virus_associated_DNA_1_uid60045', 'Corchorus_yellow_vein_mosaic_betasatellite_uid192608', 'Cotton_leaf_curl_Burewala_alphasatellite_uid45935', 'Cotton_leaf_curl_Burewala_betasatellite_uid45933', 'Cotton_leaf_curl_Gezira_alphasatellite_uid42507', 'Cotton_leaf_curl_Gezira_virus_satellite_DNA_beta_uid15166', 'Cotton_leaf_curl_Multan_betasatellite_uid15780', 'Cotton_leaf_curl_virus_associated_DNA_1_isolate_Lucknow_uid65305', 'Cotton_leaf_curl_virus_associated_DNA_beta_uid14438', 'Cotton_leaf_curl_virus_betasatellite_uid162497', 'Cowpea_severe_leaf_curl_associated_DNA_beta_uid15157', 'Croton_yellow_vein_mosaic_alphasatellite_uid45931', 'Croton_yellow_vein_mosaic_betasatellite_uid18249', 'Cuban_alphasatellite_1_uid210798', 'Dragonfly_associated_alphasatellite_uid181244', 'Emilia_yellow_vein_virus_associated_DNA_beta_uid37893', 'Erectites_yellow_mosaic_virus_satellite_DNA_beta_uid19827', 'Eupatorium_yellow_vein_virus_satellite_DNA_beta_uid14447', 'Faba_bean_necrotic_stunt_alphasatellite_1_uid243499', 'Faba_bean_necrotic_stunt_alphasatellite_2_uid243498', 'French_bean_leaf_curl_betasatellite_Kanpur_uid169556', 'Gossypium_darwinii_symptomless_alphasatellite_uid39593', 'Gossypium_davidsonii_symptomless_alphasatellite_uid39589', 'Gossypium_mustilinum_symptomless_alphasatellite_uid39591', 'Grapevine_satellite_virus_uid208539', 'Guar_leaf_curl_alphasatellite_uid193981', 'Hedyotis_uncinella_yellow_mosaic_betasatellite_uid230991', 'Honeysuckle_yellow_vein_mosaic_disease_associated_satellite_DNA_beta_uid19863', 'Honeysuckle_yellow_vein_mosaic_virus_satellite_DNA_beta_uid14620', 'Leucas_zeylanica_yellow_vein_virus_satellite_DNA_beta_uid41305', 'Ludwigia_leaf_distortion_betasatellite__India_Amadalavalasa_Hibiscus_2007__uid29233', 'Ludwigia_yellow_vein_virus_associated_DNA_beta_uid15561', 'Luffa_puckering_and_leaf_distortion_associated_betasatellite__India_Gurdaspur_Okra_2013__uid253218', 'Luffa_puckering_and_leaf_distortion_associated_DNA_beta_uid15779', 'Malachra_yellow_vein_mosaic_virus_associated_satellite_DNA_beta_uid28727', 'Malvastrum_leaf_curl_betasatellite_uid16301', 'Malvastrum_leaf_curl_Guangdong_betasatellite_uid243827', 'Malvastrum_yellow_vein_Yunnan_virus_satellite_DNA_beta_uid14567', 'Melon_chlorotic_mosaic_virus_associated_alphasatellite_uid51413', 'Mesta_yellow_vein_mosaic_virus_alphasatellite__India_Hoshiarpur_Okra_2013__uid253217', 'Mesta_yellow_vein_mosaic_virus_associated_alphasatellite_uid174780', 'Mesta_yellow_vein_mosaic_virus_associated_DNA_beta_uid21015', 'Milk_vetch_dwarf_C10_alphasatellite_uid240682', 'Mirabilis_leaf_curl_India_virus_associated_betasatellite_uid255800', 'Mungbean_yellow_mosaic_India_virus_associated_betasatellite__India__Faizabad__Cow_Pea_2012__uid177773', 'Nanovirus_like_particle_uid14386', 'Okra_enation_leaf_curl_betasatellite__India_Sonipat_EL10_2006__uid61781', 'Okra_leaf_curl_Mali_virus_satellite_DNA_beta_uid20323', 'Okra_yellow_vein_disease_associated_sequence_uid14443', 'Olive_viral_satellite_RNA_uid230268', 'Papaya_leaf_curl_alphasatellite_uid233414', 'Papaya_leaf_curl_China_virus_satellite_DNA_beta_uid19819', 'Pepper_leaf_curl_virus_satellite_DNA_beta_uid28283', 'Pepper_leaf_curl_Yunnan_virus_satellite_DNA_beta_uid29415', 'Potato_apical_leaf_curl_disease_associated_satellite_DNA_beta_uid18323', 'Radish_leaf_curl_virus_betasatellite_uid28281', 'Sida_leaf_curl_virus_satellite_DNA_beta_uid19823', 'Sida_yellow_mosaic_virus_China_associated_DNA_beta_uid15514', 'Sida_yellow_vein_China_alphasatellite_uid238950', 'Siegesbeckia_yellow_vein_virus_associated_DNA_beta_uid17269', 'Subterranean_clover_stunt_C6_alphasatellite_uid14180', 'Synedrella_leaf_curl_alphasatellite_uid263034', 'Tobacco_leaf_chlorosis_betasatellite_uid178075', 'Tobacco_leaf_curl_disease_associated_sequence_uid14442', 'Tobacco_leaf_curl_PUSA_alphasatellite_uid56023', 'Tobacco_leaf_curl_virus_associated_DNA_beta_uid45925', 'Tomato_leaf_curl_Bangladesh_betasatellite_uid56017', 'Tomato_leaf_curl_Gandhinagar_betasatellite_uid231683', 'Tomato_leaf_curl_Hajipur_betasatellite_uid175587', 'Tomato_leaf_curl_Joydebpur_virus_satellite_DNA_beta_uid28273', 'Tomato_leaf_curl_Karnataka_alphasatellite_uid181995', 'Tomato_leaf_curl_Pakistan_virus_associated_DNA_1_uid38463', 'Tomato_leaf_curl_Patna_betasatellite_uid36541', 'Tomato_leaf_curl_Togo_betasatellite_Togo_2006__uid60629', 'Tomato_leaf_curl_Yemen_betasatellite_uid177643', 'Tomato_yellow_leaf_curl_China_betasatellite_uid181248', 'Tomato_yellow_leaf_curl_Vietnam_virus_satellite_DNA_beta_uid19829', 'TYLCCNV_Y322__satellite_DNA_beta_uid16338', 'Vernonia_yellow_vein_betasatellite_uid41303', 'Vernonia_yellow_vein_Fujian_virus_alphasatellite_uid72145', 'Vernonia_yellow_vein_Fujian_virus_betasatellite_uid72143', 'Zinnia_leaf_curl_disease_associated_sequence_uid14440', 'Zinnia_leaf_curl_virus_associated_DNA_beta_uid14538']\n",
    "\n",
    "\n",
    "def getTaxidFromGenomeFolder(genome_folder):\n",
    "    #Takes an rpt file for a genome and returns the taxid in it.\n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        with open(os.path.join('all_rpt', folder, os.listdir(os.path.join('all_rpt', folder))[0])) as rpt_file:\n",
    "            for line in rpt_file:\n",
    "                if line.startswith('Taxid: '):\n",
    "                    return int(line[7:-1])\n",
    "                \n",
    "\n",
    "def getHostTaxidFromGenomeFolder(genome_folder):\n",
    "    #Takes a folder containing an asn file and returns the taxid of the host of the virus described by that asn file. \n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        with open('folderToHostTaxid.txt', 'r') as taxid2host:\n",
    "            for line in taxid2host:\n",
    "                if genome_folder in line:\n",
    "                    return int(line.split('\\t')[1][:-1])\n",
    "        return 0\n",
    "    \n",
    "\n",
    "def getGenomeLengthFromGenomeFolder(genome_folder):\n",
    "    total = 0\n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        for file in os.listdir(os.path.join('all_ptt', folder)):\n",
    "            with open(os.path.join('all_ptt', folder, file)) as ptt_file:\n",
    "                line = ptt_file.readline()\n",
    "                index = line.find('- 1..')\n",
    "                if index == -1:\n",
    "                    return -1\n",
    "                num = line[index + 5 : -1]\n",
    "                total += int(num)\n",
    "        return total\n",
    "    \n",
    "\n",
    "def getCodingNumberAndOverlapNumberFromGenomeFolder(folder):\n",
    "    totalCoding = 0\n",
    "    totalOverlap = 0\n",
    "    for file in os.listdir(os.path.join('all_ptt', folder)):\n",
    "        #these folders contain genomes with misannotation of gene overlap,non-coding DNA, or host information from NCBI. \n",
    "        if folder not in badFolders and folder not in sats:\n",
    "            with open(os.path.join('all_ptt', folder, file)) as ptt_file:\n",
    "                line = ptt_file.readline()\n",
    "                index = line.find('- 1..')\n",
    "                repliconLength = int(line[index + 5 : -1])\n",
    "                indicesThatOverlap = [0] * repliconLength\n",
    "                line = ptt_file.readline()\n",
    "                line = ptt_file.readline()\n",
    "\n",
    "                uniqueIndices = set([])\n",
    "                for line in ptt_file:                                     \n",
    "                    #this part is to collect beginning and end indices for genes\n",
    "                    line = line.split('\\t')[0]\n",
    "                    [start, end] = [int(x) for x in line.split('..')]\n",
    "                    #this is put in place to distinguish linear from circular genomes. Here, we're looking for linear genomes\n",
    "                    if start < end:                    \n",
    "                        for i in range(start, end):\n",
    "                            if i in uniqueIndices:\n",
    "                                indicesThatOverlap[i] = 1\n",
    "                            else:\n",
    "                                uniqueIndices.add(i)\n",
    "                    #here were are looking for circular genomes to be able to properly count noncoding bases\n",
    "                    elif start > end:\n",
    "                        for i in range(start, repliconLength):\n",
    "                            if i in uniqueIndices:\n",
    "                                indicesThatOverlap[i] = 1\n",
    "                            else:\n",
    "                                uniqueIndices.add(i)\n",
    "                        for i in range(1, end):\n",
    "                            if i in uniqueIndices:\n",
    "                                indicesThatOverlap[i] = 1\n",
    "                            else:\n",
    "                                uniqueIndices.add(i)          \n",
    "                \n",
    "                totalCoding += len(uniqueIndices)\n",
    "                #totalOverlap += len(bigList) - len(uniqueIndices)\n",
    "                totalOverlap += sum(indicesThatOverlap)\n",
    "    return (totalCoding, totalOverlap)\n",
    "\n",
    "\n",
    "def getProtienNamesFromGenomeFolder(folder):\n",
    "    protVec=[]\n",
    "    for file in os.listdir(os.path.join('all_ptt', folder)):\n",
    "        #these folders contain genomes with misannotation of gene overlap, non-coding DNA, or host information from NCBI. \n",
    "        if folder not in badFolders and folder not in sats:\n",
    "            with open(os.path.join('all_ptt', folder, file)) as ptt_file:\n",
    "                line = ptt_file.readline()\n",
    "                line = ptt_file.readline()\n",
    "                line = ptt_file.readline()               \n",
    "                for line in ptt_file:                    \n",
    "                    #this is to collect the name of proteins from ptt files                   \n",
    "                    protVec.append(line.split('\\t')[-1].rstrip())                                    \n",
    "    return (protVec)\n",
    "\n",
    "\n",
    "genomeLengths = []\n",
    "lengthData = []\n",
    "counter = 0\n",
    "length_lists = []\n",
    "hostTaxidList = []\n",
    "taxidList = []\n",
    "folders = []\n",
    "proteinNames=[]\n",
    "dbObject = ete.NCBITaxa()\n",
    "\n",
    "for folder in os.listdir('all_faa'):\n",
    "    #these entries had problems with their gene overlap, protein, and noncoding annotation so they're excluded\n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        length_lists.append([])\n",
    "        for fileName in os.listdir(os.path.join('all_faa', folder)):\n",
    "            with open(os.path.join('all_faa', folder, fileName), \"r\") as faa_file:\n",
    "                proteins = SeqIO.parse(faa_file, 'fasta')\n",
    "                for protein in proteins:\n",
    "                    length_lists[counter].append(float(len(protein.seq)))\n",
    "\n",
    "        hostTaxid = getHostTaxidFromGenomeFolder(folder)\n",
    "        # These taxids come viruses with unidentified hosts \n",
    "        if hostTaxid and hostTaxid != 11086 and hostTaxid != 10036 and hostTaxid != 37965 and hostTaxid != 10640  and hostTaxid != 32644 and hostTaxid != 1 and hostTaxid != 212035 and hostTaxid != 1051671:\n",
    "            taxidList.append(getTaxidFromGenomeFolder(folder))\n",
    "            hostTaxidList.append(hostTaxid)\n",
    "            genomeLengths.append(getGenomeLengthFromGenomeFolder(folder))\n",
    "            folders.append(folder)\n",
    "            proteinNames.append(getProtienNamesFromGenomeFolder(folder))\n",
    "            counter += 1\n",
    "        else:\n",
    "            length_lists = length_lists[:-1]\n",
    "\n",
    "        if counter % 100 == 0:\n",
    "            print(str(int(np.round(float(counter) * 100.0 / 2701.0))) + '% done.')\n",
    "\n",
    "taxidLineages = [dbObject.get_lineage(t) for t in taxidList]\n",
    "englishLineages = [[str(rank) for rank in dbObject.translate_to_names(lin)]\n",
    "                   for lin in taxidLineages]\n",
    "hostTaxidLineages = [dbObject.get_lineage(t) for t in hostTaxidList]\n",
    "hostEnglishLineages = [[str(rank) for rank in dbObject.translate_to_names(lin)] \n",
    "                       for lin in hostTaxidLineages]\n",
    "\n",
    "for i in range(len(length_lists)):\n",
    "    lengthData.append({'length_list' : length_lists[i],\n",
    "                       'taxid' : taxidList[i],\n",
    "                       'hostTaxid' : hostTaxidList[i],\n",
    "                       'englishLineage' : englishLineages[i],\n",
    "                       'taxidLineage' : taxidLineages[i],\n",
    "                       'hostEnglishLineage' : hostEnglishLineages[i],\n",
    "                       'hostTaxidLineage' : hostTaxidLineages[i],\n",
    "                       'mean' : np.mean(length_lists[i]),\n",
    "                       'std' : np.std(length_lists[i]),\n",
    "                       'median' : np.median(length_lists[i]),\n",
    "                       'numProteins' : len(length_lists[i]),\n",
    "                       'genomeLength' : genomeLengths[i],\n",
    "                       'proteinNames': proteinNames[i],\n",
    "                       'folder' : folders[i]})\n",
    "                       \n",
    "lengthData = [virus for virus in lengthData if virus['genomeLength'] != -1 and virus['median'] != -1]\n",
    "\n",
    "for (i, virus) in enumerate(lengthData):\n",
    "    print(str(i) + '/' + str(len(lengthData)))\n",
    "    (totalCoding, totalOverlap) = getCodingNumberAndOverlapNumberFromGenomeFolder(virus['folder'])\n",
    "    virus['percentNonCoding'] = 100 * (virus['genomeLength'] - totalCoding) / virus['genomeLength']\n",
    "    virus['percentOverlap'] = 100 * totalOverlap / virus['genomeLength']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#the shared protein coding genes between dsDNA, ssDNA and RNA viruses\n",
    "protMatds= [v['proteinNames'] for v in lengthData if 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "protMatss= [v['proteinNames'] for v in lengthData if 'ssDNA viruses' in v['englishLineage']]\n",
    "protMatRNA= [v['proteinNames'] for v in lengthData if any('dsRNA viruses','ssRNA positive-strand viruses, no DNA stage','ssRNA negative-strand viruses' in v['hostEnglishLineage']\n",
    "\n",
    "\n",
    "ds = [i for vec in protMatds for i in vec]\n",
    "ss = [i for vec in protMatss for i in vec]\n",
    "#rna = [i for vec in protMatRNA for i in vec]\n",
    "\n",
    "#len(set(a) & set(b) & set(e))\n",
    "len(set(ds) & set(ss))\n",
    "#set(e) & set(a)\n",
    "#set(e) & set(a) & set(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">gi| Leuconostoc_phage_Lmd1_uid171974\n",
      "AAAAAAAAAQQCIIAEAAEEAEAAAAAAAAAAAAAAAAAA\n",
      ">gi| Leuconostoc_phage_phiLNTR3_uid253389\n",
      "AAAAAAAAAAQQCIIAEAAEEAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Leuconostoc_phage_phiLNTR3_uid253389\n",
      "AAAAAAAAAAQQCIIAEAAEEAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Leuconostoc_phage_phiLN12_uid253392\n",
      "AAAAAAAAQQCIIAEAAEEAAEAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Leuconostoc_phage_phiLNTR2_uid253393\n",
      "AAAAAAAAAAQQCIIAEAAEEAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Leuconostoc_phage_phiLN25_uid253397\n",
      "AAAAAAAAAAQAQCIIAEAAEEAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL111M01_uid219111\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL113M01_uid219107\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P105_uid177533\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P9_1_uid177529\n",
      "QQCAIIAAAAEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL112N00_uid219110\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P1_1_uid177537\n",
      "QQCAIIAAAAEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P105_uid177533\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL111M01_uid219111\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL111M01_uid219111\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL112N00_uid219110\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL111M01_uid219111\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P105_uid177533\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P100D_uid177534\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL111M01_uid219111\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_PHL112N00_uid219110\n",
      "AQCAQIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_ATCC29399B_C_uid177539\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_ATCC29399B_T_uid177538\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Erwinia_phage_ENT90_uid184166\n",
      "EEAEEEAAAEAAAAAAIQIAAIQAAAACAAAAAAAAAAAAAAAAAAAEAAAAEEAAAAEA\n",
      ">gi| Propionibacterium_phage_P101A_uid177531\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Propionibacterium_phage_P101A_uid177531\n",
      "QQCAIIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Salmonella_phage_FSL_SP_004_uid212714\n",
      "CQIIQIEAAAAEAEEEEEEEEAEAEAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_PsP3_uid14345\n",
      "CAQIIQAAAAAAAAAEEAAAEEAAAAEAAAAAAAAAAAAAAA\n",
      ">gi| Yersinia_phage_L_413C_uid14280\n",
      "CQIIQAAAAAAAAAEEAAAAEEAAAAEAAAAAAAAAAAAA\n",
      ">gi| Clostridium_phage_phiMMP04_uid179417\n",
      "QCAIIIAAEAAAEAAAAAEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Vibrio_phage_K139_uid14144\n",
      "AAAAAAAAAAAAAAAAAAACQIIQIAEAEEAAAAAEAAAEEAAA\n",
      ">gi| Vibrio_phage_kappa_uid28503\n",
      "AAAAAAAAAAAAAAAAAAAACQIIQIAEEEEAAAAAEAAAEEAAA\n",
      ">gi| Stenotrophomonas_phage_Smp131_uid240053\n",
      "AAAAACQIIQIEAAAAEEEEEAEEEEEEEAEEAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_P2_uid14035\n",
      "CQIIQAAAAAAAAAAAEEAAAAEEAAAAEAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_fiAA91_ss_uid226726\n",
      "IQIIQIEAAAAEEEEEEEEAEEEEEAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_Fels_2_uid32273\n",
      "AEEEAEEEAAAAAEEEAEEAAAAAAEAQIIQCAAAAAAAAAAAAAAA\n",
      ">gi| Aeromonas_phage_phiO18P_uid19769\n",
      "AAAAAAAAAAAAAAAAAACQQIIQIAEEEAAAAAAAEAAAEEAAA\n",
      ">gi| Salmonella_phage_RE_2010_uid181070\n",
      "AAAAAAAAAAAAAAIQIIQIEAAAEEAEEEEEEAAEAAEEEEEEAAA\n",
      ">gi| Mannheimia_phage_phiMHaA1_uid17103\n",
      "CQIIQIEAAAAEEAEAAEEEEEAEEEAEEAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Streptococcus_phage_2972_uid15254\n",
      "AQQQCIQIIAAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Mannheimia_phage_vB_MhM_1152AP_uid212715\n",
      "IQIIQIEAAAAEEEAAEEEEEAAAEEEAEEAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Idiomarinaceae_phage_1N2_2_uid266773\n",
      "QQAAACAIIIAACAAIEIEAEAAEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Streptococcus_phage_DT1_uid15124\n",
      "AAQQICAIQIEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Streptococcus_phage_Abc2_uid42791\n",
      "QQICAIQIEEEEEEEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Burkholderia_phage_ST79_uid206488\n",
      "AAAAAAAAAAAAAAAACQIIQIEAAAAAEAEEEEEAAEEEEEEAAAAAA\n",
      ">gi| Streptococcus_phage_ALQ13_2_uid42593\n",
      "QQCIQIIIAAAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_phiCTX_uid14415\n",
      "ACQIIQIAAAAAAAEEAEEEAAAEEAAAAEAAAAAAAAAAAAAAAAA\n",
      ">gi| Psychrobacter_phage_Psymv2_uid240594\n",
      "AAAAAAAAAAAAAAAAAAAAAAQQCAIAIIAAAAAAEEEAEEEEEAAAA\n",
      ">gi| Burkholderia_phage_phiE202_uid19163\n",
      "QIIQCAAAAAAAAAAAAAAAAAAAAAEEEEAAEEEAAAAEAAAAAAAI\n",
      ">gi| Enterobacteria_phage_P88_uid271780\n",
      "AAAAAAAAAAAAAAAAAAAAAACQIIQIEAAAEEEEEEEEEEAEEEEEEAAAA\n",
      ">gi| Enterococcus_phage_phiFL2A_uid42795\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQCAIAAQAAAAAAEAAAEEEEEAAAAAAA\n",
      ">gi| Lactobacillus_phage_Lc_Nu_uid16114\n",
      "QQACAIQIIEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Bacillus_phage_Cherry_uid15784\n",
      "QQCIIAIAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Escherichia_phage_HK75_uid76733\n",
      "QQCIIAIAAEEEAEEEEAEAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_phiJL_1_uid15156\n",
      "AAAAAAAAAAQQIAQIAQAIAEAAEAAEAAAAAAAAAAAAAAAAAA\n",
      ">gi| Burkholderia_phage_phiE12_2_uid19161\n",
      "QIIQCAAAAAAAAAAAAAAAAAAAAAAAAEEEEAAEEEEAAAAAAAAAAI\n",
      ">gi| Enterobacteria_phage_Mu_uid14105\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAQIAAAAEAAEAEEAAAEEEEAAA\n",
      ">gi| Idiomarinaceae_phage_Phi1M2_2_uid267261\n",
      "QQAIAIIAAAAAAAEEEAAAAAAAAAEAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_SfV_uid14162\n",
      "QQCIIAAAAAEAAEEEEEEEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Haemophilus_phage_SuMu_uid181066\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAQCAAAAAAAAAAEEAAEAEEAEAEEAAAAA\n",
      ">gi| Streptococcus_phage_5093_uid38299\n",
      "AAAAAAAAAAAAAQQIIAIAAIIEAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Escherichia_phage_D108_uid42515\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAQIAAAAEEAAEAEEAAAEEEEAAAA\n",
      ">gi| Bacillus_phage_Gamma_uid15783\n",
      "QQCIIAIAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_gh_1_uid14265\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAEIIIEEAAAAEAAQAQAA\n",
      ">gi| Streptococcus_phage_Sfi19_uid14045\n",
      "AAQQICIQIEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Arthrobacter_phage_vB_ArS_ArV2_uid230590\n",
      "AQCIAIAAAAEAAAEEEEAEAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Yersinia_pestis_phage_phiA1122_uid14332\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAIIIIEEAAAAEAQAAAAAA\n",
      ">gi| Burkholderia_phage_phi52237_uid15422\n",
      "QIIQCAAAAAAAAAAAAAAAAAAAEAAEEEAAEEEEAAEAAAAAEAI\n",
      ">gi| Clostridium_phage_phiCD6356_uid64557\n",
      "QQCIIAIAIAEAAAEEAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Bacillus_phage_Fah_uid16382\n",
      "QQCIIQIIAEAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_ATCC_8014_B1_uid184150\n",
      "QQIIQIIQAIAEAAEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_T3_uid14336\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEIIIIEEAAAAEAQAAQAAA\n",
      ">gi| Lactobacillus_phage_KC5a_uid16663\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQQCIIIAAAAAEEAEAAAAEAAAAAAAA\n",
      ">gi| Salmonella_phage_Vi06_uid64609\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIIEEAAAAEAQAAAAA\n",
      ">gi| Yersinia_phage_Berlin_uid18481\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAEEIIEEAAAAEAQAAQAAA\n",
      ">gi| Yersinia_phage_vB_YenP_AP5_uid266772\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAEIIIIEEAAAAAEAQAQA\n",
      ">gi| Ralstonia_phage_phiRSA1_uid19481\n",
      "AAAACQIIQIEAAAAEEAEEEEEEEAEEAEAEAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Bacillus_phage_phiCM3_uid240022\n",
      "AAAAAQQCAIIIAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Cellulophaga_phage_phi17_1_uid212962\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCIAAAAAAAAAIIIIEEAEAAEAAEAAAA\n",
      ">gi| Lactobacillus_phage_Lv_1_uid33535\n",
      "QQCIIIIEEEEEIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Cronobacter_phage_Dev2_uid239722\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIAEEIIIIEAQAQA\n",
      ">gi| Cellulophaga_phage_phi12_1_uid212966\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAQCIAAAAAAAAAIIIIEAEAEAAEAAEAAAA\n",
      ">gi| Lactobacillus_phage_phiAT3_uid14475\n",
      "AQQACAIAIIEEEEEAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Cellulophaga_phage_phi18_1_uid212959\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCIAAAAAAAAAAAIIIIEEAEAAEAAAAAAA\n",
      ">gi| Citrobacter_phage_CR44b_uid239730\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIIEAEAAAAAEAAQAAAQAAA\n",
      ">gi| Halomonas_phage_phiHAP_1_uid28763\n",
      "QQACIAAAEEEEEAAEAEAEEAEEEAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Erwinia_phage_vB_EamP_L1_uid181229\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIIEAEAEAAAAQAAAQAAA\n",
      ">gi| Rhodobacter_phage_RcapMu_uid76743\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAACIAAAAIAAAAAAAAEAAEEEAE\n",
      ">gi| Kluyvera_phage_Kvp1_uid32673\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIIEEAAAAEEAQAAQAA\n",
      ">gi| Enterococcus_phage_phiFL3A_uid42787\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQQQCAIQIAAAAEAAAAEAAAAAA\n",
      ">gi| Yersinia_phage_phiYeO3_12_uid14591\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEIIIIEEAAAAAEAQAAQAAA\n",
      ">gi| Citrobacter_phage_CR8_uid239727\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIIEEAAAAEAAQAAAQAA\n",
      ">gi| Enterobacteria_phage_K1F_uid15880\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAEIIIEEAAAAAAQAAA\n",
      ">gi| Enterobacteria_phage_HK97_uid14592\n",
      "QQCAIAIAAEAEAEAAAAAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Shigella_phage_SfIV_uid227000\n",
      "QQCIIAAAAEAAEEEEEEEAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Burkholderia_phage_KS9_uid39771\n",
      "AAAAAAAAAAAAAAAAAAAAQCQIAIAAEEEEEAEEEEAAAAAAAAAAAA\n",
      ">gi| Cronobacter_phage_phiES15_uid173212\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAQQAAAIAAAAAEAAEAEEAEAAAAA\n",
      ">gi| Mycobacterium_phage_Leo_uid209361\n",
      "AQCAAAIAAAAAEAAEEEAEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_Lrm1_uid30879\n",
      "QQACIIQIIIEAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQ\n",
      ">gi| Staphylococcus_phage_StB27_uid184157\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCIQIIIAIEEAEEAAEAAEAAAA\n",
      ">gi| Salmonella_phage_ST64B_uid14228\n",
      "QAAAIIAAAAAAAAAEEAAEIEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_prophage_Lj965_uid14351\n",
      "AAAAAAAAAAAAAAAAAAAAAAQQCAIQAIAAAAEAAEAAAAAAAA\n",
      ">gi| Pseudomonas_phage_Phi_S1_uid197298\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIIIEEAAAAEAQAQA\n",
      ">gi| Stenotrophomonas_phage_S1_uid32787\n",
      "AAQQAACIAAAAAAAEAEAEAEAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_phiPSA2_uid253322\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEIIIEEAAAAEAAQAQA\n",
      ">gi| Streptomyce_phage_TG1_uid177524\n",
      "QQCIIAAEAEEEAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterococcus_phage_IME_EF4_uid240030\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEAEIIIIEIICAQQ\n",
      ">gi| Acinetobacter_phage_Petty_uid240050\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIQIAEEAAAEAAQQAA\n",
      ">gi| Streptococcus_phage_Sfi21_uid14133\n",
      "QQICAIQIEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_HK022_uid14048\n",
      "QQIIIAIAAAAEAAEAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Bacillus_phage_WBeta_uid16329\n",
      "QQCIIQIAAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_StB20_uid184156\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQACIIQIAAEAAAEEEEAAAEAAAEAAAAAA\n",
      ">gi| Enterococcus_phage_EfaCPT1_uid266785\n",
      "QACIIAIIAAEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQ\n",
      ">gi| Pseudomonas_phage_phiIBB_PF7A_uid64561\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAEIIIIEEAAAAEAQAQAAAAAAAA\n",
      ">gi| Clostridium_phage_phiCD38_2_uid67249\n",
      "QQIIAAIAAAAAAAAEEEAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Klebsiella_phage_KP32_uid42779\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAEIIIEEAAAAEAAQAAQA\n",
      ">gi| Listeria_phage_LP_030_3_uid253385\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAQQCIAAAIIIEAAAEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Bacillus_phage_PBC1_uid167662\n",
      "QQAACIAIAAIIAIAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Vibrio_phage_X29_uid253324\n",
      "AAQAAAACAAIAAAAAEEAAAEAAEAAEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterococcus_phage_IME_EF3_uid239731\n",
      "AAAQQACIIIAIIIIEEEEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_Sha1_uid181084\n",
      "AAAAAAAAAAAAQQICAIQIIAEAAEAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Temperate_phage_phiNIH1_1_uid14145\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAQIIAAAIAIIIEAAEAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_DW2_uid253383\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCIAAIAIIAEEEEEEAAEAAAEAA\n",
      ">gi| Bacillus_phage_BtCS33_uid169233\n",
      "QQCAIIIAEAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Salmonella_phage_vB_SenS_Ent2_uid239739\n",
      "QQCAIIAAQIIAAIAIAAEAAAAAAAEAAAEEAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_vB_SepiS_phiIPLA7_uid171982\n",
      "QQCIIIAIIAAEAAEAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_phig1e_uid14315\n",
      "AAAAAAAAAAAAAAAIIAAAAIAIIIIIIIAIIQQAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Escherichia_phage_ep3_uid266771\n",
      "AAAAAAAAAAAAAAAAAAAAAAQQAIIAAAAAAEEEEEAAEEAEEEAAAAA\n",
      ">gi| Salmonella_phage_vB_SenS_Ent1_uid181993\n",
      "QQCAIIAAQIIAAACCAAEAAAAAAAEAAAEEAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Cyanophage_PP_uid227004\n",
      "AAAAAAAAAAAAAAAAAAAAAAAEEEAAEAEAIACAQAAAA\n",
      ">gi| Staphylococcus_phage_phiSauS_IPLA88_uid33001\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCAIAQIAQIAEAAEAAAEAAAAEAA\n",
      ">gi| Enterobacteria_phage_phiP27_uid14599\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQACIIAAAAAAEAEEAAAEEEAA\n",
      ">gi| Enterococcus_phage_IME_EFm1_uid253326\n",
      "AQQACIIIIIIEAEEEAEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_LF1_uid181083\n",
      "QQACIIQIIIEAAEAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_SA13_uid213078\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCIAIAIIAAEEAEEAAEAAAEEAAAA\n",
      ">gi| Salmonella_phage_SETP13_uid226727\n",
      "QQCAIEAAAAIIEAACAEAEAAAAAAEAAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Salmonella_phage_SETP7_uid226725\n",
      "QQCAIEAAAQIIEAACAAAEAAAAAEAAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Salmonella_phage_vB_SenS_Ent3_uid249091\n",
      "QQCAIIAAQIIAAIAIAAEAAAAAAAEAAAEEAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Burkholderia_phage_KL1_uid171970\n",
      "AAAAAAQCIAIAAAIEAEEAEAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_vB_Pae_Kakheti25_uid167052\n",
      "AAAAAQQCIQIAAAAAEEEEEAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_SA12_uid212955\n",
      "AAAAAAAAAAAAAAAAAAAAAAQQCIAAIAIAAAEAAEEAAAAAAAEAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_MPK6_uid227001\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIEEAAAAEEEQQAAAAA\n",
      ">gi| Pseudomonas_phage_vB_Pae_TbilisiM32_uid167051\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACQIEEAAAIEEEAAAAAAAA\n",
      ">gi| Edwardsiella_phage_eiAU_uid240025\n",
      "EEEEEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAIAAAAAAEAAA\n",
      ">gi| Acinetobacter_phage_IME_AB3_uid240027\n",
      "AAAAAAAAAAAAAAAAAAAAAAEAAIAEEEEEEIAAIQICQQAAAAAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_vB_PaeS_SCH_Ab26_uid253340\n",
      "AAAAQQCIQIAAAAAEEEEEAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Pseudomonas_phage_phikF77_uid36373\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIEEAAAAEEEAQAAAAAA\n",
      ">gi| Phormidium_phage_Pf_WMP3_uid19801\n",
      "AAAAAAAAAAAAAAAAAAAAAAAEAAAAAEEIQCAQAAAAA\n",
      ">gi| Staphylococcus_phage_phiRS7_uid229900\n",
      "QQACIIAQIIAEAEAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Lactobacillus_phage_A2_uid14602\n",
      "QQCAIQIIIEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQ\n",
      ">gi| Enterobacteria_phage_JL1_uid179426\n",
      "QAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAEEEEEEAAEAAAAAAAAAIA\n",
      ">gi| Pseudomonas_phage_LUZ19_uid28741\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIEEAAAAEEEQQAAAAAAA\n",
      ">gi| Staphylococcus_phage_vB_SepiS_phiIPLA5_uid171981\n",
      "QQCIAAIAQIAAEEEEEACAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_phiBU01_uid271781\n",
      "AAAAAAAAAAEEAEAAAIIACQAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_SP6_uid14291\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIEEAAAEAQQAAAAAAAEAAA\n",
      ">gi| Staphylococcus_phage_TEM123_uid167573\n",
      "AAAAAAAAAAAAEAAAAAEEAAAAEAAIIQICQQAAAAAAAAA\n",
      ">gi| Phage_Paz_uid230593\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIQIAEEAAAEEEEAQQAAAAAA\n",
      ">gi| Phage_Prado_uid230587\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIQIAEEAAAEEEEAQQAAAAA\n",
      ">gi| Shigella_phage_EP23_uid80919\n",
      "AQAIAAAAAAAAAEAEAEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Synechococcus_phage_S_CBP4_uid266792\n",
      "AAAAAAAAAAAAAAAAAAAAIIAAAIAAEEAAAEAAAAAAAAAQAA\n",
      ">gi| Lelliottia_phage_phD2B_uid266769\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIEEAAAEAQQAQAAAAAA\n",
      ">gi| Bacillus_phage_phIS3501_uid181213\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQCAIQIAAEAEEAAAAAAAAA\n",
      ">gi| Vibrio_phage_vB_VchM_138_uid181217\n",
      "AAAAAAAAAAAAAAAAAAAAQQCAAAAIAAAAIAIAAAAAAAAAAAAAAAAAAEAAAEAEAEEAAAA\n",
      ">gi| Enterobacteria_phage_SSL_2009a_uid34919\n",
      "QAIAAAAAAAAAEAEAAEEEEEEAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Salmonella_phage_E1_uid29079\n",
      "AQACIAAAEEAAAAAIIAAAAAAAAAAAEAAAEAAAEAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_StauST398_1_uid206473\n",
      "AAAAAAAAEEAAAAAAAEEEAEAEIIAIAICQQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Staphylococcus_phage_SMSAP5_uid181240\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAEEEAAEEAAAIAIQQ\n",
      ">gi| Streptococcus_phage_TP_J34_uid188154\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAQQCIQAIAAAAAEAAEEEAAAAAAAAAAAAAAA\n",
      ">gi| Mycobacterium_phage_BTCU_1_uid209077\n",
      "EEAAAEAAQCIQIIAAAAEEEAAAAEAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_RTP_uid16178\n",
      "AAAAAAAAAAAAAAAAAAQAQACIIAAAAAAAEAAAEEEAEEEAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_cdtI_uid19737\n",
      "QQACQAAEEEEEEEEEEEAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Iodobacteriophage_phiPLPE_uid30965\n",
      "AAAAAAAAAAAAAAAAQQCAAAAAIAIAIAAAAAAEAEAAAAAAAAAAAAAAAAEAAAEAEAEAAAAAAAAAAAAAEAAAAAAA\n",
      ">gi| Enterobacteria_phage_lambda_uid14204\n",
      "QQIIIIIIQIEEEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_T1_uid14496\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAQQCAIAAAAAAAAEAAEEEEEEAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Cronobacter_phage_ESP2949_1_uid181234\n",
      "EEEEAEAAAAAAAAAIICQQAAAAAAAAAAAAAAAAAAEAAAAEE\n",
      ">gi| Burkholderia_phage_phiE125_uid14330\n",
      "QQACIIAAAAAEEAEEAEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Puniceispirillum_phage_HMO_2011_uid213071\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAEEEAEEEAAEEEEEAAAAAIAAACQAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Burkholderia_phage_BcepNazgul_uid14305\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEEEEEEAAAAAIAIACIQQAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_BP_4795_uid14287\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQQIACAAIAAEEAEEEEEEAAEAAAAAAAAAA\n",
      ">gi| Burkholderia_phage_AH2_uid171971\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEEAAEEEAEAAAIAICIQQAAAAAAAAAA\n",
      ">gi| Salmonella_phage_FSL_SP_088_uid212711\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEAEAAEEEAAAAAIAICIQQAAAAAAAAAA\n",
      ">gi| Salmonella_phage_FSL_SP_030_uid212718\n",
      "AAAAAAAAAAQQICIAIAAAAAEEEAAEAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacter_phage_Enc34_uid181238\n",
      "AAAAAAAAAQQICIAIAAAAAEEEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Stx2_converting_phage_1717_uid32213\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAICCAIAEEAEEEAEEEAEAAAAA\n",
      ">gi| Acinetobacter_phage_133_uid64541\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAAEEEAEEEECCIEQQEEIIIIIIAIIAAAAAAAAEEEEEEEEAAAAAAAAAIAAAAAAAAAAAAAAAAAAEAEAAAAAAAAAAAAAAAAEEEEAAAAAAAAA\n",
      ">gi| Salmonella_phage_S16_uid191122\n",
      "AAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAAEEEEEEECCCEQQEAECIIAIIIIIAAAAAIIAAAAAAAEEEEEEEEAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAEAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Aeromonas_phage_Aes012_uid195532\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIAAEEAEEEEEEECCCEQQEEIIIIIIIAAAAAAAIAAAAEEEEEEEEAAAAAAIAAAAAAAAAEAAAAAAAAAAAAAEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_3S_uid266636\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAAIEEAAAEEEAAAEEEECCCEQQEECIIIIIAAAIAAAAAIAIAAAAEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Aeromonas_phage_phiAS4_uid59727\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEEEEEAAAAAIIAAAAAAIAAIIIIIIIAEEEQQEIIIIAEEEEEEEEEEEAEEAIAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_RB49_uid14301\n",
      "AAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAAIEEAAEEEAAAAEEEEACCEQQEECIIIIIAIAAAAIAIAAAAEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Acinetobacter_phage_Acj61_uid60117\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAAIEEAAEEEAEEEECCIEQQEAEIIIIIIIIIAAAAEEEEEEEEAAAAAAAAIAAAAAAAEAAAAAAAAAAAAAEEEEAAAAAAAA\n",
      ">gi| Enterobacteria_phage_Phi1_uid20789\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAAAIEEAAEEEAAAEEEECCACEQQEEIIIIIIAAIAAAAIAIAAAAEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Shigella_phage_Shfl2_uid66347\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAIIIAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAEAEAIEEAAEEEEEEECCIEQQEEIIIIIIIAAAIIAAAAAAEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_JSE_uid38263\n",
      "AAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAAAIEEAAEEEAAAEEEEACCAEQQEECIIIIIAAIAAAAIAIAAAAEEEEEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Escherichia_phage_wV7_uid181232\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAIAEAEAIEEAAEEEEEEECCCEQQEEIIIIIIIAAAIIAAAAAEEEEEEEEAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_ime09_uid181233\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAEAEAIEEAAEEEEEEECCCEQQEEIIIIIIIAAAIIAAAAAEEEEEEEEAAAAAAAAAAAAAAIEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Serratia_phage_PS2_uid248461\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAIEEAAEEEEEEECCCEQQEECIIIIIIAAAAAAAAAAIIAAAAEEEEEEEEAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEEAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_vB_EcoM_ACG_C40_uid179415\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAAAEEEEEEECCCEQQEEIIIIIIIAAAAIIIAAAAAAAEEEEEEEEAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_RB69_uid15141\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAEAEAIEEAAEEEEEEEACCEQQEECIIIIIAAIAAAIIAAAAAEEEEEEEEAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_Bp7_uid181212\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAIAAAAAAAAAAAAAAAEEEEEEEEAAAAAAIIAAIIAIIIIIIIAEAEQQECCCEEEEEEEAEEIAEAEAI\n",
      ">gi| Escherichia_phage_e11_2_uid248462\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAEAEAAIEEAAEEEEEEEACCEQQEEIIIIIIIAAAIIAAAAAAEEAEEEEEAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Yersinia_phage_phiR1_RT_uid184143\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIAAAAAAAAAAAAAAAAAAAAAAAEAIAEEAAEEEEEAEAIIAEQQAQEEIIIIIIAIAAAIAAAIAAAAEEEEEEEEAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_T4_uid14044\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIIAAAAAAAAAAAAAIAEAEAIEEAAAAAAEEEEEEEACCEQQQQQEECIIIIIIAIAAAIIAAAAAEEAEEEEEEAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_vB_EcoM_VR7_uid61099\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAIAAAAAAAAAAAAAAAAAAAAAAAIAAIAAEAEAAIEEAAEEEEEEEACCEQQEAECIIIIIAIIAAAAAIIAAAAEEEEEEEEAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Acinetobacter_phage_Acj9_uid60121\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAIEEAAEEEEEEECCIAEQAQEEIIIIIIIIIAAAAAAAEEEAEEEEAAAAAAAAIAAAAAAAAAAAEAAAAAAAAAAAAAAAAEEEEAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_JS98_uid27983\n",
      "AAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAAEEEEEEEACCEQQEECIIIIIAIIAAIIAAAAEEEEEEEEAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Shigella_phage_SP18_uid56019\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAIIAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAEEEEAEEECCACEQQEEIIIIIIAIAAAIIAAAAAAEEEEEEEEAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAEEAEAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_JS10_uid38265\n",
      "AAAAAAAAAAAAAAAAAAAAAAAIIAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAAAEEEEEEEACCEQQEECIIIIIAIIAAAEIIIAAAAEEEEEEEEAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAA\n",
      ">gi| Enterobacteria_phage_IME08_uid50177\n",
      "AAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEEAAEEEEEEECCCEQQEEIIIIIIAAIIEAAIIIAAAAEEEEEEEEAAAAAAAAAAAAAAAIAAAAAAAAAAAAAEAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAA\n",
      ">gi| Klebsiella_phage_KP27_uid185314\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAIAAIEEAEEEEEEECIIEQQEEIIIIIIAIAAAAIAAIIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAA\n",
      ">gi| Klebsiella_phage_KP15_uid47333\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAAIEEAEEEEEEEACCEQQEECIIIIIAIAAAAIAAIIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAA\n",
      ">gi| Enterobacteria_phage_RB16_uid51699\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAIAAAIEEAAEEEEEEEACCAEQQEECIIIIIAIAAAAIAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAA\n",
      ">gi| T4likevirus_uid266638\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEAAAEEEEEEECCCEQQEAEIIIIIIIAAAIIAAAAAAAAEEEEEEAAAAAAAAAAAAAAAIEEAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAEAIEAAAEEEEEEECCCEQQEEIIIIIIIAAAIIAAAAAAAEEEEEEAAAAAAAAAAAAAAAIEEAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAEEEEEAAAAAAAAAAAAAAAAAAAAAAAAAA\n"
     ]
    }
   ],
   "source": [
    "#want to produce gene order sequences\n",
    "from collections import Counter\n",
    "\n",
    "\"\"\"\n",
    "protMat= [v['proteinNames'] for v in lengthData]\n",
    "genomeLen= [v['genomeLength'] for v in lengthData]\n",
    "virusName = [v['folder'] for v in lengthData]\n",
    "virusLineage = [v['englishLineage'] for v in lengthData]\n",
    "hostLineage = [v['hostEnglishLineage'] for v in lengthData]\n",
    "\"\"\"\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "genomeLen= [v['genomeLength'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage']  and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "virusName = [v['folder'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage']  and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "virusLineage = [v['englishLineage'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage']  and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "hostLineage = [v['hostEnglishLineage'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage']  and 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "\n",
    "#ordering the name list and the protMat to be sorted based on the genome lengths. You can used virusName instead to order based on folder names. \n",
    "protMat = [protMat for (genomeLen, protMat) in sorted (zip(genomeLen, protMat), key=lambda pair:pair[0])]\n",
    "virusName = [virusName for (genomeLen, virusName) in sorted (zip(genomeLen, virusName), key=lambda pair:pair[0])]\n",
    "virusLineage = [virusLineage for (genomeLen, virusLineage) in sorted (zip(genomeLen, virusLineage), key=lambda pair:pair[0])]\n",
    "hostLineage = [hostLineage for (genomeLen, hostLineage) in sorted (zip(genomeLen, hostLineage), key=lambda pair:pair[0])]\n",
    "\n",
    "capsid = ['coat', 'capsid', 'head']\n",
    "matrix = ['matrix', 'tegument']\n",
    "#env = ['envelope'] \n",
    "portal = ['portal','neck', 'head-tail connector', 'head-tail joining'] \n",
    "tail= [ 'tail','baseplate', 'tape measure']\n",
    "packing = ['terminase', 'TerL','TerS', 'scaffold protein', 'prohead protease', 'DNA packaging']\n",
    "lysis = ['holin', 'endolysin', 'lysozyme','lysis', 'lysin', 'N-acetylmuramoyl-L-alanine amidase', 'cell wall hydrolase']\n",
    "nuclease = ['nuclease']\n",
    "integrase =['integrase']\n",
    "polymerase =['polymerase']\n",
    "helicase =['helicase']\n",
    "replication =['Rep protein','rep protein','replicase', 'replication']\n",
    "regulationModification = ['transcriptional enhancer', 'transcriptional regulator', 'expression regulator','promoter', 'antirepressor','anti-repressor', 'thioredoxin','DNA methylase','methyltransferase','ADP-ribosylase','ADP-ribosyltransferase']\n",
    "metabolismSynth = ['thymidine kinase', 'deaminase', 'thymidylate synthase','dihydrofolate reductase', 'ribonucleotide reductase', 'dNMP kinase','dCMP hydroxymethylase','deoxynucleoside monophosphate kinase','amidase','phosphate starvation-inducible protein' ]\n",
    "entryInjection = ['movement protein','injection']\n",
    "trna = ['tRNA']\n",
    "#determing and storing the color-code information for each viral gene\n",
    "colormat=[]\n",
    "virusNameFiltered =[]\n",
    "for vec in protMat:\n",
    "    colorvec=[]\n",
    "    for i in vec:          \n",
    "        if any(substring in i for substring in capsid):\n",
    "            colorvec.append('I')\n",
    "        elif any(substring in i for substring in portal):\n",
    "            colorvec.append('C')         \n",
    "        elif any(substring in i for substring in tail):\n",
    "            colorvec.append('E')          \n",
    "        elif any(substring in i for substring in packing):\n",
    "            colorvec.append('Q')         \n",
    "        else:       \n",
    "            colorvec.append('A')\n",
    "    #colormat.append(colorvec)\n",
    "                  \n",
    "    if colorvec.count('A')/len(colorvec) < 0.85 and len(colorvec)>39:                \n",
    "        colormat.append(colorvec)\n",
    "        #getting the index of the vectors in protMat that passed through the filter. Then I use the index to get the virus names from the virusName matrix \n",
    "        index = protMat.index(vec)\n",
    "        virusNameFiltered.append(virusName[index])\n",
    "        print('>gi|',virusName[index])#,'_',virusLineage[index][4],'_',hostLineage[index][3:6])\n",
    "        print(str(colorvec).replace(\",\",\"\").replace(\"'\",\"\").replace(\" \", \"\").strip('[]'))\n",
    "        #note T4likevirus_uid266638 virus genome has been copied twice, which is why we have taken this virus out in our plots.\n",
    "        #originally we thought it's a genome with its gene order pattern repeated twice, until we looked further at the paper wherein the phage was first sequenced. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#defining a function that gives us both the gene order sequence, as well as the COG labels of the proteins that the genes encode. \n",
    "from collections import Counter\n",
    "def printAminoAcidSeq(a):\n",
    "    protMat= [v['proteinNames'] for v in lengthData if a in v['folder']]\n",
    "    print(protMat)\n",
    "    genomeLen= [v['genomeLength'] for v in lengthData if a in v['folder']]\n",
    "    virusName = [v['folder'] for v in lengthData if a in v['folder']]\n",
    "    virusLineage = [v['englishLineage'] for v in lengthData if a in v['folder']]\n",
    "    hostLineage = [v['hostEnglishLineage'] for v in lengthData if a in v['folder']]\n",
    "    \n",
    "    capsid = ['coat', 'capsid', 'head']\n",
    "    matrix = ['matrix', 'tegument']\n",
    "    neck = ['neck', 'head-tail connector', 'head-tail joining']\n",
    "    env = ['envelope'] \n",
    "    portal = ['portal'] \n",
    "    tail= [ 'tail','baseplate', 'tape measure']\n",
    "    packing = ['terminase', 'TerL','TerS', 'scaffold protein', 'prohead protease', 'DNA packaging']\n",
    "    lysis = ['holin', 'endolysin', 'lysozyme','lysis', 'lysin', 'N-acetylmuramoyl-L-alanine amidase', 'cell wall hydrolase']\n",
    "    nuclease = ['nuclease']\n",
    "    integrase =['integrase']\n",
    "    polymerase =['polymerase']\n",
    "    helicase =['helicase']\n",
    "    replicase =['Rep protein','rep protein','replicase', 'replication']\n",
    "    regulationModification = ['transcriptional enhancer', 'transcriptional regulator', 'expression regulator','promoter', 'antirepressor','anti-repressor', 'thioredoxin','DNA methylase','methyltransferase','ADP-ribosylase','ADP-ribosyltransferase']\n",
    "    metabolismSynth = ['thymidine kinase', 'deaminase', 'thymidylate synthase','dihydrofolate reductase', 'ribonucleotide reductase', 'dNMP kinase','dCMP hydroxymethylase','deoxynucleoside monophosphate kinase','amidase','phosphate starvation-inducible protein' ]\n",
    "    entryInjection = ['movement protein','injection']\n",
    "\n",
    "    #determing and storing the color-code information for each viral protein\n",
    "    for vec in protMat:\n",
    "        colorvec=[]\n",
    "        for i in vec:          \n",
    "            # make sure the search is not case sensitive        \n",
    "            if any(substring in i for substring in capsid):\n",
    "                colorvec.append('I')\n",
    "            elif any(substring in i for substring in matrix):\n",
    "                colorvec.append('R')\n",
    "            elif any(substring in i for substring in neck):\n",
    "                colorvec.append('N')\n",
    "            elif any(substring in i for substring in env):\n",
    "                colorvec.append('D')\n",
    "            elif any(substring in i for substring in portal):\n",
    "                colorvec.append('C')         \n",
    "            elif any(substring in i for substring in tail):\n",
    "                colorvec.append('E')\n",
    "            elif any(substring in i for substring in polymerase):\n",
    "                colorvec.append('T')\n",
    "            elif any(substring in i for substring in helicase):\n",
    "                colorvec.append('L')\n",
    "            elif any(substring in i for substring in replicase):\n",
    "                colorvec.append('S')\n",
    "            elif any(substring in i for substring in integrase):\n",
    "                colorvec.append('K')            \n",
    "            elif any(substring in i for substring in nuclease):\n",
    "                colorvec.append('P')            \n",
    "            elif any(substring in i for substring in packing):\n",
    "                colorvec.append('Q')\n",
    "            elif any(substring in i for substring in lysis):\n",
    "                colorvec.append('G')           \n",
    "            elif any(substring in i for substring in regulationModification):\n",
    "                colorvec.append('Y')\n",
    "            elif any(substring in i for substring in metabolismSynth):\n",
    "                colorvec.append('burlywood')       \n",
    "            elif any(substring in i for substring in entryInjection):\n",
    "                colorvec.append('H')           \n",
    "            else:       \n",
    "                colorvec.append('A')    \n",
    "\n",
    "        print('>gi|',virusName[0],'_',virusLineage[0][4],'_',hostLineage[0][3])\n",
    "        print(str(colorvec).replace(\",\",\"\").replace(\"'\",\"\").replace(\" \", \"\").strip('[]'))\n",
    "\n",
    "    print(len(colorvec)) \n",
    "    return()\n",
    "\n",
    "#getting this information for the podoviruses in pattern C. \n",
    "virusList =['Pseudomonas_phage_gh_1_uid14265','Yersinia_pestis_phage_phiA1122_uid14332','Enterobacteria_phage_T3_uid14336','Salmonella_phage_Vi06_uid64609','Yersinia_phage_Berlin_uid18481','Yersinia_phage_vB_YenP_AP5_uid266772','Synechococcus_phage_S_CBP4_uid266792','Cronobacter_phage_Dev2_uid239722','Citrobacter_phage_CR44b_uid239730','Erwinia_phage_vB_EamP_L1_uid181229','Kluyvera_phage_Kvp1_uid32673','Yersinia_phage_phiYeO3_12_uid14591','Pseudomonas_phage_Phi_S1_uid197298','Pseudomonas_phage_phiPSA2_uid253322','Citrobacter_phage_CR8_uid239727','Pseudomonas_phage_phiIBB_PF7A_uid64561','Klebsiella_phage_KP32_uid42779','Enterobacteria_phage_K1F_uid15880','Pseudomonas_phage_phikF77_uid36373','Pseudomonas_phage_MPK6_uid227001','Pseudomonas_phage_LUZ19_uid28741','Lelliottia_phage_phD2B_uid266769','Phage_Paz_uid230593','Phage_Prado_uid230587','Acinetobacter_phage_Petty_uid240050']\n",
    "for a in virusList:\n",
    "    printAminoAcidSeq(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/Users/Gita/anaconda/lib/python3.4/site-packages/matplotlib/backend_bases.py\u001b[0m in \u001b[0;36menter_notify_event\u001b[0;34m(self, guiEvent, xy)\u001b[0m\n\u001b[1;32m   1966\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_lastx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_lasty\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1967\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1968\u001b[0;31m     \u001b[0;32mdef\u001b[0m \u001b[0menter_notify_event\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mguiEvent\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mxy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1969\u001b[0m         \"\"\"\n\u001b[1;32m   1970\u001b[0m         \u001b[0mBackend\u001b[0m \u001b[0mderived\u001b[0m \u001b[0mclasses\u001b[0m \u001b[0mshould\u001b[0m \u001b[0mcall\u001b[0m \u001b[0mthis\u001b[0m \u001b[0mfunction\u001b[0m \u001b[0mwhen\u001b[0m \u001b[0mentering\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    },
    {
     "data": {
      "text/plain": [
       "(['structurald',\n",
       "  'rmd',\n",
       "  'integrased',\n",
       "  'helicased',\n",
       "  'replicased',\n",
       "  'ligased',\n",
       "  'polyd',\n",
       "  'nucleard',\n",
       "  'portald',\n",
       "  'recombd',\n",
       "  'lysisd',\n",
       "  'metd',\n",
       "  'matrixd',\n",
       "  'otherd',\n",
       "  'translationd',\n",
       "  'taild',\n",
       "  'polymerased',\n",
       "  'entryd',\n",
       "  'hypd',\n",
       "  'packingd',\n",
       "  'capsidd',\n",
       "  'envd',\n",
       "  'nucleased'],\n",
       " [[81,\n",
       "   0,\n",
       "   0,\n",
       "   7,\n",
       "   127,\n",
       "   0,\n",
       "   343,\n",
       "   0,\n",
       "   0,\n",
       "   0,\n",
       "   8,\n",
       "   2,\n",
       "   61,\n",
       "   1244,\n",
       "   0,\n",
       "   0,\n",
       "   225,\n",
       "   118,\n",
       "   163,\n",
       "   0,\n",
       "   360,\n",
       "   170,\n",
       "   0],\n",
       "  [110,\n",
       "   48,\n",
       "   2,\n",
       "   0,\n",
       "   353,\n",
       "   0,\n",
       "   0,\n",
       "   50,\n",
       "   0,\n",
       "   0,\n",
       "   7,\n",
       "   0,\n",
       "   0,\n",
       "   1105,\n",
       "   0,\n",
       "   0,\n",
       "   0,\n",
       "   80,\n",
       "   302,\n",
       "   7,\n",
       "   408,\n",
       "   17,\n",
       "   0],\n",
       "  [1137,\n",
       "   1629,\n",
       "   245,\n",
       "   1120,\n",
       "   693,\n",
       "   352,\n",
       "   1,\n",
       "   0,\n",
       "   617,\n",
       "   263,\n",
       "   1604,\n",
       "   1668,\n",
       "   607,\n",
       "   28880,\n",
       "   143,\n",
       "   3939,\n",
       "   1365,\n",
       "   35,\n",
       "   64428,\n",
       "   1262,\n",
       "   2581,\n",
       "   1522,\n",
       "   1947],\n",
       "  [42,\n",
       "   412,\n",
       "   4,\n",
       "   310,\n",
       "   74,\n",
       "   38,\n",
       "   0,\n",
       "   0,\n",
       "   1,\n",
       "   5,\n",
       "   106,\n",
       "   541,\n",
       "   606,\n",
       "   9876,\n",
       "   53,\n",
       "   4,\n",
       "   363,\n",
       "   11,\n",
       "   11139,\n",
       "   152,\n",
       "   581,\n",
       "   1178,\n",
       "   218],\n",
       "  [1062,\n",
       "   1190,\n",
       "   228,\n",
       "   794,\n",
       "   618,\n",
       "   313,\n",
       "   1,\n",
       "   0,\n",
       "   613,\n",
       "   256,\n",
       "   1496,\n",
       "   1111,\n",
       "   1,\n",
       "   18782,\n",
       "   90,\n",
       "   3925,\n",
       "   997,\n",
       "   24,\n",
       "   51526,\n",
       "   1104,\n",
       "   1982,\n",
       "   322,\n",
       "   1721],\n",
       "  [374,\n",
       "   454,\n",
       "   142,\n",
       "   223,\n",
       "   234,\n",
       "   33,\n",
       "   0,\n",
       "   0,\n",
       "   273,\n",
       "   75,\n",
       "   652,\n",
       "   260,\n",
       "   0,\n",
       "   9589,\n",
       "   13,\n",
       "   1252,\n",
       "   225,\n",
       "   2,\n",
       "   16642,\n",
       "   599,\n",
       "   679,\n",
       "   45,\n",
       "   486],\n",
       "  [444,\n",
       "   578,\n",
       "   39,\n",
       "   416,\n",
       "   283,\n",
       "   204,\n",
       "   0,\n",
       "   0,\n",
       "   263,\n",
       "   149,\n",
       "   492,\n",
       "   694,\n",
       "   1,\n",
       "   6216,\n",
       "   70,\n",
       "   2242,\n",
       "   447,\n",
       "   3,\n",
       "   25483,\n",
       "   299,\n",
       "   932,\n",
       "   237,\n",
       "   889],\n",
       "  [225,\n",
       "   91,\n",
       "   37,\n",
       "   120,\n",
       "   84,\n",
       "   71,\n",
       "   1,\n",
       "   0,\n",
       "   61,\n",
       "   23,\n",
       "   302,\n",
       "   110,\n",
       "   0,\n",
       "   2550,\n",
       "   2,\n",
       "   364,\n",
       "   291,\n",
       "   19,\n",
       "   7175,\n",
       "   168,\n",
       "   313,\n",
       "   32,\n",
       "   289],\n",
       "  [0,\n",
       "   29,\n",
       "   0,\n",
       "   29,\n",
       "   2,\n",
       "   7,\n",
       "   0,\n",
       "   0,\n",
       "   0,\n",
       "   0,\n",
       "   3,\n",
       "   31,\n",
       "   0,\n",
       "   1860,\n",
       "   1,\n",
       "   0,\n",
       "   15,\n",
       "   7,\n",
       "   931,\n",
       "   0,\n",
       "   29,\n",
       "   18,\n",
       "   15],\n",
       "  [2,\n",
       "   122,\n",
       "   0,\n",
       "   31,\n",
       "   22,\n",
       "   13,\n",
       "   0,\n",
       "   0,\n",
       "   0,\n",
       "   1,\n",
       "   3,\n",
       "   79,\n",
       "   3,\n",
       "   1510,\n",
       "   2,\n",
       "   0,\n",
       "   148,\n",
       "   0,\n",
       "   689,\n",
       "   9,\n",
       "   4,\n",
       "   112,\n",
       "   7],\n",
       "  [1,\n",
       "   73,\n",
       "   0,\n",
       "   150,\n",
       "   10,\n",
       "   1,\n",
       "   0,\n",
       "   0,\n",
       "   1,\n",
       "   0,\n",
       "   1,\n",
       "   157,\n",
       "   601,\n",
       "   2726,\n",
       "   2,\n",
       "   0,\n",
       "   76,\n",
       "   2,\n",
       "   323,\n",
       "   142,\n",
       "   371,\n",
       "   970,\n",
       "   40],\n",
       "  [1, 0, 0, 0, 6, 0, 0, 0, 0, 0, 0, 0, 0, 435, 0, 0, 0, 0, 19, 0, 35, 0, 0],\n",
       "  [0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 865, 0, 0, 26, 0, 42, 0, 81, 42, 0],\n",
       "  [32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 189, 0, 0, 0, 0, 0, 0, 27, 0, 0]])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# creating two functions for plotting the functional gene categories of each viral group\n",
    "def proteinBar (protMat):\n",
    "    protDict={} \n",
    "    capsid = ['coat', 'capsid', 'head', 'CP', 'Cap protein', 'cap protein']\n",
    "    matrix = ['matrix', 'tegument']\n",
    "    env =['envelope protein', 'envelope glycoprotein', 'membrane protein', 'membrane glycoprotein', 'glycoprotein', 'spike protein']\n",
    "    portal = ['portal','neck', 'head-tail connector', 'head-tail joining'] \n",
    "    tail= [ 'tail','baseplate', 'tape measure']\n",
    "    packing = ['terminase', 'TerL','TerS', 'scaffold protein', 'prohead protease', 'DNA packaging']\n",
    "    lysis = ['holin', 'endolysin', 'lysozyme','lysis', 'lysin', 'N-acetylmuramoyl-L-alanine amidase', 'cell wall hydrolase','hydrolase']\n",
    "    nuclease = ['nuclease', 'DNAse', 'DNase','RNAse', 'RNase']\n",
    "    integrase =['integrase']\n",
    "    polymerase =['polymerase']\n",
    "    helicase =['helicase']\n",
    "    replicase =['topisomerase', 'primase','Rep protein','rep protein','replicase', 'replication', 'holliday junction resolvase','clamp loader', 'sliding clamp']\n",
    "    regulationModification = ['transcriptional enhancer', 'transcriptional repressor','transcriptional activation protein', 'transcription regulator', 'transcription factor', 'transcriptional regulator','transcriptional activator','expression regulator','promoter', 'antirepressor','anti-repressor', 'thioredoxin','DNA methylase','methyltransferase','ADP-ribosylase','ADP-ribosyltransferase', 'sigma factor', 'phosphatase', 'apoptosis regulator'\n",
    "                             'growth factor', 'ubiquitin']\n",
    "    metabolismSynth = ['deaminase', 'thymidylate synthase','dihydrofolate reductase', 'ribonucleotide reductase','ribonucleotide-reductase', 'ribonucleoside-diphosphate reductase','anaerobic NTP reductase','oxidoreductase', 'kinase','dUTPase', 'ATPase','dCMP hydroxymethylase','amidase','phosphate starvation-inducible protein', 'deoxyuridine triphosphatase', 'dCTP pyrophosphatase', 'uracil DNA glycosylase' ]\n",
    "    entryInjection = ['movement protein','injection', 'fusion protein']\n",
    "    hyp=['hypothetical protein', 'putative uncharacterised protein']\n",
    "    ligase =['ligase']\n",
    "    structural=['structural protein']\n",
    "    recomb = ['recombination','recombinase']\n",
    "    translation = ['translation initiation', 'translation elongation', 'tRNA', 'translation factor']\n",
    "    nuclear=['nuclear shuttle protein', 'nuclear transporter']\n",
    "    poly=['polyprotein', 'readthrough protein', 'read-through protein']\n",
    "\n",
    "    capsidVec=[]\n",
    "    matrixVec=[]\n",
    "    envVec=[]\n",
    "    portalVec=[]        \n",
    "    tailVec=[]\n",
    "    polymeraseVec=[]\n",
    "    helicaseVec=[]\n",
    "    replicaseVec=[]\n",
    "    integraseVec=[]       \n",
    "    nucleaseVec=[]           \n",
    "    packingVec=[]\n",
    "    lysisVec=[]    \n",
    "    rmVec=[]\n",
    "    translationVec=[]        \n",
    "    metVec=[]      \n",
    "    entryVec=[]  \n",
    "    hypVec=[]\n",
    "    otherVec=[]\n",
    "    ligaseVec =[]\n",
    "    structuralVec=[]\n",
    "    recombVec=[]\n",
    "    nuclearVec=[]\n",
    "    polyVec=[]\n",
    "\n",
    "    for vec in protMat:\n",
    "        for i in vec:          \n",
    "            if any(substring in i for substring in capsid):\n",
    "                capsidVec.append(i)\n",
    "            elif any(substring in i for substring in matrix):\n",
    "                matrixVec.append(i)\n",
    "            elif any(substring in i for substring in env):\n",
    "                envVec.append(i)  \n",
    "            elif any(substring in i for substring in portal):\n",
    "                portalVec.append(i)        \n",
    "            elif any(substring in i for substring in tail):\n",
    "                tailVec.append(i)\n",
    "            elif any(substring in i for substring in polymerase):\n",
    "                polymeraseVec.append(i)\n",
    "            elif any(substring in i for substring in helicase):\n",
    "                helicaseVec.append(i)\n",
    "            elif any(substring in i for substring in replicase):\n",
    "                replicaseVec.append(i)\n",
    "            elif any(substring in i for substring in integrase):\n",
    "                integraseVec.append(i)       \n",
    "            elif any(substring in i for substring in nuclease):\n",
    "                nucleaseVec.append(i)           \n",
    "            elif any(substring in i for substring in packing):\n",
    "                packingVec.append(i)\n",
    "            elif any(substring in i for substring in lysis):\n",
    "                lysisVec.append(i)    \n",
    "            elif any(substring in i for substring in regulationModification):\n",
    "                rmVec.append(i)\n",
    "            elif any(substring in i for substring in translation):\n",
    "                translationVec.append(i)        \n",
    "            elif any(substring in i for substring in metabolismSynth):\n",
    "                metVec.append(i)      \n",
    "            elif any(substring in i for substring in entryInjection):\n",
    "                entryVec.append(i)            \n",
    "            elif any(substring in i for substring in ligase):\n",
    "                ligaseVec.append(i)\n",
    "            elif any(substring in i for substring in structural):\n",
    "                structuralVec.append(i)\n",
    "            elif any(substring in i for substring in recomb):\n",
    "                recombVec.append(i)\n",
    "            elif any(substring in i for substring in nuclear):\n",
    "                nuclearVec.append(i)\n",
    "            elif any(substring in i for substring in poly):\n",
    "                polyVec.append(i)\n",
    "            elif any(substring in i for substring in hyp):\n",
    "                hypVec.append(i)\n",
    "            else:       \n",
    "                otherVec.append(i)\n",
    "    protDict={'capsidd':capsidVec, 'matrixd': matrixVec, 'envd': envVec, 'portald': portalVec, 'taild': tailVec, 'structurald': structuralVec,'polymerased':polymeraseVec, \n",
    "              'replicased':replicaseVec, 'rmd':rmVec,'translationd':translationVec, 'helicased':helicaseVec,'integrased': integraseVec,'ligased':ligaseVec,'nucleased': nucleaseVec, 'recombd': recombVec,'metd':metVec,'entryd': entryVec,'nucleard': nuclearVec,'packingd': packingVec, 'lysisd':lysisVec, 'polyd': polyVec,'hypd': hypVec, 'otherd': otherVec}\n",
    "\n",
    "    return(protDict)\n",
    "\n",
    "#putting all dictionaries in a list\n",
    "protDictList=[]\n",
    "\n",
    "#indices of proteDictList = 0 RNA, 1 ssDNA, 2 dsDNA, 3 euk dsDNA, 4 bact dsDNA, 5 bact sipho, 6 bact myo, 7 bact podo, and then a bunch of eukaryotic dsDNA VIRUSES\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'dsRNA viruses' in  v['englishLineage'] or 'ssRNA positive-strand viruses, no DNA stage' in v['englishLineage']\n",
    "         or 'ssRNA negative-strand viruses' in v['englishLineage'] or 'Metaviridae' in v['englishLineage'] or 'Retroviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'ssDNA viruses' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'dsDNA viruses, no RNA stage' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'dsDNA viruses, no RNA stage' in v['englishLineage']] \n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] and 'dsDNA viruses, no RNA stage' in v['englishLineage']] \n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] and 'Siphoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] and 'Myoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Bacteria' in v['hostEnglishLineage'] and 'Podoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'Baculoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'Poxviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'Herpesvirales' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'Papillomaviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'Adenoviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "protMat= [v['proteinNames'] for v in lengthData if 'Eukaryota' in v['hostEnglishLineage'] and 'Polyomaviridae' in v['englishLineage']]\n",
    "protDictList.append(proteinBar(protMat))\n",
    "\n",
    "def getXYs(ind):     \n",
    "    #ind is a vector containing the list of indices (viral groups) we're interested in plotting together\n",
    "    #labels is a vector of strings that describe the viral groups being plotted\n",
    "    bar_width = 0.17\n",
    "    #m is the number of groups we're plotting \n",
    "    m = len(ind)\n",
    "    #the multiplier for bar_width so the bars can be placed next to each other\n",
    "    n= 0\n",
    "    #starting and updating the index of labels that we need to pull out. Corresponds to the labe of the viral group we're plotting\n",
    "    n2 =-1\n",
    "    #yvec will contain the number of proteins associated with each protein category, and all viral groups \n",
    "    yvec =[]\n",
    "    \n",
    "    for i in ind:\n",
    "    #i corresponds to the indices of the protein groups we want to plot from protDictList\n",
    "        n=n+1\n",
    "        n2 =n2 +1\n",
    "        #initializing the x and y coordinates for each bar plot\n",
    "        y=[]\n",
    "        x=[]\n",
    "    \n",
    "        for key in protDictList[i]: \n",
    "            #going through different protein categories (keys) to build the x vector labels\n",
    "            #the x values are just the number of categories, multiplied by n (which keeps getting updated) so the bar plots don't overlap\n",
    "            x.append(key)\n",
    "            #counting the number of items associated with each protein category (or key) to build the y vector\n",
    "            y.append(len(protDictList[i].get(key)))    \n",
    "        \n",
    "        xvec = list(range(len(y)))\n",
    "        xvalnew = [n + xval * m for xval in xvec] \n",
    "        yvec.append(y)\n",
    "\n",
    "\n",
    "        plt.bar(xvalnew, y, align='center')#,color=colvec[n2], label =labvec[n2])\n",
    "        plt.xticks(xvalnew, x)\n",
    "    plt.ylim(0,5000)\n",
    "    plt.show()\n",
    "    return(x, yvec)\n",
    "\n",
    "\n",
    "#calling the function to output x and y values for plotting\n",
    "getXYs(list(range(14)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
