{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Percent noncoding DNA/RNA across all viruses with known hosts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Gita/anaconda/lib/python3.4/site-packages/matplotlib/__init__.py:872: UserWarning: axes.color_cycle is deprecated and replaced with axes.prop_cycle; please use the latter.\n",
      "  warnings.warn(self.msg_depr % (key, alt_key))\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import ete3 as ete\n",
    "import os\n",
    "import Bio\n",
    "import random\n",
    "import scipy.stats as stats\n",
    "from functools import *\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "import matplotlib.patches as mpatches\n",
    "import seaborn as sns\n",
    "\n",
    "\n",
    "#filename that will be the name of both the stat text file and the plot. \n",
    "filename = 'percentNoncoding'\n",
    "inputFileNameTXT = filename + '.txt'\n",
    "fo = open(inputFileNameTXT, 'w')\n",
    "#writing the first line here, and the rest as part of the for loop \n",
    "fo.write('\\t'+'Noncoding DNA Statistics For Different Viral Groups (% noncoding DNA)'+'\\n')\n",
    "fo.write('\\t'+ 'Min' +'\\t'+ 'Max' + '\\t'+ '25th Percentile'+'\\t'+ 'Median'+ '\\t' + '75th Percentile' +'\\t' + 'Mean' + '\\t' +'Stdev.' +'\\n')\n",
    "\n",
    "#as part of quality control the following viruses had to be omitted from our analysis\n",
    "#they contain various types of misannotations or mislabeled hosts \n",
    "badFolders = ['Acyrthosiphon_pisum_secondary_endosymbiont_phage_1_uid14047', 'Aureococcus_anophagefferens_virus_MM_2014_uid258005', 'Dragonfly_associated_microphage_1_uid177547', 'Enterobacter_phage_PG7_uid240014', 'Escherichia_phage_vB_EcoM_JS09_uid248321', 'Lactobacillus_phage_LL_H_uid19803', 'Providencia_phage_Redjac_uid177540', 'Pseudoalteromonas_phage_RIO_1_uid206039', 'Sputnik_virophage_2_uid243067', 'Sputnik_virophage_3_uid243065', 'Staphylococcus_phage_6ec_uid253318', 'Streptococcus_phage_DCC1738_uid253219', 'Streptococcus_phage_IC1_uid253220', 'Streptococcus_phage_K13_uid253223', 'Streptococcus_phage_SMP_uid18529', 'Vibrio_phage_CP_T1_uid181062', 'Vibrio_phage_vpms1_uid212709', 'Vibrio_phage_VPUSM_8_uid227006',  'Abaca_bunchy_top_virus_uid28697','Shallot_yellow_stripe_virus_uid15745','Equine_rhinitis_A_virus_uid15205','African_pouched_rat_arterivirus_uid274797','Spilanthes_yellow_vein_virus_uid19779', 'Velvet_bean_severe_mosaic_virus_uid41175','Paramecium_bursaria_Chlorella_virus_FR483_uid18305','Human_endogenous_retrovirus_K113_uid222261','Acholeplasma_phage_MV_L1_uid14573']\n",
    "#Another group of vviruses with misannotations, this time in their taxonomy (they don't belong to any Baltimore group)\n",
    "unknowns= ['Bacillus_phage_phBC6A51_uid15021', 'Bacillus_phage_phBC6A52_uid15022', 'Badger_feces_associated_gemycircularvirus_uid281125', 'Chaetoceros_DNA_virus_7_uid237558', 'Chaetoceros_lorenzianus_DNA_Virus_uid63565', 'Chaetoceros_tenuissimus_DNA_virus_uid60753', 'Cladosporium_cladosporioides_virus_1_uid258308', 'Enterobacteria_phage_YYZ_2008_uid32231', 'European_mountain_ash_ringspot_associated_virus_uid39973', 'Faecal_associated_gemycircularvirus_3_uid268557', 'Faecal_associated_gemycircularvirus_4_uid268556', 'Faecal_associated_gemycircularvirus_5_uid268555', 'Faecal_associated_gemycircularvirus_6_uid268554', 'Faecal_associated_gemycircularvirus_7_uid268553', 'Gentian_ovary_ring_spot_virus_uid256090', 'Geobacillus_virus_E2_uid19797', 'Glossina_pallidipes_salivary_gland_hypertrophy_virus_uid28839', 'Haloarcula_hispanica_icosahedral_virus_2_uid109269', 'Haloarcula_phage_SH1_uid15535', 'Halovirus_PH1_uid196975', 'Halovirus_VNH_1_uid262927', 'Human_genital_associated_circular_DNA_virus_1_uid281399', 'Lactobacillus_johnsonii_prophage_Lj771_uid28145', 'Magnaporthe_oryzae_RNA_virus_uid272442', 'Mongoose_feces_associated_gemycircularvirus_a_uid281407', 'Mongoose_feces_associated_gemycircularvirus_b_uid281406', 'Mongoose_feces_associated_gemycircularvirus_c_uid281404', 'Mongoose_feces_associated_gemycircularvirus_d_uid281405', 'Mycoplasma_phage_phiMFV1_uid14387', 'Penicillium_roqueforti_dsRNA_mycovirus_1_uid258162', 'Phage_Gifsy_1_uid32269', 'Phage_Gifsy_2_uid32271', 'Pigeonpea_cryptic_virus_uid244664', 'Planaria_asexual_strain_specific_virus_like_element_type_1_uid14140', 'Pseudomonas_phage_phi_Pto_bp6g_uid240724', 'Rhizoctonia_fumigata_mycovirus_uid283068', 'Rhodococcus_phage_REQ2_uid81171', 'Rhodococcus_phage_REQ3_uid81175', 'Rose_rosette_virus_uid64937', 'Rosellinia_necatrix_fusarivirus_1_uid255787', 'Rosellinia_necatrix_megabirnavirus_1_W779_uid41609', 'Salisaeta_icosahedral_phage_1_uid167575', 'Salmonella_phage_Fels_1_uid29267', 'Sodalis_phage_phiSG1_uid16583', 'Staphylococcus_phage_phi2958PVL_uid32173', 'Staphylococcus_phage_tp310_1_uid20659', 'Staphylococcus_phage_tp310_3_uid20663', 'Stenotrophomonas_phage_phiSMA9_uid15493', 'Streptococcus_phage_20617_uid239271', 'Streptococcus_phage_phiBHN167_uid227353', 'Streptococcus_pyogenes_phage_315_1_uid14533', 'Streptococcus_pyogenes_phage_315_2_uid14528', 'Streptococcus_pyogenes_phage_315_3_uid14529', 'Streptococcus_pyogenes_phage_315_4_uid14530', 'Streptococcus_pyogenes_phage_315_5_uid14531', 'Streptococcus_pyogenes_phage_315_6_uid14532', 'Tanay_virus_uid246129', 'Thermococcus_prieurii_virus_1_uid84407', 'Thermus_phage_IN93_uid14235', 'Thermus_phage_P23_77_uid40235']\n",
    "badFolders=badFolders+unknowns\n",
    "#these are satellite viruses, which we will exclude from further analysis in this manuscript \n",
    "sats = ['Ageratum_conyzoides_associated_symptomless_alphasatellite_uid259293', 'Ageratum_enation_alphasatellite_uid181994', 'Ageratum_leaf_curl_betasatellite_uid195929', 'Ageratum_leaf_curl_Cameroon_betasatellite_uid36669', 'Ageratum_yellow_leaf_curl_betasatellite_uid14439', 'Ageratum_yellow_vein_China_alphasatellite_uid237561', 'Ageratum_yellow_vein_China_virus_associated_DNA_beta_uid15515', 'Ageratum_yellow_vein_Singapore_alphasatellite_uid14232', 'Ageratum_yellow_vein_virus_satellite_DNA_beta_uid14444', 'Alternanthera_yellow_vein_virus_satellite_DNA_beta_uid19833', 'Andrographis_yellow_vein_leaf_curl_betasatellite_uid243492', 'Bhendi_yellow_vein_India_betasatellite__India_Aurangabad_OY164_2006__uid61557', 'Bhendi_yellow_vein_mosaic_betasatellite__India_Coimbator_OYCO1_2005__uid61777', 'Bhendi_yellow_vein_mosaic_virus_associated_alphasatellite_uid174781', 'Bhendi_yellow_vein_mosaic_virus_satellite_DNA_beta_uid14445', 'Black_medic_leafroll_alphasatellite_1_uid243500', 'Cardamom_bushy_dwarf_virus_satellite_uid230064', 'Cassava_mosaic_Madagascar_alphasatellite_uid175666', 'Chili_leaf_curl_Bhatinda_betasatellite_uid206467', 'Chilli_leaf_curl_alphasatellite_uid253224', 'Chilli_leaf_curl_Multan_alphasatellite_uid39933', 'Chilli_leaf_curl_virus_satellite_DNA_beta_uid14441', 'Cleome_leaf_crumple_virus_associated_DNA_1_uid60045', 'Corchorus_yellow_vein_mosaic_betasatellite_uid192608', 'Cotton_leaf_curl_Burewala_alphasatellite_uid45935', 'Cotton_leaf_curl_Burewala_betasatellite_uid45933', 'Cotton_leaf_curl_Gezira_alphasatellite_uid42507', 'Cotton_leaf_curl_Gezira_virus_satellite_DNA_beta_uid15166', 'Cotton_leaf_curl_Multan_betasatellite_uid15780', 'Cotton_leaf_curl_virus_associated_DNA_1_isolate_Lucknow_uid65305', 'Cotton_leaf_curl_virus_associated_DNA_beta_uid14438', 'Cotton_leaf_curl_virus_betasatellite_uid162497', 'Cowpea_severe_leaf_curl_associated_DNA_beta_uid15157', 'Croton_yellow_vein_mosaic_alphasatellite_uid45931', 'Croton_yellow_vein_mosaic_betasatellite_uid18249', 'Cuban_alphasatellite_1_uid210798', 'Dragonfly_associated_alphasatellite_uid181244', 'Emilia_yellow_vein_virus_associated_DNA_beta_uid37893', 'Erectites_yellow_mosaic_virus_satellite_DNA_beta_uid19827', 'Eupatorium_yellow_vein_virus_satellite_DNA_beta_uid14447', 'Faba_bean_necrotic_stunt_alphasatellite_1_uid243499', 'Faba_bean_necrotic_stunt_alphasatellite_2_uid243498', 'French_bean_leaf_curl_betasatellite_Kanpur_uid169556', 'Gossypium_darwinii_symptomless_alphasatellite_uid39593', 'Gossypium_davidsonii_symptomless_alphasatellite_uid39589', 'Gossypium_mustilinum_symptomless_alphasatellite_uid39591', 'Grapevine_satellite_virus_uid208539', 'Guar_leaf_curl_alphasatellite_uid193981', 'Hedyotis_uncinella_yellow_mosaic_betasatellite_uid230991', 'Honeysuckle_yellow_vein_mosaic_disease_associated_satellite_DNA_beta_uid19863', 'Honeysuckle_yellow_vein_mosaic_virus_satellite_DNA_beta_uid14620', 'Leucas_zeylanica_yellow_vein_virus_satellite_DNA_beta_uid41305', 'Ludwigia_leaf_distortion_betasatellite__India_Amadalavalasa_Hibiscus_2007__uid29233', 'Ludwigia_yellow_vein_virus_associated_DNA_beta_uid15561', 'Luffa_puckering_and_leaf_distortion_associated_betasatellite__India_Gurdaspur_Okra_2013__uid253218', 'Luffa_puckering_and_leaf_distortion_associated_DNA_beta_uid15779', 'Malachra_yellow_vein_mosaic_virus_associated_satellite_DNA_beta_uid28727', 'Malvastrum_leaf_curl_betasatellite_uid16301', 'Malvastrum_leaf_curl_Guangdong_betasatellite_uid243827', 'Malvastrum_yellow_vein_Yunnan_virus_satellite_DNA_beta_uid14567', 'Melon_chlorotic_mosaic_virus_associated_alphasatellite_uid51413', 'Mesta_yellow_vein_mosaic_virus_alphasatellite__India_Hoshiarpur_Okra_2013__uid253217', 'Mesta_yellow_vein_mosaic_virus_associated_alphasatellite_uid174780', 'Mesta_yellow_vein_mosaic_virus_associated_DNA_beta_uid21015', 'Milk_vetch_dwarf_C10_alphasatellite_uid240682', 'Mirabilis_leaf_curl_India_virus_associated_betasatellite_uid255800', 'Mungbean_yellow_mosaic_India_virus_associated_betasatellite__India__Faizabad__Cow_Pea_2012__uid177773', 'Nanovirus_like_particle_uid14386', 'Okra_enation_leaf_curl_betasatellite__India_Sonipat_EL10_2006__uid61781', 'Okra_leaf_curl_Mali_virus_satellite_DNA_beta_uid20323', 'Okra_yellow_vein_disease_associated_sequence_uid14443', 'Olive_viral_satellite_RNA_uid230268', 'Papaya_leaf_curl_alphasatellite_uid233414', 'Papaya_leaf_curl_China_virus_satellite_DNA_beta_uid19819', 'Pepper_leaf_curl_virus_satellite_DNA_beta_uid28283', 'Pepper_leaf_curl_Yunnan_virus_satellite_DNA_beta_uid29415', 'Potato_apical_leaf_curl_disease_associated_satellite_DNA_beta_uid18323', 'Radish_leaf_curl_virus_betasatellite_uid28281', 'Sida_leaf_curl_virus_satellite_DNA_beta_uid19823', 'Sida_yellow_mosaic_virus_China_associated_DNA_beta_uid15514', 'Sida_yellow_vein_China_alphasatellite_uid238950', 'Siegesbeckia_yellow_vein_virus_associated_DNA_beta_uid17269', 'Subterranean_clover_stunt_C6_alphasatellite_uid14180', 'Synedrella_leaf_curl_alphasatellite_uid263034', 'Tobacco_leaf_chlorosis_betasatellite_uid178075', 'Tobacco_leaf_curl_disease_associated_sequence_uid14442', 'Tobacco_leaf_curl_PUSA_alphasatellite_uid56023', 'Tobacco_leaf_curl_virus_associated_DNA_beta_uid45925', 'Tomato_leaf_curl_Bangladesh_betasatellite_uid56017', 'Tomato_leaf_curl_Gandhinagar_betasatellite_uid231683', 'Tomato_leaf_curl_Hajipur_betasatellite_uid175587', 'Tomato_leaf_curl_Joydebpur_virus_satellite_DNA_beta_uid28273', 'Tomato_leaf_curl_Karnataka_alphasatellite_uid181995', 'Tomato_leaf_curl_Pakistan_virus_associated_DNA_1_uid38463', 'Tomato_leaf_curl_Patna_betasatellite_uid36541', 'Tomato_leaf_curl_Togo_betasatellite_Togo_2006__uid60629', 'Tomato_leaf_curl_Yemen_betasatellite_uid177643', 'Tomato_yellow_leaf_curl_China_betasatellite_uid181248', 'Tomato_yellow_leaf_curl_Vietnam_virus_satellite_DNA_beta_uid19829', 'TYLCCNV_Y322__satellite_DNA_beta_uid16338', 'Vernonia_yellow_vein_betasatellite_uid41303', 'Vernonia_yellow_vein_Fujian_virus_alphasatellite_uid72145', 'Vernonia_yellow_vein_Fujian_virus_betasatellite_uid72143', 'Zinnia_leaf_curl_disease_associated_sequence_uid14440', 'Zinnia_leaf_curl_virus_associated_DNA_beta_uid14538']\n",
    "\n",
    "\n",
    "def getTaxidFromGenomeFolder(genome_folder):\n",
    "    #Takes an rpt file for a genome and returns the taxid in it.  \n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        with open(os.path.join('all_rpt', folder, os.listdir(os.path.join('all_rpt', folder))[0])) as rpt_file:\n",
    "            for line in rpt_file:\n",
    "                if line.startswith('Taxid: '):\n",
    "                    return int(line[7:-1])\n",
    "                \n",
    "\n",
    "def getHostTaxidFromGenomeFolder(genome_folder):\n",
    "    #Takes a folder containing an asn file and returns the taxid of the host of the virus described by that asn file.\n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        with open('folderToHostTaxid.txt', 'r') as taxid2host:\n",
    "            for line in taxid2host:\n",
    "                if genome_folder in line:\n",
    "                    return int(line.split('\\t')[1][:-1])\n",
    "        return 0\n",
    "    \n",
    "\n",
    "def getGenomeLengthFromGenomeFolder(genome_folder):\n",
    "    total = 0\n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        for file in os.listdir(os.path.join('all_ptt', folder)):\n",
    "            with open(os.path.join('all_ptt', folder, file)) as ptt_file:\n",
    "                line = ptt_file.readline()\n",
    "                index = line.find('- 1..')\n",
    "                if index == -1:\n",
    "                    return -1\n",
    "                num = line[index + 5 : -1]\n",
    "                total += int(num)\n",
    "        return total\n",
    "    \n",
    "\n",
    "def getCodingNumberAndOverlapNumberFromGenomeFolder(folder):\n",
    "    totalCoding = 0\n",
    "    totalOverlap = 0\n",
    "    for file in os.listdir(os.path.join('all_ptt', folder)):\n",
    "        #these folders contain genomes with misannotation of gene overlap,non-coding DNA, or host information from NCBI. \n",
    "        if folder not in badFolders and folder not in sats:\n",
    "            with open(os.path.join('all_ptt', folder, file)) as ptt_file:\n",
    "                line = ptt_file.readline()\n",
    "                index = line.find('- 1..')\n",
    "                repliconLength = int(line[index + 5 : -1])\n",
    "                indicesThatOverlap = [0] * repliconLength\n",
    "                line = ptt_file.readline()\n",
    "                line = ptt_file.readline()\n",
    "\n",
    "                uniqueIndices = set([])\n",
    "                for line in ptt_file:\n",
    "                    line = line.split('\\t')[0]\n",
    "                    [start, end] = [int(x) for x in line.split('..')]\n",
    "                    #this is put in place to distinguish linear from circular genomes. Here, we're looking for linear genomes\n",
    "                    if start < end:                    \n",
    "                        for i in range(start, end):\n",
    "                            if i in uniqueIndices:\n",
    "                                indicesThatOverlap[i] = 1\n",
    "                            else:\n",
    "                                uniqueIndices.add(i)\n",
    "                    #here were are looking for circular genomes to be able to properly count noncoding bases\n",
    "                    elif start > end:\n",
    "                        for i in range(start, repliconLength):\n",
    "                            if i in uniqueIndices:\n",
    "                                indicesThatOverlap[i] = 1\n",
    "                            else:\n",
    "                                uniqueIndices.add(i)\n",
    "                        for i in range(1, end):\n",
    "                            if i in uniqueIndices:\n",
    "                                indicesThatOverlap[i] = 1\n",
    "                            else:\n",
    "                                uniqueIndices.add(i)\n",
    "                           \n",
    "                totalCoding += len(uniqueIndices)\n",
    "                #totalOverlap += len(bigList) - len(uniqueIndices)\n",
    "                totalOverlap += sum(indicesThatOverlap)\n",
    "    return (totalCoding, totalOverlap)\n",
    "\n",
    "\n",
    "\n",
    "def makePlot(virusGroups, hostNames, plotPositionVert):\n",
    "    \"\"\"\n",
    "    Makes a plot by plotting the current values of virusGroups and adding a legend\n",
    "    using hostNames.\n",
    "    \"\"\"\n",
    "    #plotPositionVert will let the program know where in the 7x3 subplots to place the figures that are generated. \n",
    "    #noncodingByGroup is a list of lists containing noncoding percentage per virus group across different classifications\n",
    "    noncodingByGroup = [[x['percentNonCoding'] for x in group] for group in virusGroups]\n",
    "    #yVec is normalized frequency. Basically 1/N, where N is the number of viruses per viral group. \n",
    "    yVec = [len([x['percentNonCoding'] for x in group]) for group in virusGroups]     \n",
    "    \n",
    "    #boxWidth is the width of the box plot\n",
    "    boxWidth = 0.05 \n",
    "    #matOfOnes will be a list of lists, containing ones. It has the same structure as noncodingByGroup\n",
    "    matOfOnes=[]\n",
    "    #matWeight will hold 1/N values. \n",
    "    matWeight=[]\n",
    "    #ylimitVec will hold max histogram bar height, so the plot y limit can be determined based off of it. \n",
    "    \n",
    "    counter = range(len(yVec))\n",
    "    #per each classification system, we're looping through viral groups, which correspond to j. \n",
    "    for j in counter:\n",
    "        vecOfOnes= np.zeros(yVec[j]) + 1.\n",
    "        matOfOnes.append(vecOfOnes)\n",
    "        matWeight.append([k / yVec[j] for k in vecOfOnes])   \n",
    "        \n",
    "        #we want to make sure that the nucleotide classification plots end up at the 15th and 18th subplot positions\n",
    "        if plotPositionVert is 4:\n",
    "            if j is 0:\n",
    "                plotPositionHor = 14\n",
    "            elif j is 1: \n",
    "                plotPositionHor = 17\n",
    "            else:\n",
    "                plotPositionHor = 20\n",
    "            ax1 = plt.subplot(7,3, plotPositionHor)\n",
    "        \n",
    "        #all other subplots are placed automatically according to their plotPositionVert, which is specified later in the code\n",
    "        else:\n",
    "            plotPositionHor = 3*j + plotPositionVert\n",
    "            ax1 = plt.subplot(7,3, plotPositionHor)\n",
    "        \n",
    "        #calculating the number of bins needed according to Freedman-Diaconis rule. Bin width = 2 * IQR * N^(-1/3), where N is number of datapoints, IQR is interquartile range. \n",
    "        #number of bins can be calculated from bin width according to bin number = (max(data) - min(data))/bin width\n",
    "        iqr=np.percentile(noncodingByGroup[j], 75, interpolation='higher') - np.percentile(noncodingByGroup[j], 25, interpolation='lower')\n",
    "        bottom = binWidth=2*iqr*len(noncodingByGroup[j])**(-1/3)\n",
    "        top = max(noncodingByGroup[j])-min(noncodingByGroup[j])\n",
    "        binNumFreed =int(np.ceil(top/bottom))\n",
    "        \n",
    "        #plotting the normalized histograms, and ensuring bins of equal length by setting 'bins' equal to an array with equal spacings\n",
    "        n, bins, patches = plt.hist(noncodingByGroup[j], weights=matWeight[j], bins=binNumFreed, color=colors[j], linewidth=0)\n",
    "        maxHeight= np.round(np.amax(n),1) + 0.2\n",
    "        \n",
    "        #instead of a legend, since there is only one figure per plot, I'm annotating the plot with virus group name, and number of viruses per that group \n",
    "        ax1.annotate(hostNames[j] + ' (N = ' + str(yVec[j]) + ')', xy=(0.95, 0.8), xycoords='axes fraction',\n",
    "                horizontalalignment='right', verticalalignment='top', fontsize=11, color='dimgrey')   \n",
    "        \n",
    "        #drawing the boxplots. \n",
    "        bp = plt.boxplot(noncodingByGroup[j], showmeans=True, sym='', widths=boxWidth, vert=False, \n",
    "                         boxprops=dict(color=colors[j]),positions=[maxHeight-0.05],\n",
    "                         capprops=dict(markerfacecolor=colors[j]),medianprops=dict(color=colors[j]), \n",
    "                         meanprops=dict(markerfacecolor=colors[j]), whis='range')        \n",
    "        plt.setp(bp['whiskers'], color=colors[j])\n",
    "        plt.setp(bp['caps'], color=colors[j])       \n",
    "              \n",
    "        maxBarHeight = maxHeight - 0.2\n",
    "        xmin = 0\n",
    "        #xmax is set to 25 for shortrange\n",
    "        xmax = 40\n",
    "        numticksy=2\n",
    "        numticksx=5\n",
    "        \n",
    "        plt.xlim(xmin,xmax)        \n",
    "        plt.ylim(0,maxHeight)\n",
    "        ax1.set_yticks(np.linspace(0,maxBarHeight,numticksy))\n",
    "        ax1.set_yticklabels(np.linspace(0,maxBarHeight, numticksy), color='dimgrey' )\n",
    "        ax1.set_xticks(np.linspace(xmin,xmax,numticksx*2-1))\n",
    "            \n",
    "        #The following lines are made so that I can have more ticks but everyother label is skipped.     \n",
    "        labelList= np.linspace(xmin,xmax,numticksx)\n",
    "        #a list of zeros as placeholders\n",
    "        newLabelList = [0] * (len(labelList) * 2 - 1)\n",
    "        #making a list of original tick labels interspaced with zeros. In the code, I make zero labels invisible. \n",
    "        newLabelList[0::2] = labelList\n",
    "        \n",
    "        ax1.set_xticklabels(newLabelList,color='dimgrey')\n",
    "        plt.tick_params(axis='both', which='both', labelsize=12) \n",
    "        \n",
    "        #making the zero xlabels invisible. They were just placeholders. \n",
    "        for label in ax1.set_xticklabels(newLabelList)[1::2]:\n",
    "            label.set_visible(False)        \n",
    "        #this piece is written so that the x axis of all plots except the very last don't get labeled. \n",
    "        if j in counter[:-1]:\n",
    "            ax1.set_xticklabels([])\n",
    "        \n",
    "        #just added to be able to perserve text when taking the plot image into illustrator version 6\n",
    "        plt.rcParams['pdf.fonttype'] = 42  \n",
    "        \n",
    "        #writing the stats to the file we started earlier \n",
    "        minimum =\"%.1f\" %min(noncodingByGroup[j])\n",
    "        maximum= \"%.1f\" %max(noncodingByGroup[j])\n",
    "        iqr1 = \"%.1f\" %np.percentile(noncodingByGroup[j], 25, interpolation='lower')\n",
    "        med =\"%.1f\" %np.median(noncodingByGroup[j])\n",
    "        iqr3 = \"%.1f\" %(np.percentile(noncodingByGroup[j], 75, interpolation='higher'))\n",
    "        avg =\"%.1f\" %np.mean(noncodingByGroup[j])\n",
    "        std =\"%.1f\" %np.std(noncodingByGroup[j])\n",
    "        \n",
    "        fo.write(hostNames[j]+' (N = ' + str(yVec[j]) + ')' +'\\t'+ minimum +'\\t'+ maximum + '\\t'+ iqr1 + '\\t'+ med+ '\\t' + iqr3 + '\\t' + avg +'\\t'+ std + '\\n')\n",
    "\n",
    "genomeLengths = []\n",
    "lengthData = []\n",
    "counter = 0\n",
    "length_lists = []\n",
    "hostTaxidList = []\n",
    "taxidList = []\n",
    "folders = []\n",
    "dbObject = ete.NCBITaxa()\n",
    "for folder in os.listdir('all_faa'):\n",
    "#these entries had problems with their gene overlap, protein, or noncoding annotation, need to be excluded    \n",
    "    if folder not in badFolders and folder not in sats:\n",
    "        length_lists.append([])\n",
    "        for fileName in os.listdir(os.path.join('all_faa', folder)):\n",
    "            with open(os.path.join('all_faa', folder, fileName), \"r\") as faa_file:\n",
    "                proteins = SeqIO.parse(faa_file, 'fasta')\n",
    "                for protein in proteins:\n",
    "                    length_lists[counter].append(float(len(protein.seq)))\n",
    "\n",
    "        hostTaxid = getHostTaxidFromGenomeFolder(folder)\n",
    "        #viruses with unidentified hosts. Now they are out. \n",
    "        if hostTaxid and hostTaxid != 11086 and hostTaxid != 10036 and hostTaxid != 37965 and hostTaxid != 10640  and hostTaxid != 32644 and hostTaxid != 1 and hostTaxid != 212035 and hostTaxid != 1051671:\n",
    "            taxidList.append(getTaxidFromGenomeFolder(folder))\n",
    "            hostTaxidList.append(hostTaxid)\n",
    "            genomeLengths.append(getGenomeLengthFromGenomeFolder(folder))\n",
    "            folders.append(folder)\n",
    "            counter += 1\n",
    "        else:\n",
    "            length_lists = length_lists[:-1]\n",
    "\n",
    "        #Uncomment the following lines to see the code's progress while its running        \n",
    "        \"\"\"\n",
    "        if counter % 100 == 0:\n",
    "            print(str(int(np.round(float(counter) * 100.0 / 2701.0))) + '% done.')\n",
    "        \"\"\"\n",
    "\n",
    "taxidLineages = [dbObject.get_lineage(t) for t in taxidList]\n",
    "englishLineages = [[str(rank) for rank in dbObject.translate_to_names(lin)]\n",
    "                   for lin in taxidLineages]\n",
    "hostTaxidLineages = [dbObject.get_lineage(t) for t in hostTaxidList]\n",
    "hostEnglishLineages = [[str(rank) for rank in dbObject.translate_to_names(lin)] \n",
    "                       for lin in hostTaxidLineages]\n",
    "\n",
    "for i in range(len(length_lists)):\n",
    "    lengthData.append({'length_list' : length_lists[i],\n",
    "                       'taxid' : taxidList[i],\n",
    "                       'hostTaxid' : hostTaxidList[i],\n",
    "                       'englishLineage' : englishLineages[i],\n",
    "                       'taxidLineage' : taxidLineages[i],\n",
    "                       'hostEnglishLineage' : hostEnglishLineages[i],\n",
    "                       'hostTaxidLineage' : hostTaxidLineages[i],\n",
    "                       'mean' : np.mean(length_lists[i]),\n",
    "                       'std' : np.std(length_lists[i]),\n",
    "                       'median' : np.median(length_lists[i]),\n",
    "                       'numProteins' : len(length_lists[i]),\n",
    "                       'genomeLength' : genomeLengths[i],\n",
    "                       'folder' : folders[i]})\n",
    "                       \n",
    "lengthData = [virus for virus in lengthData if virus['genomeLength'] != -1]\n",
    "\n",
    "for (i, virus) in enumerate(lengthData):\n",
    "    #print(str(i) + '/' + str(len(lengthData)))\n",
    "    (totalCoding, totalOverlap) = getCodingNumberAndOverlapNumberFromGenomeFolder(virus['folder'])\n",
    "    virus['percentNonCoding'] = 100 * (virus['genomeLength'] - totalCoding) / virus['genomeLength']\n",
    "    virus['percentOverlap'] = 100 * totalOverlap / virus['genomeLength']\n",
    "\n",
    "\n",
    "plt.figure(1, figsize=(15, 9))\n",
    "\n",
    "#Make the host-classified plot\n",
    "hostNames = 'Eukaryota;Bacteria;Archaea'.split(';')\n",
    "virusGroups = [[genome for genome in lengthData if taxon in genome['hostEnglishLineage']]\n",
    "               for taxon in hostNames]\n",
    "plotPositionVert =4\n",
    "hostNames = [x + ' Viruses' for x in hostNames]\n",
    "colors = ['salmon','cornflowerblue','teal']\n",
    "makePlot(virusGroups, hostNames, plotPositionVert)\n",
    "\n",
    "    \n",
    "#Make the Baltimore-classified plot\n",
    "virusGroups = []\n",
    "virusGroups.append([g for g in lengthData if 'dsDNA viruses, no RNA stage' in g['englishLineage']])\n",
    "virusGroups.append([g for g in lengthData if 'ssDNA viruses' in g['englishLineage']])\n",
    "virusGroups.append([g for g in lengthData if 'dsRNA viruses' in g['englishLineage']])\n",
    "virusGroups.append([g for g in lengthData if 'ssRNA positive-strand viruses, no DNA stage'\n",
    "                 in g['englishLineage']])\n",
    "virusGroups.append([g for g in lengthData if 'ssRNA negative-strand viruses' in g['englishLineage']])\n",
    "virusGroups.append([g for g in lengthData if 'Metaviridae' in g['englishLineage']\n",
    "                    or 'Retroviridae' in g['englishLineage']])\n",
    "virusGroups.append([g for g in lengthData if 'Caulimoviridae' in g['englishLineage']\n",
    "                    or 'Hepadnaviridae' in g['englishLineage']])\n",
    "\n",
    "hostNames = ['Group I (dsDNA)', 'Group II (ssDNA)', 'Group III (dsRNA)', 'Group IV (+ssRNA)',\n",
    "             'Group V (-ssRNA)', 'Group VI (ssRNA-RT)', 'Group VII (dsDNA-RT)', 'Unclassified']\n",
    "plotPositionVert =1\n",
    "colors =  ['gold', 'gold', 'maroon', 'maroon', 'maroon', 'maroon', 'gold']\n",
    "makePlot(virusGroups, hostNames,plotPositionVert)\n",
    "\n",
    "    \n",
    "#Make the Nucleic-Acid classified plot\n",
    "newVirusGroups = []\n",
    "newVirusGroups.append(virusGroups[0] + virusGroups[1] + virusGroups[6])\n",
    "newVirusGroups.append(virusGroups[2] + virusGroups[3] + virusGroups[4] + virusGroups[5])\n",
    "DNAviruses = newVirusGroups[0]\n",
    "RNAviruses = newVirusGroups[1]\n",
    "#ax3 = plt.subplot(223)\n",
    "plotPositionVert = 2\n",
    "colors = ['gold','maroon']\n",
    "makePlot(newVirusGroups, ['DNA Viruses', 'RNA Viruses'],plotPositionVert)\n",
    "\n",
    "\n",
    "#Make the ICTV classified plot\n",
    "hostNames = ['Caudovirales', 'Herpesvirales', 'Ligamenvirales', 'Mononegavirales', 'Nidovirales',\n",
    "             'Picornavirales', 'Tymovirales']\n",
    "virusGroups = [[genome for genome in lengthData if taxon in genome['englishLineage']]\n",
    "               for taxon in hostNames]\n",
    "\n",
    "colors = [\"navy\"]*7\n",
    "plotPositionVert =3\n",
    "makePlot(virusGroups, hostNames, plotPositionVert)\n",
    "\n",
    "plt.savefig(filename + '.pdf')\n",
    "fo.close()\n",
    "\n",
    "plt.show() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#creating a function to plot the histograms of various sub viral groups. \n",
    "#numcol and numrow refer to the number of rows and columns to be generated for subplots. Colors refers to the color used for each subplot.\n",
    "#allViruses is a list of lists, containing relevant data to be plotted for different viral groups\n",
    "def histPlots(allViruses, hostNames, colors, numcol, numrow):\n",
    "    input_var = input(\"Enter file name: \")\n",
    "    inputFileName = input_var + '.pdf'\n",
    "\n",
    "    plt.figure(1, figsize=(15, 9))   \n",
    "    #boxWidth is the width of the box plot\n",
    "    boxWidth = 0.05 \n",
    "    #matWeight will hold 1/N values. \n",
    "    matWeight=[]\n",
    "       \n",
    "    counter = range(len(allViruses))\n",
    "    #per each classification system, we're looping through viral groups, which correspond to j. \n",
    "    for j in counter:\n",
    "        #a vector of ones the length of allViruses[j] which is a genome length vector of one of the three domains of life, depending on j. \n",
    "        vecOfOnes= np.zeros(len(allViruses[j])) + 1.\n",
    "        \n",
    "        #matOfOnes.append(vecOfOnes)\n",
    "        matWeight.append([k / len(allViruses[j]) for k in vecOfOnes]) \n",
    "        \n",
    "        #need 3 rows, 1 column for this current plot\n",
    "        ax1 = plt.subplot(numcol,numrow, j+1)\n",
    "        \n",
    "        #calculating the number of bins needed according to Freedman-Diaconis rule. Bin width = 2 * IQR * N^(-1/3), where N is number of datapoints, IQR is interquartile range. \n",
    "        #number of bins can be calculated from bin width according to bin number = (max(data) - min(data))/bin width  \n",
    "        iqr=np.percentile(allViruses[j], 75, interpolation='higher') - np.percentile(allViruses[j], 25, interpolation='lower')\n",
    "        bottom =binWidth=2*iqr*len(allViruses[j])**(-1/3)\n",
    "        top = max(allViruses[j])- min(allViruses[j])\n",
    "        \n",
    "        if iqr > 0:            \n",
    "            binNumFreed =int(np.ceil(top/bottom))\n",
    "            #plotting the normalized histograms, and ensuring bins of equal length by setting 'bins' equal to an array with equal spacings          \n",
    "            n, bins, patches = plt.hist(allViruses[j], weights=matWeight[j], bins = binNumFreed, color=colors[j], linewidth=0)\n",
    "            maxHeight= np.round(np.amax(n),1) + 0.2\n",
    "            \n",
    "            \n",
    "        elif iqr ==0 or max(allViruses[j])==0: \n",
    "            # this step is to ensure that we don't get infinity while calculating the number of bins (see binNumFreed)\n",
    "            n, bins, patches = plt.hist(allViruses[j], weights=matWeight[j],bins=5, color=colors[j], linewidth=0)\n",
    "            maxHeight= np.round(np.amax(n),1) + 0.2\n",
    "            \n",
    "\n",
    "        ax1.annotate(hostNames[j] + ' (N = ' + str(len(allViruses[j])) + ')', xy=(0.95, 0.8), xycoords='axes fraction',\n",
    "                horizontalalignment='right', verticalalignment='top', fontsize=11, color='dimgrey')   \n",
    "\n",
    "        #drawing the boxplots. \n",
    "        bp = plt.boxplot(allViruses[j], showmeans=True, sym='', widths=boxWidth, vert=False, \n",
    "                            boxprops=dict(color=colors[j]),positions=[maxHeight-0.05],\n",
    "                            capprops=dict(markerfacecolor=colors[j]),medianprops=dict(color=colors[j]), \n",
    "                            meanprops=dict(markerfacecolor=colors[j]), whis='range')        \n",
    "        plt.setp(bp['whiskers'], color=colors[j])\n",
    "        plt.setp(bp['caps'], color=colors[j]) \n",
    "                \n",
    "        maxBarHeight = maxHeight - 0.2\n",
    "        \n",
    "        #the scale values used to create num proteins plot\n",
    "        \"\"\"    \n",
    "        xmin = 0\n",
    "        xmax = 2\n",
    "        numticksy=2\n",
    "        numticksx=3\n",
    "        \"\"\"\n",
    "        \n",
    "        xmin = 0\n",
    "        #set to 25 for short-range\n",
    "        xmax =40\n",
    "        numticksy=2\n",
    "        numticksx=5\n",
    "        \n",
    "                \n",
    "        plt.xlim(xmin,xmax)        \n",
    "        plt.ylim(0,maxHeight)\n",
    "        ax1.set_yticks(np.linspace(0,maxBarHeight,numticksy))\n",
    "        ax1.set_yticklabels(np.linspace(0,maxBarHeight, numticksy), color='dimgrey' )\n",
    "        ax1.set_xticks(np.linspace(xmin,xmax,numticksx*2-1))\n",
    "            \n",
    "        #The following lines are made so that I can have more ticks but everyother label is skipped.     \n",
    "        labelList= np.linspace(xmin,xmax,numticksx)\n",
    "        #a list of zeros as placeholders\n",
    "        newLabelList = [0] * (len(labelList) * 2 - 1)\n",
    "        #making a list of original tick labels interspaced with zeros. In the code, I make zero labels invisible. \n",
    "        newLabelList[0::2] = labelList\n",
    "        \n",
    "        ax1.set_xticklabels(newLabelList,color='dimgrey')\n",
    "        plt.tick_params(axis='both', which='both', labelsize=12) \n",
    "        \n",
    "        #making the zero xlabels invisible. They were just placeholders. \n",
    "        for label in ax1.set_xticklabels(newLabelList)[1::2]:\n",
    "            label.set_visible(False)        \n",
    "        #this piece is written so that the x axis of all plots except the very last don't get labeled. \n",
    "        if j in counter[:-1]:\n",
    "            ax1.set_xticklabels([])                \n",
    "                \n",
    "        #just added to be able to perserve text when taking the plot image into illustrator version 6    \n",
    "        plt.rcParams['pdf.fonttype'] = 42 \n",
    "\n",
    "    plt.savefig(inputFileName)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Noncoding percentages of eukaryotic dsDNA viral groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter file name: dsDNAeukNoncodingPercentage\n",
      "Enter file name: dsDNAeukNoncodingPercentage\n"
     ]
    }
   ],
   "source": [
    "virusesdsDNAEUKS = [virus['percentNonCoding'] for virus in lengthData if 'dsDNA viruses, no RNA stage' in virus['englishLineage'] and 'Eukaryota' in virus['hostEnglishLineage']]\n",
    "virusesBacu = [virus['percentNonCoding'] for virus in lengthData if 'Baculoviridae' in virus['englishLineage']]\n",
    "virusesPoxv = [virus['percentNonCoding'] for virus in lengthData if 'Poxviridae' in virus['englishLineage']]\n",
    "virusesHerp = [virus['percentNonCoding'] for virus in lengthData if 'Herpesvirales' in virus['englishLineage']]\n",
    "virusesPapi = [virus['percentNonCoding'] for virus in lengthData if 'Papillomaviridae' in virus['englishLineage']]\n",
    "virusesAden = [virus['percentNonCoding'] for virus in lengthData if 'Adenoviridae' in virus['englishLineage']]\n",
    "virusesPoly = [virus['percentNonCoding'] for virus in lengthData if 'Polyomaviridae' in virus['englishLineage']]\n",
    "\n",
    "allViruses= [virusesdsDNAEUKS, virusesBacu,virusesPoxv,virusesHerp,virusesPapi,virusesAden,virusesPoly]\n",
    "hostNames= ['All Eukaryotic dsDNA viruses', 'Baculoviridae', 'Poxviridae', 'Herpesvirales', 'Papillomaviridae', 'Adenoviridae', 'Polyomaviridae']\n",
    "colors=['lightcoral', 'lightcoral','lightcoral','lightcoral','lightcoral','lightcoral','lightcoral']\n",
    "numrow = 7\n",
    "numcol = 3\n",
    "histPlots(allViruses, hostNames, colors, numrow, numcol)\n",
    "\n",
    "input_var = input(\"Enter file name: \")\n",
    "filename = input_var+ '.txt'\n",
    "fo = open(filename, 'w')\n",
    "#writing the first line here, and the rest as part of the for loop \n",
    "fo.write('\\t'+'Noncoding DNA Statistics For Different Viral Groups (% of the genome that is noncoding)'+'\\n')\n",
    "fo.write('\\t'+ 'Min' +'\\t'+ 'Max' + '\\t'+ '25th Percentile'+'\\t'+ 'Median'+ '\\t' + '75th Percentile' +'\\t' + 'Mean' + '\\t' +'Stdev.' +'\\n')\n",
    "\n",
    "for j in range(len(allViruses)):  \n",
    "    #writing the stats to the file we started earlier \n",
    "    minimum =\"%.1f\" %min(allViruses[j])\n",
    "    maximum= \"%.1f\" %max(allViruses[j])\n",
    "    iqr1 = \"%.1f\" %np.percentile(allViruses[j], 25, interpolation='lower')\n",
    "    med =\"%.1f\" %np.median(allViruses[j])\n",
    "    iqr3 = \"%.1f\" %(np.percentile(allViruses[j], 75, interpolation='higher'))\n",
    "    avg =\"%.1f\" %np.mean(allViruses[j])\n",
    "    std =\"%.1f\" %np.std(allViruses[j])\n",
    "    fo.write(hostNames[j]+' (N = ' + str(len(allViruses[j])) + ')' +'\\t'+ minimum +'\\t'+ maximum + '\\t'+ iqr1 + '\\t'+ med+ '\\t' + iqr3 + '\\t' + avg +'\\t'+ std + '\\n')\n",
    "\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter file name: allvirusesabsNoncoding\n",
      "Enter file name: allvirusesabsNoncoding\n"
     ]
    }
   ],
   "source": [
    "#plotting all viruses\n",
    "viruses = [virus['percentNonCoding'] for virus in lengthData ]\n",
    "allViruses= [viruses]\n",
    "hostNames= ['All viruses']\n",
    "colors=['grey']\n",
    "numrow = 7\n",
    "numcol = 3\n",
    "histPlots(allViruses, hostNames, colors, numrow, numcol)\n",
    "\n",
    "input_var = input(\"Enter file name: \")\n",
    "filename = input_var+ '.txt'\n",
    "fo = open(filename, 'w')\n",
    "#writing the first line here, and the rest as part of the for loop \n",
    "fo.write('\\t'+'Noncoding DNA Statistics For Different Viral Groups (% of the genome that is noncoding)'+'\\n')\n",
    "fo.write('\\t'+ 'Min' +'\\t'+ 'Max' + '\\t'+ '25th Percentile'+'\\t'+ 'Median'+ '\\t' + '75th Percentile' +'\\t' + 'Mean' + '\\t' +'Stdev.' +'\\n')\n",
    "\n",
    "for j in range(len(allViruses)):  \n",
    "    #writing the stats to the file we started earlier \n",
    "    minimum =\"%.1f\" %min(allViruses[j])\n",
    "    maximum= \"%.1f\" %max(allViruses[j])\n",
    "    iqr1 = \"%.1f\" %np.percentile(allViruses[j], 25, interpolation='lower')\n",
    "    med =\"%.1f\" %np.median(allViruses[j])\n",
    "    iqr3 = \"%.1f\" %(np.percentile(allViruses[j], 75, interpolation='higher'))\n",
    "    avg =\"%.1f\" %np.mean(allViruses[j])\n",
    "    std =\"%.1f\" %np.std(allViruses[j])\n",
    "    fo.write(hostNames[j]+' (N = ' + str(len(allViruses[j])) + ')' +'\\t'+ minimum +'\\t'+ maximum + '\\t'+ iqr1 + '\\t'+ med+ '\\t' + iqr3 + '\\t' + avg +'\\t'+ std + '\\n')\n",
    "\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Noncoding percentages of bacterial dsDNA viral groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter file name: dsDNAbacNoncodingPercentage\n",
      "Enter file name: dsDNAbacNoncodingPercentage\n"
     ]
    }
   ],
   "source": [
    "virusesdsDNABACT = [virus['percentNonCoding'] for virus in lengthData if 'dsDNA viruses, no RNA stage' in virus['englishLineage'] and 'Bacteria' in virus['hostEnglishLineage']]\n",
    "virusesCaudo = [virus['percentNonCoding'] for virus in lengthData if 'Caudovirales' in virus['englishLineage'] and 'Bacteria' in virus['hostEnglishLineage']]\n",
    "virusesSiph = [virus['percentNonCoding'] for virus in lengthData if 'Caudovirales' in virus['englishLineage'] and 'Siphoviridae' in virus['englishLineage']]\n",
    "virusesPodo = [virus['percentNonCoding'] for virus in lengthData if 'Caudovirales' in virus['englishLineage'] and 'Podoviridae' in virus['englishLineage']]\n",
    "virusesMyov = [virus['percentNonCoding'] for virus in lengthData if 'Caudovirales' in virus['englishLineage'] and 'Myoviridae' in virus['englishLineage']]\n",
    "virusesCaudoUncl = [virus['percentNonCoding'] for virus in lengthData if 'Caudovirales' in virus['englishLineage'] and 'unclassified Caudovirales' in virus['englishLineage']]\n",
    "virusesUncl = [virus['percentNonCoding'] for virus in lengthData if 'unclassified dsDNA phages' in virus['englishLineage'] and 'Bacteria' in virus['hostEnglishLineage']]\n",
    "\n",
    "\n",
    "allViruses= [virusesdsDNABACT, virusesCaudo,virusesSiph,virusesPodo,virusesMyov, virusesCaudoUncl, virusesUncl]\n",
    "\n",
    "hostNames= ['All Bacterial dsDNA viruses', 'Bacterial Caudovirales', 'Bacterial Caudovirales-Siphoviridae', 'Bacterial Caudovirales-Podoviridae', 'Bacterial Caudovirales-Myoviridae', 'Bacterial Caudovirales-Unclassified','Unclassified Bacterial Phages']\n",
    "colors=['cornflowerblue', 'cornflowerblue','cornflowerblue','cornflowerblue','cornflowerblue','cornflowerblue','cornflowerblue']\n",
    "numrow = 7\n",
    "numcol = 3\n",
    "\n",
    "histPlots(allViruses, hostNames, colors, numrow, numcol)\n",
    "\n",
    "input_var = input(\"Enter file name: \")\n",
    "filename = input_var+ '.txt'\n",
    "fo = open(filename, 'w')\n",
    "#writing the first line here, and the rest as part of the for loop \n",
    "fo.write('\\t'+'Noncoding DNA Statistics For Different Viral Groups (% of the genome that is noncoding)'+'\\n')\n",
    "fo.write('\\t'+ 'Min' +'\\t'+ 'Max' + '\\t'+ '25th Percentile'+'\\t'+ 'Median'+ '\\t' + '75th Percentile' +'\\t' + 'Mean' + '\\t' +'Stdev.' +'\\n')\n",
    "\n",
    "for j in range(len(allViruses)):  \n",
    "    #writing the stats to the file we started earlier \n",
    "    minimum =\"%.1f\" %min(allViruses[j])\n",
    "    maximum= \"%.1f\" %max(allViruses[j])\n",
    "    iqr1 = \"%.1f\" %np.percentile(allViruses[j], 25, interpolation='lower')\n",
    "    med =\"%.1f\" %np.median(allViruses[j])\n",
    "    iqr3 = \"%.1f\" %(np.percentile(allViruses[j], 75, interpolation='higher'))\n",
    "    avg =\"%.1f\" %np.mean(allViruses[j])\n",
    "    std =\"%.1f\" %np.std(allViruses[j])\n",
    "    fo.write(hostNames[j]+' (N = ' + str(len(allViruses[j])) + ')' +'\\t'+ minimum +'\\t'+ maximum + '\\t'+ iqr1 + '\\t'+ med+ '\\t' + iqr3 + '\\t' + avg +'\\t'+ std + '\\n')\n",
    "\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Noncoding percentages of archaeal dsDNA viral groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter file name: dsDNAarcNoncodingPercentage\n",
      "Enter file name: dsDNAarcNoncodingPercentage\n"
     ]
    }
   ],
   "source": [
    "virusesdsDNAArch = [virus['percentNonCoding'] for virus in lengthData if 'dsDNA viruses, no RNA stage' in virus['englishLineage'] and 'Archaea' in virus['hostEnglishLineage']]\n",
    "virusesCaudo = [virus['percentNonCoding'] for virus in lengthData if 'Caudovirales' in virus['englishLineage'] and 'Archaea' in virus['hostEnglishLineage']]\n",
    "virusesLiga = [virus['percentNonCoding'] for virus in lengthData if 'Ligamenvirales' in virus['englishLineage'] and 'Archaea' in virus['hostEnglishLineage']]\n",
    "virusesFuse = [virus['percentNonCoding'] for virus in lengthData if 'Fuselloviridae' in virus['englishLineage'] and 'Archaea' in virus['hostEnglishLineage']]\n",
    "virusesUncl = [virus['percentNonCoding'] for virus in lengthData if 'unclassified dsDNA viruses' in virus['englishLineage'] and 'Archaea' in virus['hostEnglishLineage']]\n",
    "\n",
    "allViruses= [virusesdsDNAArch, virusesCaudo,virusesLiga, virusesFuse, virusesUncl]\n",
    "\n",
    "hostNames= ['All Archaeal dsDNA viruses', 'Archaeal Caudovirales', 'Ligamenvirales', 'Fuselloviridae', 'Archaeal dsDNA viruses- unclassified']\n",
    "colors=['teal','teal','teal','teal','teal']\n",
    "numrow = 7\n",
    "numcol = 3\n",
    "\n",
    "histPlots(allViruses, hostNames, colors, numrow, numcol)\n",
    "\n",
    "input_var = input(\"Enter file name: \")\n",
    "filename = input_var+ '.txt'\n",
    "fo = open(filename, 'w')\n",
    "#writing the first line here, and the rest as part of the for loop \n",
    "fo.write('\\t'+'Noncoding DNA Statistics For Different Viral Groups (% of the genome that is noncoding)'+'\\n')\n",
    "fo.write('\\t'+ 'Min' +'\\t'+ 'Max' + '\\t'+ '25th Percentile'+'\\t'+ 'Median'+ '\\t' + '75th Percentile' +'\\t' + 'Mean' + '\\t' +'Stdev.' +'\\n')\n",
    "\n",
    "for j in range(len(allViruses)):  \n",
    "    #writing the stats to the file we started earlier \n",
    "    minimum =\"%.1f\" %min(allViruses[j])\n",
    "    maximum= \"%.1f\" %max(allViruses[j])\n",
    "    iqr1 = \"%.1f\" %np.percentile(allViruses[j], 25, interpolation='lower')\n",
    "    med =\"%.1f\" %np.median(allViruses[j])\n",
    "    iqr3 = \"%.1f\" %(np.percentile(allViruses[j], 75, interpolation='higher'))\n",
    "    avg =\"%.1f\" %np.mean(allViruses[j])\n",
    "    std =\"%.1f\" %np.std(allViruses[j])\n",
    "    fo.write(hostNames[j]+' (N = ' + str(len(allViruses[j])) + ')' +'\\t'+ minimum +'\\t'+ maximum + '\\t'+ iqr1 + '\\t'+ med+ '\\t' + iqr3 + '\\t' + avg +'\\t'+ std + '\\n')\n",
    "\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Noncoding percentages of eukaryotic ssDNA viral groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter file name: ssDNAeukNoncodingPercentage\n",
      "Enter file name: ssDNAeukNoncodingPercentage\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "a=['Nanoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Circoviridae','unclassified','Parvoviridae','Parvoviridae','Parvoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Circoviridae','Geminiviridae','Parvoviridae','Geminiviridae','Circoviridae','Geminiviridae','Circoviridae','Parvoviridae','Circoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Nanoviridae','Geminiviridae','Geminiviridae','Parvoviridae','Geminiviridae','Parvoviridae','Anelloviridae','Circoviridae','Circoviridae','Geminiviridae','unclassified','Geminiviridae','Geminiviridae','Parvoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','unclassified','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Parvoviridae','unclassified','unclassified','Circoviridae','Circoviridae','Circoviridae','Circoviridae','Circoviridae','Geminiviridae','Parvoviridae','Geminiviridae','Geminiviridae','Parvoviridae','Geminiviridae','Geminiviridae','Geminiviridae','unclassified','unclassified','unclassified','Geminiviridae','unclassified','unclassified','unclassified','Circoviridae','Circoviridae','Circoviridae','Circoviridae','unclassified','unclassified','unclassified','unclassified','unclassified','unclassified','unclassified','unclassified','unclassified','unclassified','unclassified','Circoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Parvoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Nanoviridae','Parvoviridae','Parvoviridae','Circoviridae','Parvoviridae','Circoviridae','Geminiviridae','Geminiviridae','Circoviridae','unclassified','Circoviridae','Parvoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Circoviridae','Circoviridae','Circoviridae','Circoviridae','Parvoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Parvoviridae','Parvoviridae','Parvoviridae','Circoviridae','Circoviridae','Circoviridae','Parvoviridae','Parvoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Circoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Circoviridae','Geminiviridae','Parvoviridae','Parvoviridae','Parvoviridae','Parvoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Parvoviridae','Geminiviridae','Geminiviridae','Nanoviridae','Nanoviridae','Circoviridae','Parvoviridae','Parvoviridae','Geminiviridae','Geminiviridae','Geminiviridae','unclassified','Parvoviridae','Parvoviridae','Parvoviridae','Parvoviridae','Parvoviridae','Circoviridae','Parvoviridae','Parvoviridae','Parvoviridae','Parvoviridae','unclassified','unclassified','unclassified','Circoviridae','Circoviridae','unclassified','unclassified','unclassified','unclassified','unclassified','Geminiviridae','Geminiviridae','Parvoviridae','Geminiviridae','Geminiviridae','Circoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','unclassified','unclassified','Anelloviridae','Anelloviridae','Anelloviridae','Anelloviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Anelloviridae','Anelloviridae','Anelloviridae','Anelloviridae','Geminiviridae','Parvoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Circoviridae','Geminiviridae','Geminiviridae','Circoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Anelloviridae','Anelloviridae','Anelloviridae','Anelloviridae','Anelloviridae','Anelloviridae','Anelloviridae','Anelloviridae','Anelloviridae','Anelloviridae','Anelloviridae','unclassified','Geminiviridae','Geminiviridae','Parvoviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Geminiviridae','Circoviridae']\n",
    "b =[[x,a.count(x)] for x in set(a)]\n",
    "print(b)\n",
    "\"\"\"\n",
    "virusesssDNAEUKS = [virus['percentNonCoding'] for virus in lengthData if 'ssDNA viruses' in virus['englishLineage'] and 'Eukaryota' in virus['hostEnglishLineage']]\n",
    "virusesGemi = [virus['percentNonCoding'] for virus in lengthData if 'Geminiviridae' in virus['englishLineage'] and 'Eukaryota' in virus['hostEnglishLineage']]\n",
    "virusesCirc = [virus['percentNonCoding'] for virus in lengthData if 'Circoviridae' in virus['englishLineage'] and 'Eukaryota' in virus['hostEnglishLineage']]\n",
    "virusesParv = [virus['percentNonCoding'] for virus in lengthData if 'Parvoviridae' in virus['englishLineage'] and 'Eukaryota' in virus['hostEnglishLineage']]\n",
    "virusesAnel = [virus['percentNonCoding'] for virus in lengthData if 'Anelloviridae' in virus['englishLineage'] and 'Eukaryota' in virus['hostEnglishLineage']]\n",
    "virusesNano = [virus['percentNonCoding'] for virus in lengthData if 'Nanoviridae' in virus['englishLineage'] and 'Eukaryota' in virus['hostEnglishLineage']]\n",
    "virusesUncl = [virus['percentNonCoding'] for virus in lengthData if 'unclassified ssDNA viruses' in virus['englishLineage'] and 'Eukaryota' in virus['hostEnglishLineage']]\n",
    "\n",
    "allViruses= [virusesssDNAEUKS, virusesGemi,virusesCirc,virusesParv,virusesAnel,virusesNano,virusesUncl]\n",
    "hostNames= ['All Eukaryotic ssDNA viruses', 'Geminiviridae', 'Circoviridae', 'Parvoviridae', 'Anelloviridae', 'Nanoviridae', 'Unclassified Eukaryotic ssDNA viruses']\n",
    "colors=['lightcoral', 'lightcoral','lightcoral','lightcoral','lightcoral','lightcoral','lightcoral']\n",
    "numrow = 7\n",
    "numcol = 3\n",
    "histPlots(allViruses, hostNames, colors, numrow, numcol)\n",
    "\n",
    "input_var = input(\"Enter file name: \")\n",
    "filename = input_var+ '.txt'\n",
    "fo = open(filename, 'w')\n",
    "#writing the first line here, and the rest as part of the for loop \n",
    "fo.write('\\t'+'Noncoding DNA Statistics For Different Viral Groups (% of the genome that is noncoding)'+'\\n')\n",
    "fo.write('\\t'+ 'Min' +'\\t'+ 'Max' + '\\t'+ '25th Percentile'+'\\t'+ 'Median'+ '\\t' + '75th Percentile' +'\\t' + 'Mean' + '\\t' +'Stdev.' +'\\n')\n",
    "\n",
    "for j in range(len(allViruses)):  \n",
    "    #writing the stats to the file we started earlier \n",
    "    minimum =\"%.1f\" %min(allViruses[j])\n",
    "    maximum= \"%.1f\" %max(allViruses[j])\n",
    "    iqr1 = \"%.1f\" %np.percentile(allViruses[j], 25, interpolation='lower')\n",
    "    med =\"%.1f\" %np.median(allViruses[j])\n",
    "    iqr3 = \"%.1f\" %(np.percentile(allViruses[j], 75, interpolation='higher'))\n",
    "    avg =\"%.1f\" %np.mean(allViruses[j])\n",
    "    std =\"%.1f\" %np.std(allViruses[j])\n",
    "    fo.write(hostNames[j]+' (N = ' + str(len(allViruses[j])) + ')' +'\\t'+ minimum +'\\t'+ maximum + '\\t'+ iqr1 + '\\t'+ med+ '\\t' + iqr3 + '\\t' + avg +'\\t'+ std + '\\n')\n",
    "\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Noncoding percentages of bacterial ssDNA viral groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter file name: ssDNAbacNoncodingPercentage\n",
      "Enter file name: ssDNAbacNoncodingPercentage\n"
     ]
    }
   ],
   "source": [
    "#a = [[virus['englishLineage'],virus['hostEnglishLineage']] for virus in lengthData if 'ssDNA viruses' in virus['englishLineage'] and 'Bacteria' in virus['hostEnglishLineage']]\n",
    "\n",
    "virusesssDNABact = [virus['percentNonCoding'] for virus in lengthData if 'ssDNA viruses' in virus['englishLineage'] and 'Bacteria' in virus['hostEnglishLineage']]\n",
    "virusesMicr = [virus['percentNonCoding'] for virus in lengthData if 'Microviridae' in virus['englishLineage'] and 'Bacteria' in virus['hostEnglishLineage']]\n",
    "virusesInov = [virus['percentNonCoding'] for virus in lengthData if 'Inoviridae' in virus['englishLineage'] and 'Bacteria' in virus['hostEnglishLineage']]\n",
    "\n",
    "allViruses= [virusesssDNABact, virusesMicr,virusesInov]\n",
    "hostNames= ['All Bacterial ssDNA viruses', 'Microviridae', 'Inoviridae']\n",
    "colors=['cornflowerblue', 'cornflowerblue','cornflowerblue']\n",
    "numrow = 7\n",
    "numcol = 3\n",
    "histPlots(allViruses, hostNames, colors, numrow, numcol)\n",
    "\n",
    "input_var = input(\"Enter file name: \")\n",
    "filename = input_var+ '.txt'\n",
    "fo = open(filename, 'w')\n",
    "#writing the first line here, and the rest as part of the for loop \n",
    "fo.write('\\t'+'Noncoding DNA Statistics For Different Viral Groups (% of the genome that is noncoding)'+'\\n')\n",
    "fo.write('\\t'+ 'Min' +'\\t'+ 'Max' + '\\t'+ '25th Percentile'+'\\t'+ 'Median'+ '\\t' + '75th Percentile' +'\\t' + 'Mean' + '\\t' +'Stdev.' +'\\n')\n",
    "\n",
    "for j in range(len(allViruses)):  \n",
    "    #writing the stats to the file we started earlier \n",
    "    minimum =\"%.1f\" %min(allViruses[j])\n",
    "    maximum= \"%.1f\" %max(allViruses[j])\n",
    "    iqr1 = \"%.1f\" %np.percentile(allViruses[j], 25, interpolation='lower')\n",
    "    med =\"%.1f\" %np.median(allViruses[j])\n",
    "    iqr3 = \"%.1f\" %(np.percentile(allViruses[j], 75, interpolation='higher'))\n",
    "    avg =\"%.1f\" %np.mean(allViruses[j])\n",
    "    std =\"%.1f\" %np.std(allViruses[j])\n",
    "    fo.write(hostNames[j]+' (N = ' + str(len(allViruses[j])) + ')' +'\\t'+ minimum +'\\t'+ maximum + '\\t'+ iqr1 + '\\t'+ med+ '\\t' + iqr3 + '\\t' + avg +'\\t'+ std + '\\n')\n",
    "\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###testing statistical significance \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H: dsDNA euk has greater percentage of noncoding than dsDNA bac MannwhitneyuResult(statistic=163976.0, pvalue=2.6467182369338602e-18)\n",
      "H: dsDNA arc has greater percentage of noncoding than dsDNA bac MannwhitneyuResult(statistic=21524.0, pvalue=0.034388131859286346)\n",
      "H: dsDNA euk has greater percentage of noncoding than dsDNA arc MannwhitneyuResult(statistic=6655.0, pvalue=0.020602856798038065)\n",
      "H: ssDNA euk has greater percentage of noncoding than ssDNA bac MannwhitneyuResult(statistic=11145.0, pvalue=0.027573939400795663)\n",
      "H: DNA viruses have greater percentage of noncoding than RNA MannwhitneyuResult(statistic=879787.0, pvalue=1.8143054459067447e-70)\n",
      "H: ssDNA viruses have greater percentage of noncoding than dsDNA MannwhitneyuResult(statistic=389966.0, pvalue=7.1408666682628568e-53)\n"
     ]
    }
   ],
   "source": [
    "dsDNAeuks= [virus['percentNonCoding'] for virus in lengthData if 'dsDNA viruses, no RNA stage' in virus['englishLineage'] and 'Eukaryota' in virus['hostEnglishLineage']]\n",
    "dsDNAbact= [virus['percentNonCoding'] for virus in lengthData if 'dsDNA viruses, no RNA stage' in virus['englishLineage'] and 'Bacteria' in virus['hostEnglishLineage']]\n",
    "dsDNAarchs= [virus['percentNonCoding'] for virus in lengthData if 'dsDNA viruses, no RNA stage' in virus['englishLineage'] and 'Archaea' in virus['hostEnglishLineage']]\n",
    "ssDNAbact= [virus['percentNonCoding'] for virus in lengthData if 'ssDNA viruses' in virus['englishLineage'] and 'Bacteria' in virus['hostEnglishLineage']]\n",
    "ssDNAeuks= [virus['percentNonCoding'] for virus in lengthData if 'ssDNA viruses' in virus['englishLineage'] and 'Eukaryota' in virus['hostEnglishLineage']]\n",
    "\n",
    "#For comparing DNA and RNA viruses. Note, variables DNAviruses and RNAviruses was defined in the first script at the very top\n",
    "rna = [virus['percentNonCoding'] for virus in RNAviruses]\n",
    "dna = [virus['percentNonCoding'] for virus in DNAviruses]\n",
    "#for comparing ssDNA and dsDNA viruses\n",
    "ssdna = [virus['percentNonCoding'] for virus in DNAviruses if 'ssDNA viruses' in virus['englishLineage']]\n",
    "dsdna = [virus['percentNonCoding'] for virus in DNAviruses if 'dsDNA viruses, no RNA stage' in virus['englishLineage']]\n",
    "\n",
    "be = stats.mannwhitneyu(dsDNAeuks, dsDNAbact, alternative='greater')\n",
    "ba = stats.mannwhitneyu(dsDNAarchs, dsDNAbact, alternative='greater')\n",
    "ea = stats.mannwhitneyu(dsDNAeuks, dsDNAarchs, alternative='greater')\n",
    "beSS = stats.mannwhitneyu(ssDNAeuks, ssDNAbact, alternative='greater')\n",
    "dnarna = stats.mannwhitneyu(dna,rna, alternative='greater')\n",
    "ssds = stats.mannwhitneyu(ssdna,dsdna, alternative='greater')\n",
    "\n",
    "print('H: dsDNA euk has greater percentage of noncoding than dsDNA bac', be)\n",
    "print('H: dsDNA arc has greater percentage of noncoding than dsDNA bac',ba)\n",
    "print('H: dsDNA euk has greater percentage of noncoding than dsDNA arc',ea)\n",
    "print('H: ssDNA euk has greater percentage of noncoding than ssDNA bac',beSS)\n",
    "print('H: DNA viruses have greater percentage of noncoding than RNA',dnarna)\n",
    "print('H: ssDNA viruses have greater percentage of noncoding than dsDNA',ssds)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
